--- 
title: "R community analysis"
author: "Matthew R. Gemmell"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
favicon: figures/NEOF_favicon.png
description: NEOF book for the R community analysis workflow
cover-image: "figures/NEOF.png"
---

```{r, echo=FALSE}
#Change colour, border, and text of code chunks
#Check style.css for .Rchunk
#https://stackoverflow.com/questions/65627531/change-r-chunk-background-color-in-bookdown-gitbook
#https://bookdown.org/yihui/rmarkdown-cookbook/chunk-styling.html
knitr::opts_chunk$set(class.source="Rchunk") 
```

```{r cite-packages, include = FALSE}
# automatically create a bib database for R packages
# add any packages you want to cite here
knitr::write_bib(c(
  .packages(), 'bookdown', 'webexercises'
), 'packages.bib')
```

<center>
![](figures/NEOF.png){style="border-radius: 15px; width: 300px"}
</center>

# (PART\*) Intro {-}

# Introduction
<center>
![](figures/R_community.png){style="width:200px"}
</center>


A lot of different analyses and visualisations can be carried out with community data. This includes taxonomy and functional abundance tables from 16S rRNA and Shotgun metagenomics analysis. This workshop will teach you how to use R with the `phyloseq` R object; a specialised object containing an abundance, taxonomy, and metadata table. 

The workshop will use a 16S dataset that has been pre-analysed with QIIME2 to create the ASV table, taxonomy table, and phylogenetic tree. [Supplementary materials](https://neof-workshops.github.io/R_community_whqkt8/Supplemental/) will show how to import Bracken shotgun metagenomic abundance data and generic abundance data frames into a phyloseq object.
 
The sessions will start with a brief presentation followed by self-paced computer practicals guided by this online interactive book. This book contains theory and practice code. This will be reinforced with multiple choice questions that will recap concepts and aid in interpretation of results.

At the end of the course learners will be able to:

- Import QIIME2 artifacts into a phyloseq object.
- Summarise the abundance and taxonomy contents of a phyloseq object
- Preprocess the abundance and taxonomy tables. This will include transforming sample counts, and subsetting samples & taxonomies.
- Understand the grammar of graphics (ggplot2) used by phyloseq and related packages.
- Carry out alpha & beta diversity, and biomarker detection with the phyloseq object.
- Produce and customise publication quality plots.
- Run statistical analysis.

There are [supplemental](https://neof-workshops.github.io/R_community_whqkt8/Supplemental/) materials including importing other types of data into phyloseq, and the use of R plotly to produce interactive HTML based plots.

## Table of contents {-}

```{r, echo=FALSE, align="centre"}
#Create 2 column dataframe
column_1 <- c(
  #1 intro
  '[**Intro**](#datasetandworkflowchap)', 
 '[![](figures/data.png){height="150"}](#datasetandworkflowchap)',
 #3 Setup
    '[**Set-up**](#setupchap)', 
 '[![](figures/start.png){height="150"}](#setupchap)',
 #5 Data prep
   '[**Data preparation**](#preprocess_section)', 
 '[![](figures/half_tablespoon.png){height="150"}](#preprocess_section)',
 #7 Taxa relabund
   '[**Taxonomy relative abundance**](#taxa_relabund_chap)', 
 '[![](figures/taxa.png){height="150"}](#taxa_relabund_chap)',
 #9 Rarefaction
 '[**Rarefaction**](#rarefaction_chap)', 
 '[![](figures/whale.png){height="150"}](#rarefaction_chap)',
 #11 Beta
  '[**Beta diversity**](#beta_chap)', 
 '[![](figures/beta.png){height="150"}](#beta_chap)',
 #13 Summary
   '[**Summary**](#sumchap)', 
 '[![](figures/recap.png){style="height:150px; border-radius:15px; background:white"}](#sumchap)'
)
column_2 <- c(
  #2: R packages
  '[**R packages**](#rpackchap)', 
 '[![](figures/R.png){height="150"}](#rpackchap)',
 #4 Jupyter
   '[**Jupyter**](#jupyterchap)', 
 '[![](figures/jupyter_logo.png){style="height:150px; border-radius:15px; background:white"}](#jupyterchap)',
 #6 Min reads
   '[**Minimum read depth**](#mindepthchap)', 
 '[![](figures/depth_diver.png){height="150"}](#mindepthchap)',
 #8 taxa plots
 '[**Taxa plots**](#chaptaxaplots)', 
 '[![](figures/bar_chart.png){height="150"}](#chaptaxaplots)',
 #10 ALpha
 '[**Alpha diversity**](#alpha_chap)', 
 '[![](figures/alpha.png){height="150"}](#alpha_chap)',
 #12 DA
  '[**Differential abundance**](#DA_chap)', 
 '[![](figures/difference.png){height="150"}](#DA_chap)',
   #14 appendix
    '[**Appendix**](#mambainstalls)', 
 '[![](figures/mamba_logo.png){style="height:150px; border-radius:15px; background:null"}](#mambainstalls)'
)
table_df <- data.frame(column_1, column_2)
#Kable and kableextra libraries
library(knitr)
library(kableExtra)
#Create table
#ALign both column in centre (align = "cc")
table_df %>%
  kbl(align= "cc", col.names = NULL) %>%
  kable_styling(position = "center")
```

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.