<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="R community analysis" />
<meta property="og:type" content="book" />
<meta property="og:image" content="/figures/NEOF.png" />
<meta property="og:description" content="NEOF book for the R community analysis workflow" />


<meta name="author" content="Matthew R. Gemmell" />

<meta name="date" content="2023-05-09" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="NEOF book for the R community analysis workflow">

<title>R community analysis</title>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<!--bookdown:toc:end-->
<!--bookdown:toc:start-->
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="header">
<h1 class="title">R community analysis</h1>
<h4 class="author"><em>Matthew R. Gemmell</em></h4>
<h4 class="date"><em>2023-05-09</em></h4>
</div>
<p><img src="figures/NEOF.png" width="30%" style="display: block; margin: auto;" /></p>



<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p><img src="figures/R_community.png" width="20%" style="display: block; margin: auto;" /></p>
<p>A lot of different analyses and visualisations can be carried out with community data. This includes taxonomy and functional abundance tables from 16S rRNA and Shotgun metagenomics analysis. This workshop will teach you how to use R with the <code>phyloseq</code> R object; a specialised object containing an abundance, taxonomy, and metadata table.</p>
<p>The workshop will use a 16S dataset that has been pre-analysed with QIIME2 to create the ASV table, taxonomy table, and phylogenetic tree. Supplementary materials will show how to import Bracken shotgun metagenomic abundance data and generic abundance data frames into a phyloseq object.</p>
<p>The sessions will start with a brief presentation followed by self-paced computer practicals guided by this online interactive book. Thi book contains theory and practice code. This will be reinforced with multiple choice questions that will recap concepts and aid in interpretation of results.</p>
<p>At the end of the course learners will be able to:</p>
<ul>
<li>Import QIIME2 artifacts into a phyloseq object.</li>
<li>Summarise the abundance and taxonomy contents of a phyloseq object</li>
<li>Preprocess the abundance and taxonomy tables. This will include transforming sample counts, and subsetting samples &amp; taxonomies.</li>
<li>Understand the grammar of graphics (ggplot2) used by phyloseq and related packages.</li>
<li>Carry out alpha &amp; beta diversity, and biomarker detection with the phyloseq object.</li>
<li>Produce and customise publication quality plots.</li>
<li>Run statistical analysis.</li>
<li>Convert static plots into interactive html plots with plotly within R.</li>
</ul>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.</p>

</div>
<div id="dataset-workflow" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Dataset &amp; workflow</h1>
<p><img src="figures/data.png" width="15%" style="display: block; margin: auto;" /></p>
<div id="dataset" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Dataset</h2>
<p><img src="figures/freshwater_france.png" width="50%" style="display: block; margin: auto;" /></p>
<p>In this tutorial we will be using a 16S metabarcoding dataset derived from surface water from the Durance River in the south-east of France.
Two major comparisons were carried out in combination with each other.</p>
<p><a href="https://www.mdpi.com/2076-2607/8/8/1129">Link to paper</a></p>
<div id="sites" class="section level3" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> Sites</h3>
<p><img src="figures/river.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Three different sites were chosen on the Durance River. These three sites were representative of an anthropisation (transformation of land by humans) gradient along a river stream. These sites were:</p>
<ul>
<li><strong>Upper Durance sampling site (UD)</strong>: Alpine part of the river with little/no anthropisation.</li>
<li><strong>Middle Durance sampling site (MD)</strong>: Upper part of agricultural land dominated by apple and pear production.</li>
<li><strong>Lower Durance sampling site (LD)</strong>: Lower part of agricultural land with intensive production of fruits, cereals, and vegetables.</li>
</ul>
</div>
<div id="culture-media" class="section level3" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> Culture media</h3>
<p><img src="figures/petri.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Surface water was sampled and different culture media were used to produce bacterial lawns for each site. The media used were:</p>
<ul>
<li><strong>Environmental sample (ENV)</strong>: No media used, frozen at -20°C.</li>
<li><strong>TSA 10%</strong> incubated at 28°C for 2 days.</li>
<li><strong>KBC</strong> incubated at 28°C for 2 days.</li>
<li><strong>CVP</strong> incubated at 28°C for 3 days.</li>
</ul>
</div>
<div id="sum_and_qs" class="section level3" number="2.1.3">
<h3><span class="header-section-number">2.1.3</span> Summary &amp; questions</h3>
<p><img src="figures/sum_red.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Each sample and media combination was produced in replicates of three giving a total of 36 samples (3 X 4 X 3 = 36). The three replicates were cultured on three different plates with the same media. An ASV table, taxonomy table, and phylogenetic tree were produced with QIIME2 and DADA2.</p>
<p>With this data we can ask and investigate the following questions:</p>
<ul>
<li>How do the bacterial communities change across the anthropisation gradient?</li>
<li>Is there a difference in the replicates of one site and media combination? I.e. do any of the media produce inconsistent profiles?</li>
<li>Is there more difference between the sites or the media used?</li>
<li>Do the media samples differ from the ENV samples? If so, how?</li>
</ul>
</div>
</div>
<div id="workflow" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Workflow</h2>
<p><img src="figures/workflow.png" width="15%" style="display: block; margin: auto;" /></p>
<ol style="list-style-type: decimal">
<li><a href="#import_chap">Import</a>: Import QIIME2 artifacts into a <code>phyloseq</code> object with <code>qiime2R</code>.</li>
<li><a href="#sum_phyloseq_chap">Summarisations</a>: Check our <code>phyloseq</code> object with summarisations.</li>
<li><a href="#mindepthchap">Minimum depth</a>: Determine the minimum depth we should use and remove samples with lower depth.</li>
<li><a href="#taxa_relabund_chap">Taxanomic relative abundance</a>: Create taxonomic relative abundance tables.</li>
<li><a href="#chaptaxaplots">Taxa plots</a>: Produce heat maps and bar plots of taxa relative abundances.</li>
<li><a href="#family_genus_chap">Family and genus</a>: Using the last step to produce family and genus based taxa plots.</li>
<li><a href="#rarefaction_chap">Rarefaction</a>: Carry out sample depth normalisation with rarefactions. This will be used for alpha and beta diversity analysis.</li>
<li><a href="#alpha_chap">Alpha diversity</a>: Carry out alpha diversity analysis through plots and statistics.</li>
<li><a href="#beta_chap">Beta diversity</a>: Carry out beta diversity analysis through plots and statistics.</li>
<li><a href="#DA_chap">Differenital abundance anlaysis</a>: Detect biomarkers compared to a reference group with ANCOM.</li>
</ol>

</div>
</div>
<div id="r-packages" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> R Packages</h1>
<p><img src="figures/R.png" width="20%" style="display: block; margin: auto;" /></p>
<p>During this workshop we will use various R packages with their own intricacies. Before going into analysis we’ll introduce you to some of these important concepts.</p>
<div id="r-packageslibraries" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> R packages/libraries</h2>
<p><img src="figures/r_package.png" width="20%" style="display: block; margin: auto;" /></p>
<p>R packages/libraries contain additional functions, data and code for analysing, manipulating and plotting different types of data. Many common packages will be installed as default when you install R. Other more specialised packages, such as the <code>ggplot2</code> package, must be installed by the user.</p>
<p>Packages found on The Comprehensive R Archive Network (CRAN), R’s central software repository, can be installed using the following command.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;package_name&quot;</span>)</span></code></pre></div>
<p>Every time you reload R you will need to load the packages you require if they are not installed in R by default. To do this type:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;package_name&quot;</span>)</span></code></pre></div>
<p>I generally have a list of <code>library()</code> functions at the top of my R scripts (<code>.R</code> files) for all the packages I use in the script.</p>
<p>Throughout this course you will get a lot of practice installing and loading various packages.</p>
<div class="webex-solution">
<button>
R package or R Library?
</button>
<p>R packages are a collection of R functions, data, and compiled code. You can install these into a directory on your computer.</p>
<p>An R library is a directory containing a R package.</p>
<p>Because of this, the terms R package and R library may be used synonymously. We will use the term package in this workshop.</p>
</div>
<p>As we will be using a lot of packages we shall use double colons (<code>::</code>) to specify which package each function belongs to, unless the function is from base R. For example if we use the function <code>summarize_phyloseq()</code> from the package <code>microbiome</code> we would type the function like below:</p>
<p><strong>Note</strong>: Do not run the below command.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>()</span></code></pre></div>
<p>This convention has 2 benefits:</p>
<ul>
<li>We can easily tell which R package each function comes from.
<ul>
<li>This is useful for your future coding where you may copy some, but not all, commands from one script to another. You will therefore know which packages you will need to load.</li>
<li>If you need some more documentation about a function you will know what package to look up.</li>
<li>Writing your methods will be a lot easier.</li>
</ul></li>
<li>Different packages may have functions with the same name. Specifying the package will ensure you are using the correct function.</li>
</ul>
</div>
<div id="the-grammar-of-graphics" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> The grammar of graphics</h2>
<p><img src="figures/ggplot2.png" width="20%" style="display: block; margin: auto;" /></p>
<p>During this course we will be using the grammar of graphics coding approach. This approach is implemented by the R package <code>ggplot2</code> to create visualisations such as bar charts, box plots, ordination plots etc. In turn <code>ggplot2</code> is used by a host of other packages, some of which we will be using. Although <code>ggplot2</code> is R code, its structure is very different and it takes effort to learn. Thankfully, <code>ggplot2</code> is very powerful and flexible, and it produces very professional and clean plots.</p>
<p>We will use the <code>iris</code> dataset (inbuilt into R) to show an example of <code>ggplot2</code> code and its visualisation output. You don’t need to run the below code.</p>
<p><strong>Note</strong>: If you would like to see the contents of the <code>iris</code> dataset you can run the command <code>View(iris)</code> in your R instance later.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Load library</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Create new ggplot2 object using iris dataset</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x=</span>Sepal.Length, <span class="at">y=</span>Sepal.Width, <span class="at">colour=</span>Species)) <span class="sc">+</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Make the object a scatter plot </span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Add plot tile</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggtitle</span>(<span class="st">&quot;Iris Sepal length vs width&quot;</span>) <span class="sc">+</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Set x and y axis label names</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Sepal length&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Sepal width&quot;</span>)</span></code></pre></div>
<p><img src="R-community-analysis_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<p>We will not learn <code>ggplot2</code> specifically during this course. However, the structure of creating an object will be used. In the above case the initial object was built with <code>ggplot</code>. Subsequently additions and edits were carried out with <code>+</code> and various other functions.</p>
<p>An important concept of the grammar of graphics is aesthetics. Aesthetics are the parts of a graphic/plot. In the above command we set the aesthetics with the function <code>aes()</code> within the <code>ggplot()</code> function. The X aesthetic (i.e. what values are assigned to the x axis) was set as the Sepal length values from the column <code>Sepal.Length</code> of the dataframe <code>iris</code>. In turn the Y axis values are set to the Sepal width and the colouring of the points are set to the Species.</p>
<p>That was a quick introduction to the grammar of graphics. We will be using this to create visualisations with a <code>phyloseq</code> object using various R packages specifically designed for community abundance data within <code>phyloseq</code> objects.</p>
<p>For more resources on <code>ggplot2</code> please see the <a href="#ggplot2_appendix">appendix</a> of this book.</p>
</div>
<div id="phyloseq" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> phyloseq</h2>
<p><img src="figures/phyloseq_logo.png" width="20%" style="display: block; margin: auto;" /></p>
<p>In this book we will be working with <a href="https://joey711.github.io/phyloseq/"><code>phyloseq</code></a> objects to preprocess our dataset, create visualisations, and carry out statistical analyses. This is a very popular object type for community abundance datasets as it contains the abundance table, metadata, and taxonomy table in one object, optionally containing the phylogenetic tree and reference sequences if wanted/required.</p>
<p><img src="figures/phyloseq_input.png" width="80%" style="display: block; margin: auto;" /></p>
<p>For more info on <code>phyloseq</code> and associated packages please see the <a href="#phyloseq_appendix">appendix</a>.</p>

</div>
</div>
<div id="set-up" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> Set-up</h1>
<p><img src="figures/start.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Prior to any analysis we need to setup our environment in the webVNC.</p>
<div id="cluster" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Logon instructions</h2>
<p>For this workshop we will be using Virtual Network Computing (VNC). Connect to the VNC with a browser by using the webVNC link you were sent.</p>
<p>You will now be in a logged-in Linux VNC desktop.
You will see something as below (there may be only one terminal which is fine).
If you do not see something similar please ask for assistance.</p>
<p><img src="figures/nsc200_logon.png" width="80%" style="display: block; margin: auto;" /></p>
<p>If the VNC is taking up too much/little space of your browser you can use the zoom of your browser to adjust the size.
You will most likely need to use your browser’s tool bar to accomplish this.
Ensure you can see the grey borders.</p>
<p>These instructions will not work outside of this workshop.
If you would like to install your own Linux OS on your desktop or laptop we would recommend Mint Linux</p>
<p>The following link is a guide to install Mint Linux:<br />
<a href="https://linuxmint-installation-guide.readthedocs.io/en/latest/" class="uri">https://linuxmint-installation-guide.readthedocs.io/en/latest/</a></p>
</div>
<div id="mamba" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Mamba</h2>
<p><img src="figures/mamba_logo.png" width="20%" style="display: block; margin: auto;" /></p>
<p>This workshop requires a lot of packages.
These all can be difficult to install with R.
Instead we have used Mamba forge to install R, its packages, and Jupyter-notebook (more info below).
To learn more about Mamba-forge and how to create your own environment please see the <a href="#mamba_install">appendix</a>.</p>
<p>To set-up your environment for this workshop please run the following code (you must include the full stop and space at the front of the command).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span> usercommunity</span></code></pre></div>
<p>You will have successfully activated the environment if you now see <code>(r_community)</code> at the start of your command prompt.
This indicates you are now in the mamba environment called <code>r_community</code> created by the instructor.</p>
<p>If you are interested in the use script you can look at its contents.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> /usr/local/bin/usercommunity</span></code></pre></div>
<p><strong>Tip:</strong> press <code>q</code> to quit <code>less</code>.</p>
<p>For more about mamba and how to create your own <code>r_community</code> environment please see the <a href="#mamba_install">appendix</a></p>

</div>
</div>
<div id="jupyter" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Jupyter</h1>
<p><img src="figures/jupyter_logo.png" width="20%" style="display: block; margin: auto;" /></p>
<p><a href="https://jupyter.org/"><code>Jupyter-notebook</code></a> is a nice browser based method to write, edit, and run code. It was initally created for Python coding, but has since branched out to many other languages, such as <code>R</code>.</p>
<p>We are using it in this workshop for a variety of its properties:</p>
<ul>
<li>It is popular and well maintained.</li>
<li>It is lightweight. Other heavier weight programs, such as RStudio, would struggle in our HPC due to the graphical and CPU load.</li>
<li>It is interactive and displays code output.</li>
<li>It allows for easier annotation, editing, and debugging than the command line.</li>
<li>It provides a graphical interface for changing directories and choosing files.</li>
</ul>
<p>Before carrying out any analysis we will go through a quick tutorial of <code>jupyter-notebook</code>.</p>
<div id="open-jupyter-notebook" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Open Jupyter-notebook</h2>
<p>The first step is to open <code>jupyter-notebook</code>. Run the below command in your <code>(r_community)</code> environment.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">jupyter-notebook</span></span></code></pre></div>
<p>This will open <code>jupyter-notebook</code> in firefox. We won’t need to access the linux terminal anymore. Leave the terminal running <code>jupyter-notebook</code> and full screen your <code>firefox</code> so you should see something like below.</p>
<p><img src="figures/jupyter_notebook_example_1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="create-r-notebook" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Create R notebook</h2>
<p>The next step is to create a R notebook.</p>
<ol style="list-style-type: decimal">
<li>Click on the <strong>“New”</strong> button towards the top right, right of the “Upload” button.</li>
<li>From the dropdown click <strong>“R”</strong> below “Python 3 (ipykernel)”.</li>
</ol>
<p>This will open up a new R notebook like below.</p>
<p><img src="figures/jupyter_notebook_example_2.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="cells-and-code" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Cells and code</h2>
<p><code>Jupyter-notebook</code> uses <strong>cells</strong> (the gray boxes) to separate code. This is very useful to compartmentalise our code.</p>
<p>There will already be one <strong>cell</strong>. Within the <strong>cell</strong>, type in the below commands.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">+</span><span class="dv">1</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="dv">2-3</span></span></code></pre></div>
<p>When pressing enter in <strong>cells</strong> it will create a new line. To run all commands in a <strong>cell</strong> press <code>CTRL + enter</code>.</p>
<p>Run your current <strong>cell</strong> and you should see something like below.</p>
<p><img src="figures/jupyter_notebook_example_3.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="create-new-cells" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Create new cells</h2>
<p>You can create new <strong>cells</strong> by 2 different means.</p>
<ul>
<li>Press the <code>+</code> button on the tool bar (between the floppy disk <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M380.93,57.37A32,32,0,0,0,358.3,48H94.22A46.21,46.21,0,0,0,48,94.22V417.78A46.21,46.21,0,0,0,94.22,464H417.78A46.36,46.36,0,0,0,464,417.78V153.7a32,32,0,0,0-9.37-22.63ZM256,416a64,64,0,1,1,64-64A63.92,63.92,0,0,1,256,416Zm48-224H112a16,16,0,0,1-16-16V112a16,16,0,0,1,16-16H304a16,16,0,0,1,16,16v64A16,16,0,0,1,304,192Z" style="fill:none;stroke:#000;stroke-linecap:round;stroke-linejoin:round;stroke-width:32px"></path></svg> and scissors <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M103.48,224a71.64,71.64,0,0,0,44.76-15.66l41.5,16.89,6.82-12.63a39.15,39.15,0,0,1,4.32-6.37l14.22-14.42-41.17-24.94A72,72,0,1,0,103.48,224Zm0-112a40,40,0,1,1-40,40A40,40,0,0,1,103.48,112Z"></path> <path d="M480,169l-5.52-12.58c-4.48-10.42-14.74-16-32.78-17.85-10.12-1-26.95-1.24-49.69,3.81-20,4.45-122.14,28.2-164.95,58.62C206.81,215.39,203,234.67,200,250.16c-2.78,14.14-5,25.31-18,35-15,11.14-27.27,16.38-33.58,18.6a71.74,71.74,0,1,0,24.79,38ZM255.48,256a16,16,0,1,1,16-16A16,16,0,0,1,255.48,256Zm-152,144a40,40,0,1,1,40-40A40,40,0,0,1,103.48,400Z"></path> <path d="M343.79,259.87l-83.74,48.18,27.63,13.08,3.62,1.74C310,331.92,359.74,356,410.53,359c3.89.23,7.47.34,10.78.34C442,359.31,453,354,459.75,350L480,336Z"></path></svg>). This will add a <strong>cell</strong> below your currently selected <strong>cell</strong>.</li>
<li>Click on the <strong><code>Insert</code></strong> button and use the dropdown to add a cell above or below your currently selected cell.</li>
</ul>
<p><strong>Tip:</strong> Hover over the toolbar icons to display a text based description of its function.</p>
<p>With that knowledge add a second <strong>cell</strong> below the first <strong>cell</strong>. Add the following code to your second <strong>cell</strong> but do not run it.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>num_1 <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>num_2 <span class="ot">&lt;-</span> <span class="dv">10</span></span></code></pre></div>
<p><strong>Tip:</strong> Notice there are green lines around your selected cell.</p>
<p>Insert a third <strong>cell</strong> and add the following code to it. Do not run the code.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>num_1 <span class="sc">*</span> num_2</span></code></pre></div>
</div>
<div id="running-code" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> Running code</h2>
<p>Try to run the code in the third <strong>cell</strong>. There should be an error as we have not created the objects <code>num_1</code> &amp; <code>num_2</code>. We have only written the code for these objects but not run them.</p>
<p>We can run all the code in a notebook starting from the first <strong>cell</strong> to the last <strong>cell</strong>.</p>
<p>To run all <strong>cells</strong> from the start:</p>
<ul>
<li>Click on the <strong>“Cell”</strong> button.</li>
<li>Click <strong>“Run All”</strong> from the drop-down options.</li>
</ul>
<p>You should then see something like the below in your notebook.</p>
<p><img src="figures/jupyter_notebook_example_4.png" width="100%" style="display: block; margin: auto;" /></p>
<p>There is no output printed for <strong>cell</strong> 2 because we are assigning variables. However, the correct output for Cell 3 is below it. This is because the variables were assigned in <strong>cell</strong> 2 before <strong>cell</strong> 3 was run.</p>
</div>
<div id="saving-the-file" class="section level2" number="5.6">
<h2><span class="header-section-number">5.6</span> Saving the file</h2>
<p><img src="figures/r_save.png" width="10%" style="display: block; margin: auto;" /></p>
<p>As with RStudio and other good coding interfaces we can save our notebook.</p>
<p>First we should rename the file. Rename the notebook to <strong>“jupyter_tut”</strong>:</p>
<ol style="list-style-type: decimal">
<li>Click on the name of the notebook, currently called <strong>“Untitled”</strong>.
<ul>
<li>This is at the very top of the notebook, right of the Jupyter logo.</li>
</ul></li>
<li>A pop-up called <strong>“Rename Notebook”</strong> will appear. Change the Name to <strong>“jupyter_tut”</strong>.</li>
<li>Click <strong>“Rename”</strong>.</li>
</ol>
<p>Now we can save the file. Two methods to save are:</p>
<ul>
<li>Click the floppy disk <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M380.93,57.37A32,32,0,0,0,358.3,48H94.22A46.21,46.21,0,0,0,48,94.22V417.78A46.21,46.21,0,0,0,94.22,464H417.78A46.36,46.36,0,0,0,464,417.78V153.7a32,32,0,0,0-9.37-22.63ZM256,416a64,64,0,1,1,64-64A63.92,63.92,0,0,1,256,416Zm48-224H112a16,16,0,0,1-16-16V112a16,16,0,0,1,16-16H304a16,16,0,0,1,16,16v64A16,16,0,0,1,304,192Z" style="fill:none;stroke:#000;stroke-linecap:round;stroke-linejoin:round;stroke-width:32px"></path></svg> on the toolbar.</li>
<li>Click on the <strong>“File”</strong> button. Click <strong>“Save and Checkpoint”</strong> from the dropdown options.</li>
</ul>
</div>
<div id="title-cells-with-markdown" class="section level2" number="5.7">
<h2><span class="header-section-number">5.7</span> Title cells with markdown</h2>
<p>We will be using multiple notebooks in this workshop. We will also have multiple sections per notebook. It will be useful to create header cells with markdown to create visual separation of the different sections.</p>
<p>To add a header <strong>cell</strong> to the top of our notebook:</p>
<ul>
<li>Create a new <strong>cell</strong> at the top of the notebook.</li>
<li>Click on the <strong>“Code”</strong> drop down and select <strong>“Markdown”</strong>.
<ul>
<li>The <strong>“Heading”</strong> option no longer works.</li>
</ul></li>
</ul>
<p><img src="figures/jupyter_notebook_example_5.png" width="100%" style="display: block; margin: auto;" /></p>
<ul>
<li>Add the following to the <strong>“Markdown”</strong> cell to create a first level header.
<ul>
<li>Ensure you have a space between the <code>#</code> and header text (“Tutorial”).</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tutorial</span></span></code></pre></div>
<p>Great, we can now add nice headers in our notebooks. <strong>Save</strong> the notebook once more before carrying on to the next section.</p>
<div class="webex-solution">
<button>
Markdown
</button>
<p>You won’t need to know more about <code>Markdown</code> but if you are interested please see the <a href="https://www.markdownguide.org/basic-syntax/"><code>Markdown</code> guide</a>.</p>
</div>
</div>
<div id="close-the-notebook" class="section level2" number="5.8">
<h2><span class="header-section-number">5.8</span> Close the notebook</h2>
<p>To close the notebook:</p>
<ul>
<li>Click on <strong>“File”</strong>.</li>
<li>From the dropdown options click <strong>“Close and Halt”</strong>.</li>
</ul>
<p>When you are back in the file explorer page you may not yet set the new file you saved. If so, you will need to refresh the page with the Refresh button <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M320,146s24.36-12-64-12A160,160,0,1,0,416,294" style="fill:none;stroke:#000;stroke-linecap:round;stroke-miterlimit:10;stroke-width:32px"></path> <polyline points="256 58 336 138 256 218" style="fill:none;stroke:#000;stroke-linecap:round;stroke-linejoin:round;stroke-width:32px"></polyline></svg> towards the top right.</p>
<p><img src="figures/jupyter_notebook_refresh.png" width="100%" style="display: block; margin: auto;" /></p>
<p>With that quick tutorial of <code>jupyter-notebook</code> we can start our community analysis in the next chapter.</p>
<p>For more info on <code>jupter-notebook</code> please see the <a href="#jupyter_appendix">appendix</a>.</p>
</div>
<div id="video-tutorial" class="section level2" number="5.9">
<h2><span class="header-section-number">5.9</span> Video tutorial</h2>
<div class="container">
<iframe src="https://www.youtube.com/embed/-c_6HoPMw9g" frameborder="0" allowfullscreen class="video">
</iframe>
</div>

</div>
</div>



<div id="data-prep-intro" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> Data prep intro</h1>
<p><img src="figures/half_tablespoon.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In the next 3 chapters (7-9) we will learn how to:</p>
<ul>
<li><a href="#import_chap">Import our data.</a></li>
<li><a href="#sum_phyloseq_chap">Summarise our <code>phyloseq</code> object.</a></li>
<li><a href="#mindepthchap">Determine the minimum read depth for samples to be used.</a></li>
</ul>
<p>We will also use this as a chance to reinforce how to use <code>jupyter-notebook</code> with clear instructions.
However, from chapters 10 onwards you will make more decisions on how many <strong>“Coding”</strong> and <strong>“Markdown”</strong> cells you want.</p>

</div>
<div id="import_chap" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> Import</h1>
<p><img src="figures/import.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Before carrying out any analysis we first need to import our QIIME2 artifacts into a <code>phyloseq</code> object. Thankfully there is an R package called <code>qiime2R</code>.</p>
<div id="import-notebook" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Import: notebook</h2>
<p><img src="figures/r_script.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Prior to any coding, we will create a new directory, our analysis directory.
afterwards, reate a new notebook called <strong>“01-Import.ipynb”</strong> in it.
We will be creating many notebooks, so we’ll number them so we can easily see the order of scripts.</p>
<p>First create a new directory.</p>
<ul>
<li>In the notebook file explorer, click the <strong>“New”</strong> button.</li>
<li>Select <strong>“Folder”</strong>.</li>
</ul>
<p>You will have an <strong>“Untitled Folder”</strong>. To rename it:</p>
<ul>
<li>Click on the box left of the name.</li>
<li>Press the <strong>“Rename”</strong> button that appeared.</li>
<li>Change the name to <strong>“R_community_workshop”</strong>.</li>
<li>Click <strong>“Rename”</strong>.</li>
</ul>
<p>Click on your <strong>“R_community_workshop”</strong> folder to move into it.</p>
<p>Next step is to create a new R notebook, rename it to <strong>“01-Import”</strong>, and save it.</p>
</div>
<div id="qiime2r" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> qiime2R</h2>
<p><img src="figures/qiime2r.png" width="40%" style="display: block; margin: auto;" /></p>
<p><a href="https://github.com/jbisanz/qiime2R"><code>qiime2R</code></a> is an R package for importing QIIME2 artifacts into a R <code>phyloseq</code> object. The package contains many different commands. Its function <code>read_qza()</code> can read a single artifact at a time.</p>
<p>The best way to import all your QIIME2 artifacts is with the <code>qiime2R::qza_to_phyloseq()</code> function. In your <strong>“01-Import.ipynb”</strong> script, add the following and run the commands.</p>
<p><strong>Tip</strong>: You can tab complete and/or copy and paste file paths within the webVNC.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Cell 1</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Load the package/library</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;qiime2R&quot;</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Import data</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>pseq <span class="ot">&lt;-</span> qiime2R<span class="sc">::</span><span class="fu">qza_to_phyloseq</span>(</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="st">&quot;/pub14/tea/nsc206/NEOF/R_community/data/table-dada2.qza&quot;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">tree =</span> <span class="st">&quot;/pub14/tea/nsc206/NEOF/R_community/data/rooted-tree.qza&quot;</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">taxonomy =</span> <span class="st">&quot;/pub14/tea/nsc206/NEOF/R_community/data/taxonomy.sklearn.qza&quot;</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> <span class="st">&quot;/pub14/tea/nsc206/NEOF/R_community/data/media_metadata.txt&quot;</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>This command creates a phyloseq object named <code>pseq</code>. It contains:</p>
<ul>
<li>The ASV abundance table (<code>features = "table-dada2.qza"</code>).</li>
<li>The rooted phylogenetic tree (<code>tree = "rooted-tree.qza"</code>).</li>
<li>The taxonomic classifications of the ASVs (<code>taxonomy = "taxonomy.sklearn.qza"</code>).</li>
<li>The sample metadata (<code>metadata = "media_metadata.txt"</code>)</li>
</ul>
</div>
<div id="import-summarise-phyloseq" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Import: Summarise phyloseq</h2>
<p><img src="figures/list_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Now that we have imported the data we can extract some summary information from it.</p>
<p>First we will use the <code>microbiome</code> package with its <code>summarize_phyloseq()</code> function.</p>
<p>Create a new <strong>cell</strong> and write + run the below in it.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Cell 2</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Load microbiome library</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Summary of phyloseq object</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq)</span></code></pre></div>
<p>This gives us a plethora of information:</p>
<ul>
<li>The top line tells us if the data is compositional (relative abundance).</li>
<li>We get the following list of values in a paragraph and via a list.
<ul>
<li><strong>Min. number of reads:</strong> Number of reads in the sample with the lowest number of reads.</li>
<li><strong>Max. number of reads:</strong> Number of reads in the sample with the largest number of reads.</li>
<li><strong>Total number of reads:</strong> Sum of all reads across all samples.</li>
<li><strong>Average number of reads:</strong> Sum of all reads / number of samples.</li>
<li><strong>Median number of reads:</strong> Midpoint read abundance across samples.</li>
<li><strong>Sparsity:</strong> See expandable box further down.</li>
<li><strong>Any OTU sum to 1 or less?:</strong> States if there are any ASVs with a summed abundance of 1 or less across all the samples.</li>
<li><strong>Number of singletons:</strong> Number of ASVs with a sum of 1 across all samples.</li>
<li><strong>Percent of OTUs that are singletons:</strong> Percentage of ASVs that only contain one read across all the samples.</li>
<li><strong>Number of sample variables are:</strong> Number of sample variables/groupings in our metadata.</li>
<li>The last line shows the names of the sample variables/groupings in our metadata.</li>
</ul></li>
</ul>
<div class="webex-solution">
<button>
Sparsity
</button>
<p>Sparsity is a measure of the number of 0s in a table. It can be represented by the following equation:</p>
<p><span class="math display">\[
sparsity = Z/C
\]</span></p>
<p>Where:</p>
<ul>
<li>Z = The number of cells that equal zero.</li>
<li>C = The total number of cells.</li>
</ul>
<p>Let’s look at an example of an abundance table with a small amount of ASVs and Samples.</p>
<table>
<thead>
<tr class="header">
<th></th>
<th>Sample1</th>
<th>Sample2</th>
<th>Sample3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ASV1</td>
<td>0</td>
<td>10</td>
<td>24</td>
</tr>
<tr class="even">
<td>ASV2</td>
<td>1</td>
<td>0</td>
<td>37</td>
</tr>
<tr class="odd">
<td>ASV3</td>
<td>6</td>
<td>25</td>
<td>0</td>
</tr>
<tr class="even">
<td>ASV4</td>
<td>51</td>
<td>2</td>
<td>0</td>
</tr>
</tbody>
</table>
<ul>
<li>This abundance table has 12 cells (3 samples * 4 ASVs).</li>
<li>Of these cells, 4 have an abundance of zero.</li>
<li>4/12 = 0.3333, therefore its sparsity is 0.3333.</li>
</ul>
<p>Sparsity can be any value from 0-1. The higher the value the more sparse it is, with a value of 1 meaning all the cells have an abundance of zero. The lower the value the less sparse it is, with a value of 0 meaning all the cells have an abundance of 1 or more.</p>
<p>16S data is known to be sparse so high sparsity is not unexpected. Keep in mind that lower levels of taxa (ASVs, Species, &amp; Genera) will generally have more sparse tables that higher levels of taxa (Kingdom, Phylum, Class).</p>
</div>
<p>If you would like to see how the function calculates its values you can view the <a href="https://rdrr.io/github/microbiome/microbiome/src/R/summarize_phyloseq.R">source code online</a>.</p>
</div>
<div id="save-the-phyloseq-object" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Save the phyloseq object</h2>
<p><img src="figures/r_save.png" width="15%" style="display: block; margin: auto;" /></p>
<p>When using multiple notebooks/scripts for analysis it is useful to save the R objects that will be used in different notebooks/scripts. This can be carried out with the function <code>save()</code>.</p>
<p>Write and run the following code in a third <strong>cell</strong>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Cell 3</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Save phyloseq as file</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(pseq, <span class="at">file =</span> <span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
<p>This saves our object <code>pseq</code> into the file <code>phyloseq.RData</code>. The suffix <code>.RData</code> is the normal convention.</p>
<p>We have saved our final object of the notebook. Close and halt it.</p>
</div>
<div id="import-recap" class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> Import: recap</h2>
<p><img src="figures/recap.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have imported our QIIME2 artifacts into one <code>phyloseq</code> object so we can analyse the data in R. This object has been saved into a “.RData” file which we will load in the next chapter.</p>

</div>
</div>
<div id="sum_phyloseq_chap" class="section level1" number="8">
<h1><span class="header-section-number">Chapter 8</span> Summarise phyloseq</h1>
<p><img src="figures/transform.png" width="20%" style="display: block; margin: auto;" /></p>
<p>For the next 2 chapters we will use one notebook. We will setup this notebook with packages and the <code>phyloseq</code> object we created in the last chapter.</p>
<p>After this, we will summarise the <code>phyloseq</code> object. We will investigate the read depth of samples and the number of ASVs in our dataset.</p>
<div id="summarise-setup" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Summarise: setup</h2>
<p><img src="figures/setup_2.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Before starting analysis create and save a new <code>R</code> notebook called <strong>“02-Preprocess.ipynb”</strong> in the analysis directory (when renaming you don’t need to include the suffix).</p>
<p>It is useful to add a title to the top of the notebook.
Create a <strong>“Markdown”</strong> <strong>cell</strong> and add the following first level header:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Preprocessing data notebook</span></span></code></pre></div>
<p>The first section of our notebook will be used for setup.
This will involve loading packages and data we need.</p>
<p>For good documentation add a second level heading to the first <strong>“Markdown”</strong> cell.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Setup</span></span></code></pre></div>
<p>To decrease the level of a heading add another <code>#</code>.</p>
<ul>
<li><code>#</code>: 1st level header.</li>
<li><code>##</code>: 2nd level header.</li>
<li><code>###</code>: 3rd level header etc…</li>
</ul>
<p>I like to load all the packages to be used in the notebook in this section. We will explain their uses later in this chapter. Add the below to a new <strong>“Code”</strong> cell in your notebook:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Packages</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;vegan&quot;</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span></code></pre></div>
<p>Our last bit of set-up is to load in our abundance <code>phyloseq</code> object we created in the previous chapter.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Load the phyloseq object</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
<p>Ensure you have run the code in this cell.</p>
<div id="summarise-header" class="section level3" number="8.1.1">
<h3><span class="header-section-number">8.1.1</span> Summarise header</h3>
<p><img src="figures/header.png" width="15%" style="display: block; margin: auto;" /></p>
<p>When you load in a dataset it is always useful to check it.
We will therefore use a new section to inspect the data.</p>
<p>In a new <strong>“Markdown”</strong> cell add the following 2nd level header.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Summarise the phyloseq object</span></span></code></pre></div>
</div>
<div id="summarise-phyloseq" class="section level3" number="8.1.2">
<h3><span class="header-section-number">8.1.2</span> Summarise phyloseq</h3>
<p><img src="figures/list_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Create a new code <strong>cell</strong> and add the following annotation and code. Then run the code.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summary of phyloseq object</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq)</span></code></pre></div>
<p>We ran this code in the <strong>“01-Import.ipynb”</strong>. You can therefore check if our new output matched the output from that notebook. This should be the case since they are the same data.</p>
<p>Due to the relative large amount of output to screen we’ll put the next part into a new <strong>cell</strong>.</p>
</div>
<div id="reads-per-sample" class="section level3" number="8.1.3">
<h3><span class="header-section-number">8.1.3</span> Reads per sample</h3>
<p><img src="figures/histogram.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In our new <strong>cell</strong> we will use the command <code>microbiome::readcount()</code> to store and display the number of reads in each sample within our <code>phyloseq</code> object (<code>pseq</code>).</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of reads per sample</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>sample_depths <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>sample_depths</span></code></pre></div>
<p>With this info we can make a histogram of read numbers per sample with the base <code>R</code> function <code>hist()</code>.
We will save this as a <code>.png</code> file with the functions <code>png()</code> and <code>dev.off()</code>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Histogram</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Save as png</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">filename =</span> <span class="st">&quot;./sample_depth_histogram.png&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">100</span>, <span class="at">width =</span> <span class="dv">200</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(sample_depths, <span class="at">main =</span> <span class="st">&quot;Histogram of read depths&quot;</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div id="display-png" class="section level3" number="8.1.4">
<h3><span class="header-section-number">8.1.4</span> Display png</h3>
<p><img src="figures/display_png.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We can use the R package <a href="https://github.com/IRkernel/IRdisplay"><code>IRdisplay</code></a> to display plots from files in <code>jupyter-notebook</code>.</p>
<p>Use the function <code>display_png()</code> to display our histogram.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./sample_depth_histogram.png&quot;</span>)</span></code></pre></div>
<p>This information is very useful. We will use it in the next chapter to determine what our minimum read depth should be.</p>
</div>
<div id="asvs-per-sample" class="section level3" number="8.1.5">
<h3><span class="header-section-number">8.1.5</span> ASVs per sample</h3>
<p><img src="figures/nice_table.png" width="15%" style="display: block; margin: auto;" /></p>
<p>The last information we will look at before some preprocessing are the ASVs (Amplicon Sequnces Variants).</p>
<p>A useful <code>phyloseq</code> command is <code>otu_table()</code>. This allows us to extract the ASV/OTU/feature table. With this we can then carry out some other commands.</p>
<p>For demonstrative purposes write and run the following code in its own <strong>cell</strong> to display the ASV table.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Can extract ASV table (known as otu table in phyloseq)</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq)</span></code></pre></div>
<p>With this function we will extract the number of ASVs in the original abundance table.
We will then save this in a vector.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Each row is an ASV and each column is a sample</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Therefore we can get the number of ASVs in the data by counting the number of rows</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Let&#39;s make a new vector with this info so we can easily keep track</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>num_asvs_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq)))</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Give the 1st element a relevant name</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(num_asvs_vec)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">&quot;abundance&quot;</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co">#View current vector</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>num_asvs_vec</span></code></pre></div>
<p>Save this vector as a <code>.Rdata</code> object in a new <strong>cell</strong>.
This will allow us to load it in future notebooks.
We can remove the object so it doesn’t use RAM.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Save object as file</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(num_asvs_vec, <span class="at">file=</span> <span class="st">&quot;num_asvs_vec.RData&quot;</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove object from environment</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(num_asvs_vec)</span></code></pre></div>
</div>
</div>
<div id="summarise-recap" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Summarise: recap</h2>
<p><img src="figures/recap.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We now have an idea of some of the attributes of our dataset. This includes the number of reads and ASVs. We can use this knowledge to carry out some preprocessing.</p>

</div>
</div>
<div id="mindepthchap" class="section level1" number="9">
<h1><span class="header-section-number">Chapter 9</span> Minimum read depth</h1>
<p><img src="figures/depth_diver.png" width="20%" style="display: block; margin: auto;" /></p>
<p>It is good to remove samples with a very low read depth (number of sequencing reads).
But determining an appropriate minimum read depth is not trivial.
This value will vary from study to study.
Normally, for 16S data, a depth of at least 20K per sample is suggested.
This is the general consensus for <strong>human microbiome</strong> data.</p>
<p>In this chapter we will:</p>
<ul>
<li>Cover a brief intro to considerations of what is an acceptable minimum depth for your dataset.</li>
<li>Reinvestigate our sample depths with a previously created histogram.</li>
<li>View the depth ranges of different sample groups (site &amp; media) with box plots.</li>
<li>Create a rarefaction curve to assess if the depth of our samples have captured a good amount of biodiversity.</li>
<li>Demonstrate how to filter samples by depth.</li>
</ul>
<div id="considerations" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Considerations</h2>
<p><img src="figures/consider.png" width="15%" style="display: block; margin: auto;" /></p>
<p>There are 3 main considerations to take into account when choosing an appropriate depth for your dataset.</p>
<ul>
<li><strong>The biodiversity of your samples.</strong>
<ul>
<li>If your sample is very biodiverse, such as the human gut microbiome, you will need a good depth (&gt;20K per sample).</li>
<li>If your sample is less biodiverse, such as many geological environments or skin, then you will not need as much read depth.</li>
<li>Rarefaction curves are a good method to determine if your samples have enough depth. We will look at this in this chapter.</li>
</ul></li>
<li><strong>The biomass of your samples.</strong>
<ul>
<li>Some environments are hard to extract DNA from.</li>
<li>If this is the case for you, then people will hopefully accept that this is an unfortunate reality of life and you will use what you can.</li>
<li>Be careful of your conclusions. If you think your dataset doesn’t have as much data as it could, do not make very definitive detailed claims.</li>
</ul></li>
<li><strong>Read depth of sample groups.</strong>
<ul>
<li>It may be possible that a few samples have a much lower depth than the rest. It may be a good idea to remove them.</li>
<li>However, these may all come from the same sample group and so you have lost all information of one group.</li>
<li>For instance, you may be comparing different geological surfaces and your rock samples have much lower read depths than the various soil samples.</li>
<li>For comparisons including the lower depth sample group (e.g. rock samples) you will need to retain the lower depth samples.</li>
</ul></li>
</ul>
<p>That is a brief overview of that topic. If you are interested in more I suggest you look at papers where they have studied an environment similar to yours.</p>
</div>
<div id="minimum-read-depth-section-title" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Minimum read depth section title</h2>
<p><img src="figures/header_l2.png" width="10%" style="display: block; margin: auto;" /></p>
<p>In the next section of our <code>jupyter-notebook</code> we will investigate what the minimum read depth should be and remove samples with a lower depth.</p>
<p>Create a new <strong>“Markdown”</strong> <strong>cell</strong> and add the following 2nd level header.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Minimum read depth</span></span></code></pre></div>
</div>
<div id="read-depth-vector-and-histogram" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> Read depth vector and histogram</h2>
<p><img src="figures/histogram.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have already created a vector and histogram of the read depths across our samples. Scroll up your notebook to view these and answer the following MCQ:</p>
<ul>
<li>What is the approximate read depth range of our dataset?
<div id="radio_PHRPAMRZFI" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_PHRPAMRZFI" value=""></input> <span><strong>1-8</strong></span></label><label><input type="radio" autocomplete="off" name="radio_PHRPAMRZFI" value="answer"></input> <span><strong>10,000-18,000</strong></span></label><label><input type="radio" autocomplete="off" name="radio_PHRPAMRZFI" value=""></input> <span><strong>20,000+</strong></span></label>
</div></li>
</ul>
<p>This is a good first step but what if we want to know how the read depths vary between sample groups?</p>
</div>
<div id="sample-depth-boxplot" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> Sample depth boxplot</h2>
<p><img src="figures/boxplot.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We are going to use <code>ggplot2</code> to create a couple boxplots to show the sequencing depth ranges of the different sample groups (Site &amp; Media).</p>
<div id="creating-a-data-frame-for-boxplots" class="section level3" number="9.4.1">
<h3><span class="header-section-number">9.4.1</span> Creating a data frame for boxplots</h3>
<p><img src="figures/metadata.png" width="10%" style="display: block; margin: auto;" /></p>
<p>First, we need to create an object containing our sample names, site &amp; media information, and the depth.
We will use this object to produce our boxplots.
Thankfully the metadata in our <code>phyloseq</code> object contains all this information except the depth.</p>
<p>In a new <strong>cell</strong> write and run the below code. This will extract the sample data (metadata) to a new object and display the top 6 rows fo this new object.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract sample data as a separate R object</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>abundance_metadf <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View top 6 rows of metadata data frame</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundance_metadf)</span></code></pre></div>
<p>The 2 new functions above are:</p>
<ul>
<li><code>phyloseq::sample_data()</code>: Extracts the sample data (metadata) data frame from a <code>phyloseq</code> object.</li>
<li><code>head()</code>: Returns the first 6 parts of an R object by default.
<ul>
<li>It can be used for a vector, matrix, table, data frame, or function.</li>
<li>In the case of a data frame it returns the first 6 rows.</li>
</ul></li>
</ul>
<p>We need to add the depth information to our new data frame.
We have extracted this previously into an object called <code>sample_depths</code>.
However, before adding it we want to check if it has the same order of samples as the rows in <code>abundance_metadf</code>.</p>
<p>Write and run the below script in the same <strong>cell</strong>. The code uses <code>head()</code> to view the first 6 elements of <code>sample_depths</code> and the row names of <code>abundance_metadf</code>. Then the function <code>identical()</code> is used to see if they are identical (<code>TRUE</code>) or not (<code>FALSE</code>).</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Check if our vector of sample_depths has the same order as our metadata rows</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">names</span>(sample_depths))</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">row.names</span>(abundance_metadf))</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(<span class="fu">names</span>(sample_depths),<span class="fu">row.names</span>(abundance_metadf))</span></code></pre></div>
<p>The order of samples is identical so we can add the depth information to <code>abundance_metadf</code>. Carry this out in the same cell with the code below.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Add sample depths to metadata data frame</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>abundance_metadf[,<span class="st">&quot;depth&quot;</span>] <span class="ot">&lt;-</span> sample_depths</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View top 6 rows of edited metadata dataframe</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundance_metadf)</span></code></pre></div>
<p>Great! We will use this data frame to create 2 boxplots.</p>
</div>
<div id="ggplot2-boxplot" class="section level3" number="9.4.2">
<h3><span class="header-section-number">9.4.2</span> ggplot2 boxplot</h3>
<p><img src="figures/boxplots_groups.png" width="10%" style="display: block; margin: auto;" /></p>
<p>We are going to create 2 boxplots with <code>ggplot2</code>.
We won’t go into too much detail on how the code works here, instead learning more later in this book.</p>
<p>The code below creates a <code>ggplot2</code> boxplot. We carry this out with 2 functions:</p>
<ul>
<li><code>ggplot()</code>: This creates a <code>ggplot2</code> object, storing the information and aesthetics.
<ul>
<li>The first option is the data we want to use for plotting (<code>abundance_metadf</code>).</li>
<li>The second option is the aesthetics (<code>aes()</code>) to plot. In this case we want the depth column to be plotted on the y-axis (<code>y=depth</code>) and the site column to be plotted on the x-axis (<code>x=site</code>).</li>
</ul></li>
<li><code>+</code>: We need to have a <code>+</code> at the end of the <code>ggplot()</code> function to add the next component of the plot.</li>
<li><code>geom_boxplot()</code>: This adds a <strong>layer</strong> to our <code>ggplot2</code> object. In this case it converts the <code>ggplot</code> object, which is just information, into a boxplot.</li>
</ul>
<p>Write and run the following code in a new cell:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create ggplot2 boxplot of depth by size</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>boxplot <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(abundance_metadf, <span class="fu">aes</span>(<span class="at">y=</span>depth, <span class="at">x=</span>site)) <span class="sc">+</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p>We have a saved <code>ggplot2</code> object but now wee need to save and display it.</p>
</div>
<div id="save-ggplot-as-png" class="section level3" number="9.4.3">
<h3><span class="header-section-number">9.4.3</span> Save ggplot as png</h3>
<p><img src="figures/png_file.png" width="10%" style="display: block; margin: auto;" /></p>
<p>You can save a <code>ggplot</code> object as an image file with <code>ggsave()</code>.
This allows you to select the size and resolution (<code>dpi</code>) of the image.
This can then be displayed with <code>IRdisplay::display_png()</code>.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./boxplot_x_site_y_depth.png&quot;</span>, <span class="at">plot =</span> boxplot,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">100</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./boxplot_x_site_y_depth.png&quot;</span>)</span></code></pre></div>
<ul>
<li>Which site has the highest median depth (middle line of boxplot)?
<div id="radio_QIJHPPDNIQ" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_QIJHPPDNIQ" value="answer"></input> <span><strong>LD (Lower Durance)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_QIJHPPDNIQ" value=""></input> <span><strong>MD (Middle Durance)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_QIJHPPDNIQ" value=""></input> <span><strong>UD (Upper Durance</strong></span></label>
</div></li>
</ul>
</div>
<div id="media-by-depth-boxplot" class="section level3" number="9.4.4">
<h3><span class="header-section-number">9.4.4</span> Media by depth boxplot</h3>
<p><img src="figures/petri.png" width="10%" style="display: block; margin: auto;" /></p>
<p>We will use the same code to plot the depth by media.
You can copy and paste the code changing all hte instances of <strong>site</strong> to <strong>media</strong>.
Carry this out in the same cell and run the code.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create ggplot2 boxplot of depth by size</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(abundance_metadf, <span class="fu">aes</span>(<span class="at">y=</span>depth, <span class="at">x=</span>media)) <span class="sc">+</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./boxplot_x_media_y_depth.png&quot;</span>, <span class="at">plot =</span> boxplot,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">100</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./boxplot_x_media_y_depth.png&quot;</span>)</span></code></pre></div>
<ul>
<li>Which media has the lowest median depth (middle line of boxplot)?
<div id="radio_RFOFXXDOEH" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_RFOFXXDOEH" value="answer"></input> <span><strong>CVP</strong></span></label><label><input type="radio" autocomplete="off" name="radio_RFOFXXDOEH" value=""></input> <span><strong>ENV</strong></span></label><label><input type="radio" autocomplete="off" name="radio_RFOFXXDOEH" value=""></input> <span><strong>TSA</strong></span></label>
</div></li>
</ul>
<p>From the boxplots we can see there is no drastic difference between the depths of the different media groups. We will therefore continue and make some rarefaction curves to further assess the depth of our samples.</p>
<p>For more resources on <code>ggplot2</code> please see the <a href="#ggplot2_appendix">appendix</a> of this book.</p>
</div>
</div>
<div id="chap9rarefaction" class="section level2" number="9.5">
<h2><span class="header-section-number">9.5</span> Rarefaction curve</h2>
<p><img src="figures/rarefaction.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Our read depths appear a bit low, each sample has &lt;20K reads.
However, this might be fine for our dataset since we are using surface water samples rather than human gut microbiome samples.
Let’s see how our samples look with a rarefaction curve.</p>
<p><strong>Note</strong>: This is a quick example and we will go into more detail in the <a href="#rarefaction_chap">rarefaction chapter</a>.</p>
<p>Surprisingly, there is not a good method to produce a rarefaction curve with the <code>phyloseq</code> or <code>microbiome</code> packages.
We will therefore use the <a href="https://vegandevs.github.io/vegan/"><code>vegan</code></a> package.
<code>vegan</code> is an R package for community ecologists.
It has a variety of functions but it uses normal R data frames rather than <code>phyloseq</code> objects.
We will therefore only use it for rarefaction purposes.</p>
<div id="asv-abundance-data-frame" class="section level3" number="9.5.1">
<h3><span class="header-section-number">9.5.1</span> ASV abundance data frame</h3>
<p><img src="figures/nice_table.png" width="10%" style="display: block; margin: auto;" /></p>
<p>Before creating our rarefaction curve we will extract the ASV abundance table from the <code>phyloseq</code> object with <code>phyloseq</code>’s function <code>otu_table()</code>.
We need to transpose (<code>t()</code>) the table so it is in the correct orientation for the rarefaction function.
Additionally, we will ensure it is a data frame with the function <code>as.data.frame()</code>.</p>
<p>Carry this out in a new <strong>cell</strong>.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Rarefaction curve</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract ASV table as data frame</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>asv_abund_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq)))</span></code></pre></div>
</div>
<div id="vegans-rarecurve" class="section level3" number="9.5.2">
<h3><span class="header-section-number">9.5.2</span> vegan’s rarecurve</h3>
<p><img src="figures/vegan.png" width="10%" style="display: block; margin: auto;" /></p>
<p>With this data frame we can create a rarefaction curve with <code>vegan</code>’s <code>rarecurve()</code> function.</p>
<p>Add the following code to the same cell and run it.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Rarefaction curve</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Save as png</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">filename =</span> <span class="st">&quot;rarefaction.png&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">200</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">rarecurve</span>(</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> asv_abund_df, <span class="at">step =</span> <span class="dv">50</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Read depth&quot;</span>,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;ASVs&quot;</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./rarefaction.png&quot;</span>)</span></code></pre></div>
<p><strong>Note</strong>: You will get a warning saying <code>"most observed count data have counts 1, but smallest count is 2"</code>. This can be ignored in this case.</p>
<p>In essence, we are hoping that the majority of samples have plateau’d.
If the curves have flattened in relation to the y axis this indicates that most of the ASVs present in the sample have been captured.</p>
<p>Our samples have plateau’d or have gentle slopes towards the end.
With this we can be happy to continue and not remove samples by a minimum read depth.</p>
<p>If we saw some samples with steep curves and low depths we could carry out some more analysis with rarefaction.
However, more rarefaction functions, analysis, and theories will be covered in the <a href="#rarefaction_chap">rarefaction chapter</a>.
You could use some of these at this point to help you determine your minimum read depth for your own datasets.</p>
</div>
</div>
<div id="filtering-by-minimum-read-depth" class="section level2" number="9.6">
<h2><span class="header-section-number">9.6</span> Filtering by minimum read depth</h2>
<p><img src="figures/water_filter.png" width="10%" style="display: block; margin: auto;" /></p>
<p>What if you want to filter samples by a minimum read depth?</p>
<p>You can use the <code>subset_samples()</code> function from <code>phyloseq()</code>.
We will use our previously created vector containing read depths (<code>sample_depths</code>) to remove sample with less than 11k reads.
We have chosen this depth as an example to remove some samples.</p>
<p>Write and run the below code in a new cell.
It will create a new subsetted <code>phyloseq</code> object.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Subset and keep samples with more than 11k reads</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>pseq_min11K <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq, sample_depths <span class="sc">&gt;</span> <span class="dv">11000</span>)</span></code></pre></div>
<p>After removing samples it is also useful to remove ASVs with no abundance values.
This can occur when ASVs are only present in the samples which have been removed.</p>
<p>To remove these ASVs we can use two <code>phyloseq</code> functions:</p>
<ul>
<li><strong><code>taxa_sums()</code></strong>: Returns a vector showing sum of all taxa in the abundance table.
<ul>
<li>In our <code>phyloseq</code> object the ASVs are the taxa.</li>
<li>ASVs have long human unfriendly names that are unique to every single ASV possible.</li>
</ul></li>
<li><strong><code>prune_taxa()</code></strong>: This retains taxa/ASVs based on a provided vector.
<ul>
<li>We are creating a logical vector (<code>TRUE</code>/<code>FALSE</code>) where ASVs with 0 abundance are <code>FALSE</code> and ASVs with abundance &gt; 0 are <code>TRUE</code>.</li>
</ul></li>
</ul>
<p>We’ll first write and run some commands with <code>taxa_sums()</code> to get some practice with it. Carry this out in the same <strong>cell</strong> as the <code>subset_samples()</code> command.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Abundance sums of the 1st six ASVs</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(phyloseq<span class="sc">::</span><span class="fu">taxa_sums</span>(pseq_min11K))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View number of ASVs in our data</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(phyloseq<span class="sc">::</span><span class="fu">taxa_sums</span>(pseq_min11K))</span></code></pre></div>
<p>In the same cell add the following.
This will filter out ASVs with no abundance.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove ASVs with no abundance</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>pseq_min11k <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">prune_taxa</span>(</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  phyloseq<span class="sc">::</span><span class="fu">taxa_sums</span>(pseq_min11k) <span class="sc">&gt;</span> <span class="dv">0</span>, pseq_min11k</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Finally, summarise the contents of the <code>phyloseq</code> object.
Add the following in the same cell and then run the code in the cell.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise subsetted phyloseq</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq_min11K)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_min11K)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>pseq_min11k</span></code></pre></div>
<p>We can see that the data lost 26 ASVs (2551 - 2525).</p>
<p>Try to write your own R code in a new cell to answer the following questions:</p>
<ul>
<li>What is the difference of the minimum number of reads between <code>pseq_min11k</code> and <code>pseq</code>?
<div id="radio_WDTSOQJULZ" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_WDTSOQJULZ" value=""></input> <span><strong>4</strong></span></label><label><input type="radio" autocomplete="off" name="radio_WDTSOQJULZ" value="answer"></input> <span><strong>613</strong></span></label><label><input type="radio" autocomplete="off" name="radio_WDTSOQJULZ" value=""></input> <span><strong>42507</strong></span></label>
</div></li>
<li>How many samples were removed due to the read depth filtering?
<div id="radio_WUNCRTWIXQ" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_WUNCRTWIXQ" value="answer"></input> <span><strong>4</strong></span></label><label><input type="radio" autocomplete="off" name="radio_WUNCRTWIXQ" value=""></input> <span><strong>613</strong></span></label><label><input type="radio" autocomplete="off" name="radio_WUNCRTWIXQ" value=""></input> <span><strong>42507</strong></span></label>
</div></li>
<li>What is the difference of the minimum number of reads between <code>pseq_min11k</code> and <code>pseq</code>?
<div id="radio_UWEZFURBXL" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_UWEZFURBXL" value=""></input> <span><strong>4</strong></span></label><label><input type="radio" autocomplete="off" name="radio_UWEZFURBXL" value=""></input> <span><strong>613</strong></span></label><label><input type="radio" autocomplete="off" name="radio_UWEZFURBXL" value="answer"></input> <span><strong>42507</strong></span></label>
</div></li>
</ul>
<div class="webex-solution">
<button>
Tip
</button>
<p>Combining <code>microbiome::readcount()</code> with <code>min()</code>, <code>length()</code>, and <code>sum()</code> might help.</p>
</div>
<div class="webex-solution">
<button>
Code solutions + bonus
</button>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Difference of minimum read numbers</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_min11k)) <span class="sc">-</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq))</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of samples lost</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq)) <span class="sc">-</span> <span class="fu">length</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_min11k))</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of reads removed</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq)) <span class="sc">-</span> <span class="fu">sum</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_min11k))</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Bonus</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="co">#List the removed samples</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="fu">setdiff</span>(phyloseq<span class="sc">::</span><span class="fu">sample_names</span>(pseq), phyloseq<span class="sc">::</span><span class="fu">sample_names</span>(pseq_min11k))</span></code></pre></div>
</div>
<p>We won’t actually use this subsetted file as we want to keep all the samples in this case.
Since we not removing any samples we don’t need to remove any ASVs as they should all have a total abundance &gt; 0.
You can therfore remove the new <code>phyloseq</code> object in a new <strong>cell</strong>.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove subsetted phyloseq</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(pseq_min11K)</span></code></pre></div>
<p>Once you are finished with this notebook you can save it then close and halt it.</p>
</div>
<div id="minmum-read-depth-summary" class="section level2" number="9.7">
<h2><span class="header-section-number">9.7</span> Minmum read depth: Summary</h2>
<p><img src="figures/sum_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have assessed the read depth in this chapter and decided to not remove any samples.
This assessment included:</p>
<ul>
<li>Viewing a histogram of sample read depths.</li>
<li>Creating boxplots to compare the sample read depths across sample groups (siet and media).</li>
<li>Producing a rarefaction depth to determine if any samples did not represent a good amount of the ASVs present in the environment.</li>
</ul>
<p>Additionally, we created a new <code>phyloseq</code> object where the samples were filtered by depth.
Ultimately we did not keep the filtered <code>phyloseq</code> object but it is good to know how to do this for your future research.</p>

</div>
</div>



<div id="taxa_relabund_chap" class="section level1" number="10">
<h1><span class="header-section-number">Chapter 10</span> Taxa relative abundance</h1>
<p><img src="figures/taxa.png" width="20%" style="display: block; margin: auto;" /></p>
<p>In this section we are going to create taxonomy bar charts.
For this we are going to use relative abundance tables of different taxa levels; Genus, Family, &amp; Phyla.</p>
<p>In this chapter we are going to create a <code>phyloseq</code> object with relative abundance values of phyla.
We would use read depth filtered data if we wanted to use it for other downstream analyses.
But, as stated in the last chapter all our samples had sufficient depth, so we will use our original abundance <code>phyloseq</code> object.</p>
<div id="taxa-relative-abundance-setup" class="section level2" number="10.1">
<h2><span class="header-section-number">10.1</span> Taxa relative abundance: setup</h2>
<p><img src="figures/packages.png" width="15%" style="display: block; margin: auto;" /></p>
<p>The first steps before analysis are:</p>
<ul>
<li>Create a new notebook called <strong>“3-taxonomy_barcharts”</strong>.
<ul>
<li>We will use this to create our phyloseq objects and taxonomy bar charts.</li>
</ul></li>
<li>Add a markdown cell with the first level header: <code># Taxonomy bar charts</code>.</li>
<li>Add the below to a code cell to load in the <code>phyloseq</code> object and libraries.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
<p>From now on you will get less instructions on your notebook structure.
Please create your own coding and markdown cells where you think appropriate.</p>
</div>
<div id="relative-abundance-transformation" class="section level2" number="10.2">
<h2><span class="header-section-number">10.2</span> Relative abundance transformation</h2>
<p><img src="figures/count_to_percent.png" width="40%" style="display: block; margin: auto;" /></p>
<p>Now that we have the data loaded we can create a new <code>phyloseq</code> object by transforming the abundance values to relative abundances.</p>
<p>This is carried out with the <code>microbiome</code> function <code>transform()</code>. With it we transform the ASV abundance table within to a “compositional” table (relative abundance).</p>
<p>Run the below command in an appropriate place in your notebook:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Transform abundance table to a relative abundance (compositional) table</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>pseq_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq, <span class="st">&quot;compositional&quot;</span>)</span></code></pre></div>
<p>As always it is good to check the contents of the new ASV table.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise and check sample counts which should each amount to 1</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq_relabund)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_relabund)</span></code></pre></div>
<p>You will notice the read count for each sample is <strong>1</strong>. This abundance table contains fractional relative abundances for each ASV. This fraction is relative to the total abundance within each sample. Therefore, all the fractional relative abundance values of ASVs in one sample total 1.</p>
<p>We will use this to produce all our different taxa tables.</p>
</div>
<div id="aggregate-taxa" class="section level2" number="10.3">
<h2><span class="header-section-number">10.3</span> Aggregate taxa</h2>
<p><img src="figures/aggregate_taxa.png" width="15%" style="display: block; margin: auto;" /></p>
<p>To produce a taxa table we can use the <code>microbiome</code> function <code>aggregate_taxa()</code>.</p>
<p>Run the below command to produce a phylum based <code>phyloseq</code> object.</p>
<p><strong>Reminder</strong>: You may want a markdown cell to create a 2nd level header for this phylum section.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Check head of tax_table</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="co">#This will tell us the taxa level names on the column names</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(phyloseq<span class="sc">::</span><span class="fu">tax_table</span>(pseq_relabund))</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Phylum phyloseq</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>phylum_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_taxa</span>(pseq_relabund, <span class="st">&quot;Phylum&quot;</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s check our phyla <code>phyloseq</code>.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Head of phylum relative abundance table</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(phylum_pseq))</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of phyla</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Number of phyla: &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(phylum_pseq)))</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(phylum_pseq)</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(phylum_pseq)</span></code></pre></div>
</div>
<div id="taxa-relative-abundance-summary" class="section level2" number="10.4">
<h2><span class="header-section-number">10.4</span> Taxa Relative abundance: summary</h2>
<p><img src="figures/sum_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have produced a <code>phyloseq</code> object containing a Phylum relative abundance table. Next, we will create a taxonomy bar chart with this.</p>

</div>
</div>
<div id="chaptaxaplots" class="section level1" number="11">
<h1><span class="header-section-number">Chapter 11</span> Taxa plots</h1>
<p><img src="figures/owl_eyes.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Now that we have our Genus relative abundance <code>phyloseq</code> we can create a bar chart.</p>
<div id="simple-bar-chart" class="section level2" number="11.1">
<h2><span class="header-section-number">11.1</span> Simple bar chart</h2>
<p><img src="figures/bar_chart.png" width="15%" style="display: block; margin: auto;" /></p>
<p>To create, save, and view a simple genus bar chart we can use the following code.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Simple bar chart</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>phylum_bar <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(phylum_pseq)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./phylum_relabund_simple.png&quot;</span>, <span class="at">plot =</span> phylum_bar,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./phylum_relabund_simple.png&quot;</span>)</span></code></pre></div>
<p>This is a good start but we can improve it.</p>
<div id="editing-the-plot" class="section level3" number="11.1.1">
<h3><span class="header-section-number">11.1.1</span> Editing the plot</h3>
<p><img src="figures/plus_sign.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Components can be added to <code>ggplot2</code> objects to edit the plot.
They are added with the <code>+</code> symbol.</p>
<p>Go to the cell where you created your plot and add components so it looks like the below:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Simple bar chart</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>phylum_bar <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(phylum_pseq) <span class="sc">+</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Change/add the x and y labels</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Add a title to the plot</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Phylum relative abundance bar chart&quot;</span>)</span></code></pre></div>
<p>Now save and display the plot.</p>
</div>
</div>
<div id="taxa-heatmap" class="section level2" number="11.2">
<h2><span class="header-section-number">11.2</span> Taxa heatmap</h2>
<p><img src="figures/heatmap.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Our bar chart looks quite nice but for displaying all taxa present it may be better to use a heatmap.
Use the below code to produce and visualise a heatmap.</p>
<p><strong>Tip</strong>: I encourage you to copy and paste your previous bar chart code to edit.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Produce heatmap ggplot</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>phylum_heatmap <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(phylum_pseq, <span class="at">plot.type =</span> <span class="st">&quot;heatmap&quot;</span>) <span class="sc">+</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Phylum&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Phylum relative abundance heatmap&quot;</span>)</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./phylum_relabund_heatmap.png&quot;</span>, <span class="at">plot =</span> phylum_bar,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./phylum_relabund_heatmap.png&quot;</span>)</span></code></pre></div>
<p>With the heatmap it is much easier to tell what the relative abundances of different phyla are.</p>
<ul>
<li>What is the most abundant phyla across the samples?
<div id="radio_KMLBSHQIGF" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_KMLBSHQIGF" value=""></input> <span><strong>Chloroflexi</strong></span></label><label><input type="radio" autocomplete="off" name="radio_KMLBSHQIGF" value=""></input> <span><strong>Firmicutes</strong></span></label><label><input type="radio" autocomplete="off" name="radio_KMLBSHQIGF" value="answer"></input> <span><strong>Proteobacteria</strong></span></label>
</div></li>
</ul>
</div>
<div id="aggregateraretaxa" class="section level2" number="11.3">
<h2><span class="header-section-number">11.3</span> Aggregate rare taxa</h2>
<p><img src="figures/origami_rhino.png" width="15%" style="display: block; margin: auto;" /></p>
<p>From our bar chart and our heatmap we can see that there are a few high abundance phyla (Actinobacteria, Bacteroidetes, Firmicutes, &amp; Proteobacteria) and many low abundance phyla.</p>
<p>To make a better visualisation we will aggregate the rare phyla.
This will give us less phyla to plot so we don’t need to use so many colours.</p>
<p>To carry this out we will use the <code>microbiome</code> function <code>aggregate_rare()</code>.
The function has three main options:</p>
<ul>
<li><strong><code>level =</code></strong>: Taxa level to aggregate table.</li>
<li><strong><code>detection =</code></strong>: Detection threshold (see below).</li>
<li><strong><code>prevalence =</code></strong>: Prevalence threshold (see below).</li>
</ul>
<p>The function will aggregate taxa to our specified level.
Whilst doing this it will aggregate the rare taxa of that level to one group called “Other”.
Rare taxa are specified via the <strong>detection</strong> and <strong>prevalence</strong> thresholds.
Any taxa with an abundance &gt;= to the <strong>detection</strong> threshold in a number of samples &gt;= to the <strong>prevalence</strong> threshold will be kept.
The taxa which don’t reach these thresholds will be classified as rare and aggregated into “Other”.</p>
<p>In the below example we are aggregating the rare taxa with our ASV relative abundance <code>phyloseq</code> object.
Any phyla with a relative abundance &gt;= to 0.01 (<strong>detection</strong>) in at least 5% (5/100) of the samples will not be classified as rare.</p>
<p>More info on <a href="https://microbiome.github.io/OMA/quality-control.html#prevalence">Prevalence &amp; detection</a>.</p>
<p><strong>Note</strong>: The <strong>detection</strong> and <strong>prevalence</strong> thresholds can be given as numbers (1, 2, 1000 etc.), or percentage values (5/100, 50/100, etc.).</p>
<p>Aggregate the relative abundance table to phyla with the rare thresholds:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Aggregate rare phyla</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>phylum_rareaggregate_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Phylum&quot;</span>,</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> <span class="fl">0.01</span>, <span class="at">prevalence =</span> <span class="dv">5</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>We will plot this as a new taxonomy bar chart with some additions:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Add otu.sort to sort phyla by abundance (highest to lowest)</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>phylum_bar <span class="ot">&lt;-</span> </span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(phylum_rareaggregate_pseq, </span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">otu.sort =</span><span class="st">&quot;abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Phylum relative abundance bar chart&quot;</span>) <span class="sc">+</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Change colours of Phylum to that of the &quot;Paired&quot; palette from colour brewer</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="st">&quot;Phylum&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Paired&quot;</span>)</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png</span></span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./phylum_relabund_rareaggregate.png&quot;</span>, <span class="at">plot =</span> phylum_bar,</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./phylum_relabund_rareaggregate.png&quot;</span>)</span></code></pre></div>
<div class="webex-solution">
<button>
Colour brewer palettes
</button>
<p><img src="figures/colour_brewer_palettes.png" width="80%" style="display: block; margin: auto;" />
<a href="https://r-graph-gallery.com/38-rcolorbrewers-palettes.html">Source</a></p>
</div>
<p>That is a very nice and clean bar chart.
We can quite easily see that there are 7 main phyla (excluding other).
Additionally, it seems that the biggest difference between samples is due to the media used.</p>
</div>
<div id="plot-by-media" class="section level2" number="11.4">
<h2><span class="header-section-number">11.4</span> Plot by media</h2>
<p><img src="figures/petri.png" width="15%" style="display: block; margin: auto;" /></p>
<p>One of the many great advantages of <code>ggplot2</code> is that you can create plots tailored to your metadata.
As an example we will create another taxonomy bar chart but have one bar for each media site.</p>
<p>We’ll use a very similar code as the above bar chart but add the <code>microbiome::plot_composition</code> option <code>average_by =</code>.
With this we will average the rare aggregated relative abundance values by the four different media.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot by media</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>phylum_media_bar <span class="ot">&lt;-</span> </span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(phylum_rareaggregate_pseq, </span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">otu.sort =</span><span class="st">&quot;abundance&quot;</span>,</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">average_by =</span> <span class="st">&quot;media&quot;</span>) <span class="sc">+</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Phylum relative abundance bar chart by media&quot;</span>) <span class="sc">+</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="st">&quot;Phylum&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Paired&quot;</span>)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png</span></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./phylum_relabund_media_rareaggregate.png&quot;</span>, <span class="at">plot =</span> phylum_media_bar,</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./phylum_relabund_media_rareaggregate.png&quot;</span>)</span></code></pre></div>
<p>Now we get a good picture of which medias have the highest and lowest phlya diversity.
The Environmental samples (ENV) definitely have the highest diversity with clear presence of all phyla.</p>
Which media has the highest relative abundance of <em>Protoebacteria</em> and the lowest phyla diversity (excluding ENV)?
<div id="radio_UERCMOWWJV" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_UERCMOWWJV" value="answer"></input> <span><strong>CVP</strong></span></label><label><input type="radio" autocomplete="off" name="radio_UERCMOWWJV" value=""></input> <span><strong>KBC</strong></span></label><label><input type="radio" autocomplete="off" name="radio_UERCMOWWJV" value=""></input> <span><strong>TSA</strong></span></label>
</div>
Which media has the lowest relative abundance of <em>Protoebacteria</em> and the highest phyla diversity (excluding ENV)?
<div id="radio_FMVGICNROK" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_FMVGICNROK" value=""></input> <span><strong>CVP</strong></span></label><label><input type="radio" autocomplete="off" name="radio_FMVGICNROK" value=""></input> <span><strong>KBC</strong></span></label><label><input type="radio" autocomplete="off" name="radio_FMVGICNROK" value="answer"></input> <span><strong>TSA</strong></span></label>
</div>
<p>Later in this book we will investigate diversity more thoroughly with alpha and beta diversity visualisation and statistics.</p>
</div>
<div id="plot-by-site-and-media" class="section level2" number="11.5">
<h2><span class="header-section-number">11.5</span> Plot by site and media</h2>
<p><img src="figures/site_and_media.png" width="35%" style="display: block; margin: auto;" /></p>
<p><strong>Task</strong>: In the <code>sample_data</code> there is a column called <code>site.media</code>.
Use this column to create a phyla bar chart averaged by site and media.
Please make a good effort at the task before look at the solution in the below box.</p>
<div class="webex-solution">
<button>
Site and media bar chart solution
</button>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot by site and media</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>phylum_site_media_bar <span class="ot">&lt;-</span> </span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(phylum_rareaggregate_pseq, </span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">otu.sort =</span><span class="st">&quot;abundance&quot;</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">average_by =</span> <span class="st">&quot;site.media&quot;</span>) <span class="sc">+</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Phylum relative abundance bar chart by site &amp; media&quot;</span>) <span class="sc">+</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="st">&quot;Phylum&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Paired&quot;</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./phylum_relabund_site_media_rareaggregate.png&quot;</span>, <span class="at">plot =</span> phylum_site_media_bar,</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./phylum_relabund_site_media_rareaggregate.png&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="taxa-plots-summary" class="section level2" number="11.6">
<h2><span class="header-section-number">11.6</span> Taxa plots: summary</h2>
<p><img src="figures/sum_purple.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In this chapter we have:</p>
<ul>
<li>Created taxa bar charts with <code>microbiome</code> and <code>ggplot2</code>.</li>
<li>Saved <code>ggplot2</code> plots as <code>pngs</code> with <code>ggsave()</code>.</li>
<li>Displayed <code>pngs</code> with <code>IRdisplay::display_png()</code>.</li>
<li>Produced taxa heatmaps.</li>
<li>Aggregated rare taxa based on detection and prevalence.</li>
<li>Grouped bars in bar charts by metadata groups.</li>
</ul>
<p>We will reinforce this by producing Family and Genus based plots in the next chapter.</p>

</div>
</div>
<div id="family_genus_chap" class="section level1" number="12">
<h1><span class="header-section-number">Chapter 12</span> Family &amp; Genus plots</h1>
<p><img src="figures/plots.png" width="20%" style="display: block; margin: auto;" /></p>
<p>In this chapter you will:</p>
<ul>
<li>Create Family plots by following example code.</li>
<li>Be tasked to create Genus plots with little instruction.</li>
</ul>
<p>I highly encourage you to copy previous code and edit it.
It is a major time saver.
However, it can be easy to overlook some edits.
It is a good idea to double check your edited code to ensure you have changed the relevant options, and object &amp; file names.</p>
<div id="family-taxa-plots" class="section level2" number="12.1">
<h2><span class="header-section-number">12.1</span> Family taxa plots</h2>
<p><img src="figures/family.png" width="15%" style="display: block; margin: auto;" /></p>
<p>This part will show you a nice workflow for creating some family based taxa plots.
Run this at the bottom of your “3-taxonomy_barchart.ipynb” notebook.</p>
<div id="aggregate-families" class="section level3" number="12.1.1">
<h3><span class="header-section-number">12.1.1</span> Aggregate families</h3>
<p><img src="figures/aggregate_taxa.png" width="15%" style="display: block; margin: auto;" /></p>
<p>First we will create a <code>phyloseq</code> object by aggregating the families from the relative abundance <code>phyloseq</code> object.
Let us also then check the family <code>phyloseq</code> object.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Family phyloseq</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>family_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_taxa</span>(pseq_relabund, <span class="st">&quot;Family&quot;</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Head of phylum relative abundance table</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(family_pseq))</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of families</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Number of families: &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(family_pseq)))</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise</span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(family_pseq)</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(family_pseq)</span></code></pre></div>
<p>We have 112 families. Quite a bit but hopefully they will all fit into a heatmap.</p>
</div>
<div id="family-heatmap" class="section level3" number="12.1.2">
<h3><span class="header-section-number">12.1.2</span> Family heatmap</h3>
<p><img src="figures/heat_topology.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We will create a heatmap as before with one added <code>ggplot2</code> component.
As there are a lot of Families (112) and fewer samples (36) it will be easier to view if we have the families on the y-axis.
By default <code>microbiome::plot_composition()</code> will plot the taxa/families on the x-axis.
Thankfully we can use the <code>ggplot2()</code> function/component <code>coord_flip()</code> to flip/swap the x and y axes.</p>
<p><strong>Note</strong>: The <code>xlab()</code> and <code>ylab()</code> are specified the before this flipping.
E.g. <code>xlab()</code> is set to “Family” which will appear on the y axis because of the flip.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Family heatmap</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>family_heatmap <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(family_pseq, <span class="at">plot.type =</span> <span class="st">&quot;heatmap&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Family&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Family relative abundance heatmap&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Flip the x and y axes</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./family_relabund_heatmap.png&quot;</span>, <span class="at">plot =</span> family_heatmap,</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./family_relabund_heatmap.png&quot;</span>)</span></code></pre></div>
<p>The plot is quite large.
You can right click on the image and click “Open Image in New Tab” to view the whole image in a new tab.</p>
<p>With this heatmap we can see that the KBC and TSA samples have a high relative abundance of <em>Pseudomonadaceae</em> whilst the TSA samples have a high relative abundance of <em>Aeromonadaceae</em>.
Additionally, we can see that a lot of families have relatively low relative abundances.</p>
<p>Once you are finished looking at the plot, close the tab containing the plot.
This will help to keep the used resources low on the cluster.
Next we’ll carry out our rare aggregation and produce some bar charts.</p>
</div>
<div id="family-rare-aggregation" class="section level3" number="12.1.3">
<h3><span class="header-section-number">12.1.3</span> Family rare aggregation</h3>
<p><img src="figures/elephant_origami.png" width="15%" style="display: block; margin: auto;" /></p>
<p>To aggregate our rare families we need to choose a suitable <a href="#aggregateraretaxa">detection and prevalence threshold</a>.
Generally, this is a process of trial and error trying to get a total of 12 taxa.
Twelve is a good number as it is the total number of colours in the Paired palette from Colour brewer.
You may need more in your own analysis so it is not a hard limit.</p>
<p>Therefore, lets try out various thresholds and see how many families we get.
To get a good idea where to start we can get a summary of the mean relative abundance values of the families.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summary of row means (families) from the family otu table</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">rowMeans</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(family_pseq)))</span></code></pre></div>
<p>We have a lot of families so we will want to remove a good amount of them.
For our first attempt at rare aggregation we will use a value near the 3rd Quartile (0.0021601).
This will hopefully remove at least 3/4 of the families.</p>
<p>We’ll set up the code a bit differently so we can easily copy and paste it to try out various other values of detection and threshold.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Tests</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="co">#1</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.002</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">5</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of families = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span></code></pre></div>
<p>That gives us 56 families which is too high for a nice visualisation. We’ll therefore try a bunch of trial and error to get 12 families after aggregation.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#2 Increase prevalence threshold</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.002</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of families = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a><span class="co">#3 Increase detection threshold</span></span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.005</span></span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-17"><a href="#cb58-17" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb58-18"><a href="#cb58-18" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of families = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb58-19"><a href="#cb58-19" aria-hidden="true" tabindex="-1"></a><span class="co">#4 Increase detection threshold</span></span>
<span id="cb58-20"><a href="#cb58-20" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb58-21"><a href="#cb58-21" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb58-22"><a href="#cb58-22" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb58-23"><a href="#cb58-23" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb58-24"><a href="#cb58-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb58-25"><a href="#cb58-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-26"><a href="#cb58-26" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb58-27"><a href="#cb58-27" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of families = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb58-28"><a href="#cb58-28" aria-hidden="true" tabindex="-1"></a><span class="co">#5 Increase prevalence threshold</span></span>
<span id="cb58-29"><a href="#cb58-29" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb58-30"><a href="#cb58-30" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb58-31"><a href="#cb58-31" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb58-32"><a href="#cb58-32" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb58-33"><a href="#cb58-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb58-34"><a href="#cb58-34" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-35"><a href="#cb58-35" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb58-36"><a href="#cb58-36" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of families = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb58-37"><a href="#cb58-37" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove unwanted test_pseq</span></span>
<span id="cb58-38"><a href="#cb58-38" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(test_pseq)</span></code></pre></div>
<p>Super!
A Detection threshold of 0.02 and a prevalence threshold of 20/100 gives us 12 families.
We’ll now create a new <code>phyloseq</code> object using this info in a new cell.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Aggregate families phyla</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>family_rareaggregate_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> <span class="fl">0.02</span>, <span class="at">prevalence =</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Choosing a good detection and prevalence threshold is quite difficult.
It takes a lot of trial and practice and I find the thresholds to not be very intuitive to understand.
One piece of advise is to only change one threshold at a time during your trial and errors.</p>
</div>
<div id="family-bar-chart" class="section level3" number="12.1.4">
<h3><span class="header-section-number">12.1.4</span> Family bar chart</h3>
<p><img src="figures/chocolate_bar.png" width="15%" style="display: block; margin: auto;" /></p>
<p>With our family rare aggregated <code>phyloseq</code> object we’ll create a bar chart.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>family_bar <span class="ot">&lt;-</span> </span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(family_rareaggregate_pseq,</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>                               <span class="co">#otu.sort to sort family by abundance</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">otu.sort =</span> <span class="st">&quot;abundance&quot;</span>, <span class="at">group_by =</span> <span class="st">&quot;media&quot;</span>) <span class="sc">+</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Family relative abundance bar chart&quot;</span>) <span class="sc">+</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Change colours of families to that of the &quot;Paired&quot; palette from colour brewer</span></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="st">&quot;Family&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Paired&quot;</span>)</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png</span></span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./family_relabund_rareaggregate.png&quot;</span>, <span class="at">plot =</span> family_bar,</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./family_relabund_rareaggregate.png&quot;</span>)</span></code></pre></div>
<p>Brilliant, we are now finished with the Family taxa plots. Onto something harder.</p>
</div>
</div>
<div id="genus-taxa-plots" class="section level2" number="12.2">
<h2><span class="header-section-number">12.2</span> Genus taxa plots</h2>
<p><img src="figures/ginger.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In this section you are running your own analysis for the Genus information in the “3-taxonomy_barchart.ipynb” notebook.
Carry out the following tasks with the relative abundance <code>phyloseq</code> object (<code>pseq_relabund</code>):</p>
<ul>
<li>Create an aggregated genus <code>phyloseq</code> object and summarise it (no rare aggregation).</li>
<li>Create a heatmap of the genera relative abundances.</li>
<li>Carry out genus rare aggregation so the resulting <code>phyloseq</code> object only contains 12 genera.</li>
<li>Use the rare aggregated <code>phyloseq</code> object to create a bar chart grouped by <strong>media</strong>.</li>
</ul>
<p>In essence you are going through the workflow for the family data but for genus instead.
Again, I encourage you to copy, paste, and edit previous code.
You will most likely need to change names of objects and also change various options.</p>
<div id="genus-taxa-plots-solutions" class="section level3" number="12.2.1">
<h3><span class="header-section-number">12.2.1</span> Genus taxa plots solutions</h3>
<p><img src="figures/ginger_solutions.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Please give the task a good try before looking at the solutions in the expandable boxes.
However, if you are really stuck or want to compare your code to mine please have a look.
Please, remember that there are many ways to do the same thing.
If your code is different but it works and you understand it then that is fine.</p>
<div class="webex-solution">
<button>
Aggregate genus and summarise
</button>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Genus phyloseq</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>genus_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_taxa</span>(pseq_relabund, <span class="st">&quot;Genus&quot;</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Head of genus relative abundance table</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(genus_pseq))</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of genera</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Number of families: &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(genus_pseq)))</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(genus_pseq)</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(genus_pseq)</span></code></pre></div>
</div>
<div class="webex-solution">
<button>
Heatmap
</button>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Genus heatmap</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>genus_heatmap <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(genus_pseq, <span class="at">plot.type =</span> <span class="st">&quot;heatmap&quot;</span>) <span class="sc">+</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Genus&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Genus relative abundance heatmap&quot;</span>) <span class="sc">+</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Flip the x and y axes</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./genus_relabund_heatmap.png&quot;</span>, <span class="at">plot =</span> genus_heatmap,</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">450</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./genus_relabund_heatmap.png&quot;</span>)</span></code></pre></div>
</div>
<div class="webex-solution">
<button>
Aggregate rare genus
</button>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summary of row means (families) from the family otu table</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">rowMeans</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(genus_pseq)))</span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Tests</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co">#1</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.001</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">5</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Genus&quot;</span>,</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of genera = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a><span class="co">#2</span></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.01</span></span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">5</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Genus&quot;</span>,</span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb64-19"><a href="#cb64-19" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of genera = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb64-20"><a href="#cb64-20" aria-hidden="true" tabindex="-1"></a><span class="co">#3</span></span>
<span id="cb64-21"><a href="#cb64-21" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.01</span></span>
<span id="cb64-22"><a href="#cb64-22" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb64-23"><a href="#cb64-23" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb64-24"><a href="#cb64-24" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Genus&quot;</span>,</span>
<span id="cb64-25"><a href="#cb64-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb64-26"><a href="#cb64-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-27"><a href="#cb64-27" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb64-28"><a href="#cb64-28" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of genera = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb64-29"><a href="#cb64-29" aria-hidden="true" tabindex="-1"></a><span class="co">#4</span></span>
<span id="cb64-30"><a href="#cb64-30" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.005</span></span>
<span id="cb64-31"><a href="#cb64-31" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb64-32"><a href="#cb64-32" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb64-33"><a href="#cb64-33" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Genus&quot;</span>,</span>
<span id="cb64-34"><a href="#cb64-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb64-35"><a href="#cb64-35" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-36"><a href="#cb64-36" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb64-37"><a href="#cb64-37" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of genera = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb64-38"><a href="#cb64-38" aria-hidden="true" tabindex="-1"></a><span class="co">#5</span></span>
<span id="cb64-39"><a href="#cb64-39" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fl">0.006</span></span>
<span id="cb64-40"><a href="#cb64-40" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb64-41"><a href="#cb64-41" aria-hidden="true" tabindex="-1"></a>test_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb64-42"><a href="#cb64-42" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Genus&quot;</span>,</span>
<span id="cb64-43"><a href="#cb64-43" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> d, <span class="at">prevalence =</span> p</span>
<span id="cb64-44"><a href="#cb64-44" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-45"><a href="#cb64-45" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;Detection = &quot;</span>, d, <span class="st">&quot;, Prevalence = &quot;</span>, p,</span>
<span id="cb64-46"><a href="#cb64-46" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;, Number of genera = &quot;</span>, <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(test_pseq)))</span>
<span id="cb64-47"><a href="#cb64-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-48"><a href="#cb64-48" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove unwanted test_pseq</span></span>
<span id="cb64-49"><a href="#cb64-49" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(test_pseq)</span></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Aggregate genus phyla</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>genus_rareaggregate_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="at">level =</span> <span class="st">&quot;Genus&quot;</span>,</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">detection =</span> <span class="fl">0.006</span>, <span class="at">prevalence =</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div class="webex-solution">
<button>
Bar chart
</button>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>genus_bar <span class="ot">&lt;-</span> </span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(genus_rareaggregate_pseq,</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>                               <span class="co">#otu.sort to sort genera by abundance</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">otu.sort =</span> <span class="st">&quot;abundance&quot;</span>, <span class="at">group_by =</span> <span class="st">&quot;media&quot;</span>) <span class="sc">+</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative abundance&quot;</span>) <span class="sc">+</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Genus relative abundance bar chart&quot;</span>) <span class="sc">+</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Change colours of genera to that of the &quot;Paired&quot; palette from colour brewer</span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="st">&quot;Genus&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Paired&quot;</span>)</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./genus_relabund_rareaggregate.png&quot;</span>, <span class="at">plot =</span> genus_bar,</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./genus_relabund_rareaggregate.png&quot;</span>)</span></code></pre></div>
</div>
<p>Once you are happy you can <strong>Close and Halt</strong> the notebook.</p>

</div>
</div>
</div>



<div id="diversity-analysis-intro" class="section level1" number="13">
<h1><span class="header-section-number">Chapter 13</span> Diversity analysis intro</h1>
<p><img src="figures/diversity.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In the next 4 chapters we will carry out diversity analyses. This will include calculating and visualising diversity values plus carrying out the associated statistics.</p>
<p>In the chapters we will carry out:</p>
<ul>
<li><a href="#rarefaction_chap">Rarefaction of our ASV table.</a></li>
<li><a href="#alpha_chap">Alpha diversity of the rarefied data.</a></li>
<li><a href="#beta_chap">Beta diversity of the rarefied data.</a></li>
<li><a href="#DA_chap">Biomarker detection.</a></li>
</ul>
<p>During this we will attempt to answer our questions from <a href="#summary-questions">Chapter 2</a>:</p>
<ul>
<li>How does the bacterial communities change across the anthropisation gradient?</li>
<li>Is there a difference in the replicates of one site and media combination. I.e. do any of the media produce inconsistent profiles.</li>
<li>Is there more difference between the sites or the media used?</li>
<li>Do any of the media produce a similar taxonomic profile to the environmental sample?</li>
</ul>

</div>
<div id="rarefaction_chap" class="section level1" number="14">
<h1><span class="header-section-number">Chapter 14</span> Rarefaction</h1>
<p><img src="figures/whale.png" width="20%" style="display: block; margin: auto;" /></p>
<p>We are going to create a rarefied abundance table in this chapter.
This can be a controversial topic.
Please make your own decision if you want to do this or not in your own analysis.</p>
<div id="rarefaction-setup" class="section level2" number="14.1">
<h2><span class="header-section-number">14.1</span> Rarefaction: setup</h2>
<p><img src="figures/library_1.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We will use a new notebook called <strong>“4-rarefaction.ipynb”</strong>.
Add the following to the top of this new notebook to load the required libraries/packages and data.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;vegan&quot;</span>)</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Load ASV count vector</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;num_asvs_vec.RData&quot;</span>)</span></code></pre></div>
<p>Ensure you add markdown and code cells to this notebook to give yourself
a good structure.</p>
</div>
<div id="rarefaction-curve" class="section level2" number="14.2">
<h2><span class="header-section-number">14.2</span> Rarefaction curve</h2>
<p><img src="figures/rarefaction.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Prior to rarefying our data we need to determine the depth we want to
use. We can carry this out with a rarefaction curve.</p>
<p>A rarefaction curve is produced by randomly sampling the sequences in a
sample (without replacement). The rarefaction curve extracts
the number of unique ASVs in the first N sequences of each sample.
N is equal to the step size which
we will set as 50 (<code>step = 50</code>).
This is followed by the number of unique ASVs found in 2N, then 3N etc.</p>
<p>The plot will show the number of total unique ASVs it has found
against the depth it has currently sampled. Therefore, after 10 steps it
will be at a depth of 500 on the x-axis (50 * 10) and it will show how
many unique ASVs it has currently discovered from all 10 samplings.</p>
<div id="bird-analogy" class="section level3" number="14.2.1">
<h3><span class="header-section-number">14.2.1</span> Bird analogy</h3>
<p><img src="figures/ibyster.png" width="15%" style="display: block; margin: auto;" /></p>
<p>You are cataloguing the bird diversity (ASV diversity) in three forests/woods
in Hampshire (3 samples of our abundance table). You record your
findings over a week for each forest. Your record on the cumulative
amount of unique species you find over the 3 separate weeks are in the
table below:</p>
<table>
<caption>Cumulative bird species found in Hampshire forests/woods over a week</caption>
<colgroup>
<col width="20%" />
<col width="24%" />
<col width="27%" />
<col width="27%" />
</colgroup>
<thead>
<tr class="header">
<th>Day</th>
<th>New Forest (93900 acres)</th>
<th>Telegraph woods (50 acres)</th>
<th>Old Lords wood (4.67 acres)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>12</td>
<td>7</td>
<td>8</td>
</tr>
<tr class="even">
<td>2</td>
<td>21</td>
<td>12</td>
<td>10</td>
</tr>
<tr class="odd">
<td>3</td>
<td>32</td>
<td>16</td>
<td>11</td>
</tr>
<tr class="even">
<td>4</td>
<td>40</td>
<td>21</td>
<td>13</td>
</tr>
<tr class="odd">
<td>5</td>
<td>49</td>
<td>23</td>
<td>13</td>
</tr>
<tr class="even">
<td>6</td>
<td>60</td>
<td>24</td>
<td>13</td>
</tr>
<tr class="odd">
<td>7</td>
<td>65</td>
<td>25</td>
<td>13</td>
</tr>
</tbody>
</table>
<p>This can then be plotted in a rarefaction plot like below.</p>
<pre><code>## Loading required package: permute</code></pre>
<pre><code>## Loading required package: lattice</code></pre>
<pre><code>## This is vegan 2.6-4</code></pre>
<p><img src="R-community-analysis_files/figure-html/unnamed-chunk-171-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>In this case we can be fairly confident that seven days is enough to find all the birds we can in the Old Lords wood (green).
I.e. the sampling depth is enough.
Just like bird watching, it is very hard to find all the species present in a sample with metabarcoding or shotgun metagenomics.
This can be caused by our barcodes not picking up all species, our sampling wasn’t perfect, some species are very rare, or errors that DNA sequencing introduces.</p>
<p>A week seems like a good amount for sampling the Telegraph woods (Blue) but it is possible that some more species have not been recorded.</p>
<p>Strikingly, a week does not appear to be enough time for the New Forest (red). The amount of new species decreased to 5 on day 7 but the curve has not yet plateaued.</p>
<p>This hopefully helps you understand why rarefaction curves are a useful measure of how well we have captured the diversity of samples at different depths.</p>
</div>
<div id="rarefaction-curve-simple-plot" class="section level3" number="14.2.2">
<h3><span class="header-section-number">14.2.2</span> Rarefaction curve: simple plot</h3>
<p><img src="figures/crayon.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We’ll use <code>vegan</code> to create our rarefaction curve like we did in
<a href="#chap9rarefaction">Chapter 9</a>.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract ASV table as data frame</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>asv_abund_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq)))</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Rarefaction curve</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Save as png file</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">filename =</span> <span class="st">&quot;./rarefaction_plot.png&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>,</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Produce plot</span></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">rarecurve</span>(</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> asv_abund_df, <span class="at">step =</span> <span class="dv">50</span>,</span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Read depth&quot;</span>,</span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;ASVs&quot;</span></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./rarefaction_plot.png&quot;</span>)</span></code></pre></div>
</div>
<div id="rarefaction-curve-better-plot" class="section level3" number="14.2.3">
<h3><span class="header-section-number">14.2.3</span> Rarefaction curve: better plot</h3>
<p><img src="figures/palette_brush.png" width="15%" style="display: block; margin: auto;" /></p>
<p>It is a useful plot but we can make it better!
We will do that with the following addition:</p>
<ul>
<li>Adding extra options to <code>vegan::rarecurve()</code>:
<ul>
<li><code>lwd</code>: Sets the line width of the plot.</li>
<li><code>label</code>: Turn the sample labels on (<code>T</code>) or off (<code>F</code>). We will turn them off as it is hard to read them all.</li>
<li><code>sample</code>: Draws a vertical line at the specified depth.
Additionally, draws a horizontal line for each sample showing
how many ASVs were discovered at the sampling depth.</li>
</ul></li>
</ul>
<p>We will initially use our minimum sample depth as our rarefaction/sampling depth.
This is a good idea if it seems like a good amount as it will allow us to keep all our samples after rarefaction.
Any samples with a lower depth than are final rarefaction size will be removed.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Improved plot saved as file</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">filename =</span> <span class="st">&quot;./rarefaction_plot.png&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>,</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">rarecurve</span>(</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> asv_abund_df, <span class="at">step =</span> <span class="dv">50</span>,</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Read depth&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;ASVs&quot;</span>, <span class="at">lwd=</span><span class="dv">1</span>, <span class="at">label =</span> F,</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq))</span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./rarefaction_plot.png&quot;</span>)</span></code></pre></div>
<p>From this plot we can see that most of the samples have plateu’d quite nicely at the minimum depth.
This indicates that this is a good sampling depth, i.e. depth to be chosen for rarefying the data.
The grey horizontal lines help show how many more ASVs are found in a sample after the chosen sampling depth.
This can be seen in the sample with the highest number of ASVs where 5-20 more ASVs are discovered with its final depth of ~16,000.</p>
</div>
</div>
<div id="rarefaction-slope" class="section level2" number="14.3">
<h2><span class="header-section-number">14.3</span> Rarefaction slope</h2>
<p><img src="figures/hill_landscape.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We can calculate the slope of each sample at our specified sampling depth.
A slope of 0 shows a flat horizontal line (what we want).
A slope of 1 shows a flat vertical line (not what we want).</p>
<p>To get the slope of each sample we will use <code>vegan::rareslope()</code>.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Rarefaction slopes</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>rarefaction_slopes <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">rareslope</span>(</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> asv_abund_df, <span class="at">sample =</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq))</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="co">#View slopes from lowest to highest value</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(rarefaction_slopes)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Summary of slopes</span></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(rarefaction_slopes)</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Histogram of slopes</span></span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Save as png</span></span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">filename =</span> <span class="st">&quot;./rarefaction_slopes_histogram.png&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>,</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">100</span>, <span class="at">width =</span> <span class="dv">200</span>)</span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot</span></span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(rarefaction_slopes)</span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb73-17"><a href="#cb73-17" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./rarefaction_slopes_histogram.png&quot;</span>)</span></code></pre></div>
<p>With the calculated slopes attempt the following questions:</p>
<ol style="list-style-type: decimal">
<li>Which sample has the largest slope value?
<div id="radio_HKJAVSHYXC" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_HKJAVSHYXC" value=""></input> <span><strong>UD_ENV_rep2</strong></span></label><label><input type="radio" autocomplete="off" name="radio_HKJAVSHYXC" value="answer"></input> <span><strong>MD_ENV_rep3</strong></span></label><label><input type="radio" autocomplete="off" name="radio_HKJAVSHYXC" value=""></input> <span><strong>LD_CVP_rep1</strong></span></label>
</div></li>
<li>Which sample has the lowest slope value?
<div id="radio_DDOHKYYLXO" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_DDOHKYYLXO" value=""></input> <span><strong>UD_ENV_rep2</strong></span></label><label><input type="radio" autocomplete="off" name="radio_DDOHKYYLXO" value=""></input> <span><strong>MD_ENV_rep3</strong></span></label><label><input type="radio" autocomplete="off" name="radio_DDOHKYYLXO" value="answer"></input> <span><strong>LD_CVP_rep1</strong></span></label>
</div></li>
<li>Which samples have larger slope values; the environmental samples or media samples?
<div id="radio_UNCFCKWMHB" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_UNCFCKWMHB" value="answer"></input> <span><strong>Environmental</strong></span></label><label><input type="radio" autocomplete="off" name="radio_UNCFCKWMHB" value=""></input> <span><strong>Media</strong></span></label>
</div></li>
</ol>
<p>Overall the slopes values are low (&lt;0.01).
It is noticeable that the environmental samples have higher slopes than the media samples.
Biologically, this makes sense as there is no bias being introduced by media.</p>
</div>
<div id="rarefy-data" class="section level2" number="14.4">
<h2><span class="header-section-number">14.4</span> Rarefy data</h2>
<p><img src="figures/pangolin.png" width="20%" style="display: block; margin: auto;" /></p>
<p>The minimum sample depth appears to be a good choice for rarefying.
It keeps all the samples and at this depth the samples have a very good coverage of the ASVs present in the data.
This is represented by the good plateauing in the rarefaction plot and the low valued slopes (&lt;0.01).
In your own analyses you may need to balance the loss of samples with the loss of depth.</p>
<p>With that decision we will rarefy our data to the minimum depth.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Rarefy to minimum depth</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>  pseq, <span class="at">sample.size =</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq)),</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">rngseed =</span> <span class="dv">1000</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>The options we provided to the function <code>rarefy_even_depth</code> were:</p>
<ul>
<li><strong><code>pseq</code></strong>: The <code>phyloseq</code> object to rarefy.</li>
<li><strong><code>sample.size =</code></strong>: The sampling depth to rarefy to.</li>
<li><strong><code>rngseed</code></strong>: This is the seed used for random subsampling.
<ul>
<li>If you rarefy the data again with the same seed it will extract the same data.</li>
<li>This is useful so you and others can replicate your work.</li>
<li>Additionally, it means you will get the same output as me (the writer) so you can accurately compare your findings to mine.</li>
</ul></li>
</ul>
<p>As always, it is useful to check our data.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise and check sample counts which should each amount to 10433 (min depth)</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq_rarefy)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_rarefy)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a><span class="co">#ASV counts</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Add relative abundance ASV count</span></span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>num_asvs_vec[<span class="st">&quot;rarefied&quot;</span>] <span class="ot">&lt;-</span> <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy))</span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a>num_asvs_vec</span></code></pre></div>
<p>With this we can see that all our samples have the correct depth (10,433) and very few ASVs have been lost (2551-2498 = 53).
I would definetly be happy with this outcome.</p>
<p>Once you have viewed the outputs save your <code>phyloseq</code> object and ASV count vector.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Phyloseq save</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(pseq_rarefy, <span class="at">file =</span><span class="st">&quot;phyloseq_rarefied.RData&quot;</span>)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="co">#ASV count save</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(num_asvs_vec, <span class="at">file=</span><span class="st">&quot;num_asvs_vec.v2.RData&quot;</span>)</span></code></pre></div>
<p>Now you can close and halt the notebook.</p>
</div>
<div id="rarefaction-summary" class="section level2" number="14.5">
<h2><span class="header-section-number">14.5</span> Rarefaction: summary</h2>
<p><img src="figures/sum_orange.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In this chapter we:</p>
<ul>
<li>Determined the sampling depth to rarefy to with rarefaction curves and slopes.</li>
<li>Rarefied our samples to the minimum read depth. This retained all our samples whilst using a good depth.</li>
</ul>
<p>We now have an abundance table, relative abundance table, and rarefied abundance table. We will use these to carry out some analyses.</p>

</div>
</div>
<div id="alpha_chap" class="section level1" number="15">
<h1><span class="header-section-number">Chapter 15</span> Alpha diversity</h1>
<p><img src="figures/alpha.png" width="20%" style="display: block; margin: auto;" /></p>
<p>In this chapter we are going to produce alpha diversity plots and statistics with our rarefied data.
Alpha diversity is carried out by producing a metric for each sample.
These metrics are primarily interested in the evenness or diversity of a sample.</p>
<p>Generally, these metrics are carried out on the ASV data to see the most fine grain differences between samples.
However, you can also use different taxa levels which we will show later on.
This can be useful if there is too much difference in the ASVs from sample to sample or you can only accurately classify sequences to a high level taxonomy (such as with <code>KRAKEN2</code> output).</p>
<p>Once we have a value for each sample they can be used to compare samples and sample groups.
Box plots are used for visualisation.
Statistics are used to determine if there is a significant difference between groups.</p>
<p>We’ll start out by producing a basic box plot and statistics.
Next we’ll improve the plot with some extra R packages.
With these tools we will compare the metrics across the various metadata fields we have.</p>
<div id="alpha-setup" class="section level2" number="15.1">
<h2><span class="header-section-number">15.1</span> Alpha: setup</h2>
<p><img src="figures/gear_setup.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We will use a new notebook called <strong>“5-alpha_diversity.ipynb”</strong>.
Add the following to the top of this new notebook to load the required libraries/packages and data.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ggforce&quot;</span>)</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq_rarefied.RData&quot;</span>)</span></code></pre></div>
<p>Reminder to use markdown and code cells in the notebook.</p>
</div>
<div id="alpha-simple-plot" class="section level2" number="15.2">
<h2><span class="header-section-number">15.2</span> Alpha: simple plot</h2>
<p><img src="figures/simple_boxplot.png" width="15%" style="display: block; margin: auto;" /></p>
<p>The first step is to calculate and plot the alpha diversity metrics based on the ASV data.
Thankfully, the command <code>phyloseq::plot_richness()</code> will do both of these.
Try it out.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Simple alpha plot</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>alpha_plot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy)</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./alpha_div.png&quot;</span>, <span class="at">plot =</span> alpha_plot,</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./alpha_div.png&quot;</span>)</span></code></pre></div>
<p>The resulting plot is pretty good but is difficult to read.
We can do 3 major things to make the contents easier to read.</p>
<div id="metric-choice" class="section level3" number="15.2.1">
<h3><span class="header-section-number">15.2.1</span> Metric choice</h3>
<p><img src="figures/metric_square_1.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We can choose certain alpha diversity metrics to plot.
Currently it is showing 7 metrics (Observed, Chao1 etc.).
I like to use:</p>
<ul>
<li><strong>Observed</strong>: This is the number of observed features (ASVs, Genera, etc) in each sample.</li>
<li><strong>Chao1</strong>: This is an estimate number of the total number of features in each sample.</li>
<li><strong>Shannon</strong>: A measure of diversity where a higher number means higher diversity. Shannon’s index accounts for both abundance and evenness of the features present.</li>
</ul>
<p>We’ll choose these metrics with the <code>measures =</code> option.
Either make a new code block with new code or you can edit and run the previous code you wrote.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Simple alpha plot</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>alpha_plot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./alpha_div.png&quot;</span>, <span class="at">plot =</span> alpha_plot,</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./alpha_div.png&quot;</span>)</span></code></pre></div>
<p>To see which measures are available you can see the help page of <code>phyloseq::plot_richness()</code>.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>?phyloseq<span class="sc">::</span>plot_richness</span></code></pre></div>
<p>For a list and full description of some of the metrics please see the <a href="#alpha_appendix">APPENDIX</a>.</p>
</div>
<div id="sample-grouping-box-plot" class="section level3" number="15.2.2">
<h3><span class="header-section-number">15.2.2</span> Sample grouping &amp; box plot</h3>
<p><img src="figures/boxplots_groups.png" width="15%" style="display: block; margin: auto;" /></p>
<p>The last plot looks a bit better.
To actually make them box plots we’ll change the x axis so they are based on media rather than individual samples.</p>
<p>The <code>phyloseq::plot_richness()</code> function produces a <code>ggplot2</code> object.
To convert our dot plot into a box plot we can add the layer <code>geom_boxplt()</code> like with our <a href="#depth-boxplots">depth boxplots</a>.
This will combine samples from one media into one IQR box.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Box plot</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, </span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">&quot;media&quot;</span>,</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>)) <span class="sc">+</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>            ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./alpha_div.png&quot;</span>, <span class="at">plot =</span> alpha_plot,</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./alpha_div.png&quot;</span>)</span></code></pre></div>
</div>
<div id="png-height-and-width-choice" class="section level3" number="15.2.3">
<h3><span class="header-section-number">15.2.3</span> PNG height and width choice</h3>
<p><img src="figures/display_png.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Choosing a good height and width size can make plots easier to read.
We’ll try 150 (height) and 250 (width).</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Produce ggplot object of boxplot</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>alpha_boxplot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, </span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">&quot;media&quot;</span>,</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>) <span class="sc">+</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>                          ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Alpha_diversity_media_boxplot.png&quot;</span>, <span class="at">plot =</span> alpha_boxplot,</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">250</span>)</span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Alpha_diversity_media_boxplot.png&quot;</span>)</span></code></pre></div>
<p>How do I choose a good height and width?
This comes down to trial and error and experience.
You want something with a good aspect ratio for the plot and you want all the text to be clear.
Additionally, the size requirements of figures for journals will play a big consideration as well.
One of the nice things with creating jupyter-notebooks or R markdowns with all your code is that you can go back, edit, and rerun the code relatively easily and quickly.</p>
</div>
</div>
<div id="alpha-stats" class="section level2" number="15.3">
<h2><span class="header-section-number">15.3</span> Alpha: stats</h2>
<p><img src="figures/pear.png" width="10%" style="display: block; margin: auto;" /></p>
<p>Along with the plot it is also good to carry out statistical tests.
This will allow us to statistically determine if the alpha diversity values between groups are significantly different.</p>
<p>Before carrying out statistics we will need a data frame with the alpha diveristy values for each sample.
This can be carried out with the function<code>phyloseq::estimate_richness()</code>.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Produce data frame of all alpha diversity values</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">estimate_richness</span>(<span class="at">physeq =</span> pseq_rarefy)</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(alpha_df)</span></code></pre></div>
<p>We have all the alpha metrics in the new data frame.
Additionally, we also have the standard error for some of the metrics (e.g. <code>se.chao1</code>).</p>
<p>With this information we can carry out a pairwise comparison using Wilcoxon rank sum test (the same way as QIIME2).
Base R comes with the function <code>pairwsie.wilcox.test()</code> to carry this out.
We’ll do this using media as our groups from the <code>sample_data()</code> and the Observed metric from the alpha_df.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Paired wilcoxon test</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Observed</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.wilcox.test</span>(alpha_df<span class="sc">$</span>Observed, phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy)<span class="sc">$</span>media)</span></code></pre></div>
<p>You can ignore the warning messages in this case.</p>
<p>Observe the results.</p>
Are all the pairwise comparisons significant different (&lt;0.05)?
<div id="radio_UFOIWDEURQ" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_UFOIWDEURQ" value="answer"></input> <span><strong>Yes</strong></span></label><label><input type="radio" autocomplete="off" name="radio_UFOIWDEURQ" value=""></input> <span><strong>No</strong></span></label>
</div>
<p>The test has carried out P value adjustments using the Holm-Bonferroni method (<code>P value adjustment method: holm</code>) which is good.</p>
<p>In this case the p-values are all identical as the values do not vary greatly but there are clear differences.
The only group with a large variance is the ENV group with values going from &lt; 150 to &gt; 450.</p>
<p>For more info on Wilcoxon test <a href="https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test">Wikipedia</a> is a good start.</p>
<div id="alpha-stats-task" class="section level3" number="15.3.1">
<h3><span class="header-section-number">15.3.1</span> Alpha stats task</h3>
<p><img src="figures/exercises.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Carry out the same statistical analysis (media as the grouping) with the <code>Chao1</code> and <code>Shannon</code> metrics.</p>
<div class="webex-solution">
<button>
Alpha stats solution
</button>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Chao1</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.wilcox.test</span>(alpha_df<span class="sc">$</span>Chao1, phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy)<span class="sc">$</span>media)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Shannon</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.wilcox.test</span>(alpha_df<span class="sc">$</span>Shannon, phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy)<span class="sc">$</span>media)</span></code></pre></div>
</div>
<p>With these we can see that the differences between the Chao1 values of the different media are significantly different.</p>
<p>However, the p-values when comparing the Shannon values are not all significant.
The only significant differences are between ENV &amp; KBC, and between KBC &amp; TSA.
Going back to the boxplot, does this look like the case?
I would say yes as the IQRs for CVP, KBC, and TSA all overlap.
There is little to no overlap of the IQRs of KBC and TSA against ENV.</p>
<p>We can therefore make these claims about the alpha diversity values of the groups.</p>
<ul>
<li>ENV has the highest number of observed ASVs and the highest diversity.</li>
<li>ENV appears to have a wider range of diversities across its samples.</li>
<li>The medias have much lower numbers of ASVs and diveristy compared to ENV.</li>
<li>Within each media grouping the samples have very similar numbers of ASVs present.</li>
</ul>
</div>
</div>
<div id="alpha-diversity-violin-plot" class="section level2" number="15.4">
<h2><span class="header-section-number">15.4</span> Alpha diversity: violin plot</h2>
<p><img src="figures/violin.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Boxplots are quite nice but violin plots can be even nicer.
They show the distribution of the data in the IQR better and in fact you can easily display each value.</p>
<p>We can change our boxplot to a violin plot by changing <code>ggplot2:geom_boxplot()</code> to <code>ggplot2::geom_violin()</code>.</p>
<p>Copy and paste the previous boxplot code to a new cell at the bottom of your notebook.
Edit and run the code so it looks like the below.</p>
<p><strong>Tip:</strong> Remember to change the name of your png file in the <code>ggsave()</code> and <code>IRdisplay::display_png()</code> functions.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Produce ggplot object of violin plot</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>alpha_violinplot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, </span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">&quot;media&quot;</span>,</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>) <span class="sc">+</span></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>                          ggplot2<span class="sc">::</span><span class="fu">geom_violin</span>()</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Alpha_diversity_media_violinplot.png&quot;</span>, <span class="at">plot =</span> alpha_violinplot,</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">250</span>)</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Alpha_diversity_media_violinplot.png&quot;</span>)</span></code></pre></div>
<p>We can also add semi transparent dots representing the values for each sample.
To do this, add the <code>ggplot2</code> layer <code>ggforce::geom_sina()</code>.
This has the option <code>alpha=0.5</code> where alpha resents the transparency of the dots and 0.5 represents 50% transparency.
Alpha is a common option used for many plots.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Produce ggplot object of violin plot</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>alpha_violinplot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, </span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">&quot;media&quot;</span>,</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>) <span class="sc">+</span></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>                          ggplot2<span class="sc">::</span><span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>                          ggforce<span class="sc">::</span><span class="fu">geom_sina</span>(<span class="at">alpha=</span><span class="fl">0.5</span>)</span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb87-8"><a href="#cb87-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Alpha_diversity_media_violinplot.png&quot;</span>, <span class="at">plot =</span> alpha_violinplot,</span>
<span id="cb87-9"><a href="#cb87-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">250</span>)</span>
<span id="cb87-10"><a href="#cb87-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb87-11"><a href="#cb87-11" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Alpha_diversity_media_violinplot.png&quot;</span>)</span></code></pre></div>
</div>
<div id="alpha-site-plots-and-statistics" class="section level2" number="15.5">
<h2><span class="header-section-number">15.5</span> Alpha: Site plots and statistics</h2>
<p><img src="figures/river.png" width="15%" style="display: block; margin: auto;" /></p>
<p>With all the knowledge, skills, and code from this chapter carry out the following tasks:</p>
<ul>
<li>Produce a violin plot of Observed, Chao1, Shannon, and Inverse Simpson with the Site groups.</li>
<li>Carry out paired Wilcoxon test for Observed, Chao1, Shannon, and Inverse Simpson with the Site groups.</li>
</ul>
<div class="webex-solution">
<button>
Alpha sites solution
</button>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Produce ggplot object of violin plot</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>alpha_boxplot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, </span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">&quot;site&quot;</span>,</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>) <span class="sc">+</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>                          ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>                          ggforce<span class="sc">::</span><span class="fu">geom_sina</span>(<span class="at">alpha=</span><span class="fl">0.5</span>)</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Alpha_diversity_site_violinplot.png&quot;</span>, <span class="at">plot =</span> phylum_bar,</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">250</span>)</span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Alpha_diversity_site_violinplot.png&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Paired wilcoxon test</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Observed</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.wilcox.test</span>(alpha_df<span class="sc">$</span>Observed, phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy)<span class="sc">$</span>site)</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Chao1</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.wilcox.test</span>(alpha_df<span class="sc">$</span>Chao1, phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy)<span class="sc">$</span>site)</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Shannon</span></span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.wilcox.test</span>(alpha_df<span class="sc">$</span>Shannon, phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy)<span class="sc">$</span>site)</span></code></pre></div>
</div>
<p>The plots and stats do not show anything particularly interesting in terms of site unfortunately.
There is too much difference caused by the media.
To view this we can add the option <code>aes(colour=media)</code> to <code>ggforce_sina()</code>.
This will colour the points by media.</p>
<p>If we only do this it will give us a legend with the title “colour”.
We can fix this by adding the layer `ggplot2::labs(colour = “Media”).</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Produce ggplot object of violin plot</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>alpha_violinplot <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(<span class="at">physeq =</span> pseq_rarefy, </span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">&quot;site&quot;</span>,</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Observed&quot;</span>,<span class="st">&quot;Chao1&quot;</span>,<span class="st">&quot;Shannon&quot;</span>) <span class="sc">+</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>                          ggplot2<span class="sc">::</span><span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>                          ggforce<span class="sc">::</span><span class="fu">geom_sina</span>(<span class="at">alpha=</span><span class="fl">0.5</span>, <span class="fu">aes</span>(<span class="at">colour=</span>media)) <span class="sc">+</span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>                          ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">&quot;Media&quot;</span>)</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Alpha_diversity_site_violinplot.png&quot;</span>, <span class="at">plot =</span> alpha_violinplot,</span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">250</span>)</span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Alpha_diversity_site_violinplot.png&quot;</span>)</span></code></pre></div>
<p>When you are happy you can close and halt the notebook.</p>
</div>
<div id="alpha-summary" class="section level2" number="15.6">
<h2><span class="header-section-number">15.6</span> Alpha: summary</h2>
<p><img src="figures/sum_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In this chapter we have created alpha diversity plots and run pairwise comparisons.
The biggest difference between samples is caused by the media choice which overwrites differences we can see between sites.
The environmental samples have much higher diversirty samples than the media samples but this is expected.</p>

</div>
</div>
<div id="beta_chap" class="section level1" number="16">
<h1><span class="header-section-number">Chapter 16</span> Beta</h1>
<p><img src="figures/beta.png" width="20%" style="display: block; margin: auto;" /></p>
<p>In this chapter we are going to carry out beta diversity analysis.
This will include the production of ordination plots and statistical analysis.</p>
<p>When carrying out beta diversity all samples are compared in a pairwise manner.
Using various metrics a table of pairwise distances is created.
These distance are dissimilarity measures with smaller values indicating higher similarity and larger values representing higher dissimilarity.</p>
<p>With these dissimilarity measures we can produce scatter plots based (e.g. PCoA, &amp; NMDS plots).
Samples with low dissimilarity (i.e. similar samples) will be close to each other (cluster).
Samples with high dissimilarity will be apart from each other.
In a perfect example samples from the same group will form clusters that are distinct from the clusters of other groups.</p>
<p>In this chapter we will:</p>
<ul>
<li>Calculate beta diversity ordination metrics and plot them as a scatter plot.</li>
<li>Determine if there is a statistical difference between beta diversity metrics of sample groups.</li>
</ul>
<div id="beta-setup" class="section level2" number="16.1">
<h2><span class="header-section-number">16.1</span> Beta: setup</h2>
<p><img src="figures/books.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We will use a new notebook called <strong>“6-beta_diversity.ipynb”</strong>.
Add the following to the top of this new notebook to load the required libraries/packages and data.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ggforce&quot;</span>)</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;vegan&quot;</span>)</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq_rarefied.RData&quot;</span>)</span></code></pre></div>
</div>
<div id="beta-ordination" class="section level2" number="16.2">
<h2><span class="header-section-number">16.2</span> Beta: ordination</h2>
<p><img src="figures/ordinate.png" width="15%" style="display: block; margin: auto;" /></p>
<p>In this section we will ordinate our data with the <code>phyloseq::ordinate()</code> function.
We will ordinate using the <a href="https://en.wikipedia.org/wiki/Multidimensional_scaling">MDS method</a> (AKA PCoA).
We will be using the Weighted Unifrac metric as beta diversity dissimilarity scores.
This uses the count data along with the phylogenetic distances to calculate pairwise dissimilarity.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Ordinate data</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>ord.mds.wunifrac <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">ordinate</span>(pseq_rarefy, <span class="at">method =</span> <span class="st">&quot;MDS&quot;</span>, <span class="at">distance =</span> <span class="st">&quot;wunifrac&quot;</span>)</span></code></pre></div>
<p>To see the different methods and distances that can used please try <code>?phyloseq:ordinate</code> and <code>phyloseq::distanceMethodList</code> respectively.
More information about the different distances please see the <a href="#beta_appendix">APPENDIX</a>.</p>
</div>
<div id="beta-plot-ordination" class="section level2" number="16.3">
<h2><span class="header-section-number">16.3</span> Beta: plot ordination</h2>
<p><img src="figures/Scatterplot_2.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Now that we have our ordination we can plot it.
The script below uses <code>phyloseq::plot_ordination()</code> to plot our ordination.
It needs the <code>phyloseq</code> object used for ordination (for the metadata) and the ordination data.
We will add the options <code>color =</code> and <code>shape =</code> to colour the points by site and shape the points by media.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot ordination</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>nmds.wunifrac <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_ordination</span>(pseq_rarefy, ord.mds.wunifrac,</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">color =</span> <span class="st">&quot;site&quot;</span>, <span class="at">shape =</span> <span class="st">&quot;media&quot;</span>)</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Beta_diversity_mds_wunifrac_media_site.png&quot;</span>, <span class="at">plot =</span> nmds.wunifrac,</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">125</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Beta_diversity_mds_wunifrac_media_site.png&quot;</span>)</span></code></pre></div>
<p>In this plot we can see that there are 3 main clusters:</p>
<ol style="list-style-type: decimal">
<li>TSA</li>
<li>ENV</li>
<li>CVP &amp; KBC</li>
</ol>
<p>A very important feature of ordination plots is the percentage values on the axes.
These are the percentage variation explained values.
These values show how much of the variation of the data is explained by each axis.
We need these as we are converting high dimensional data (many pairwise comparisons) into a 2 dimensional plot.</p>
<p>In our plot axis 1 has a very high percentage variation explained value of 72.5%.
This indicates that most of the variation in our data is explained by the differences between the media samples (CVP, KBC, &amp; TSA) and the ENV samples.
This is because the ENV samples are far removed (large distances) from the media samples in terms of axis 1.</p>
<p>On the other hand, axis.2 (15.8%) seems to be primarily showing the difference between the TSA samples and the other samples, and within the TSA samples.</p>
<p>The higher the total variation explained value is the better are plot represents the beta diversity in our data.
In this case the total is &gt;88% which is very good.</p>
<ul>
<li><strong>&lt;30%</strong> is poor</li>
<li><strong>&gt;30%</strong> is not very good</li>
<li><strong>&gt;50%</strong> is good</li>
<li><strong>&gt;70%</strong> is very good</li>
<li><strong>&gt;90%</strong> is exceptional</li>
</ul>
<p>Additionally, we can see that there seems to be some separation of the sites within the different media groups. This is clear in TSA and ENV but not in CVP and KBC since the cluster is very tight (short distances).</p>
</div>
<div id="beta-statistics" class="section level2" number="16.4">
<h2><span class="header-section-number">16.4</span> Beta: statistics</h2>
<p><img src="figures/cluster_colours.png" width="15%" style="display: block; margin: auto;" /></p>
<p>For our beta diversity we will carry out PERMANOVA analysis for the media groupings.
This will be carried out for all the samples and in a pairwise manner.</p>
<p>PERMANOVA is a good test in this case to determine if there is a significance difference between the distances of groups.
This allows us to determine if the clustering of groups is significant.
More info on PERMANOVA <a href="https://en.wikipedia.org/wiki/Permutational_analysis_of_variance#:~:text=Permutational%20multivariate%20analysis%20of%20variance,are%20equivalent%20for%20all%20groups.">Wikipedia</a> is a good start.</p>
<div id="permanova" class="section level3" number="16.4.1">
<h3><span class="header-section-number">16.4.1</span> PERMANOVA</h3>
<p><img src="figures/permanova.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We will carry out a PERMANOVA test on all the samples based on media.
This will help us determine if there is any differences between groupings but not which grouping.
Prior to PERMANOVA we need to create a matrix containing all the sample paired distance values.
We will use the function <code>phyloseq::distance()</code>.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PERMANOVA</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Produce distance matrix</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>wunifrac_dist_mat <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(pseq_rarefy, <span class="at">method=</span><span class="st">&quot;wunifrac&quot;</span>)</span></code></pre></div>
<p>With the distance matrix we can carry out a PERMANOVA test with <a href="https://search.r-project.org/CRAN/refmans/vegan/html/adonis.html"><code>vegan::adonis2()</code></a>.
This requires the metadata in the form of a data frame.
In this case we can assess if there are any differences by media and by site (<code>media+site</code>).</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract metadata data frame from phyloseq</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>metadf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy))</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a><span class="co">#PERMANOVA/ADONIS of media and site</span></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>wunifrac_adonis <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(wunifrac_dist_mat <span class="sc">~</span> media<span class="sc">+</span>site, <span class="at">data =</span> metadf, <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>)</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>wunifrac_adonis</span></code></pre></div>
<p>There is a lot of information in the resulting table.
We are primarily interested in the P-value column (<code>Pr(&gt;F)</code>).
We can see that significant differences are found within the groups for media (P = 0.001) and site (P = 0.009).</p>
</div>
<div id="pairwise-permanova" class="section level3" number="16.4.2">
<h3><span class="header-section-number">16.4.2</span> Pairwise PERMANOVA</h3>
<p><img src="figures/pear.png" width="10%" style="display: block; margin: auto;" /></p>
<p>We will now see which media groups are significantly different to each other.
We will do this with pairwise PERMANOVA tests.
Unfortunately the <code>vegan</code> library does not have a function to do this so we will need to do some scripting.</p>
<p>The first step is to get every combination of groupings with the <code>combn()</code> and <code>unique()</code> functions.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Pairwise comparisons</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Get combinations of unique media values</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>cbn <span class="ot">&lt;-</span> <span class="fu">combn</span>(<span class="at">x =</span> <span class="fu">unique</span>(metadf<span class="sc">$</span>media), <span class="at">m =</span> <span class="dv">2</span>)</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>cbn</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a><span class="co">#First combination</span></span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>cbn[,<span class="dv">1</span>]</span></code></pre></div>
<p>With the first media combination (CVP &amp; ENV) we can subset our <code>phyloseq</code> object so it only contains those samples.
Additionally, we can extract the metadata from this subsetted <code>phyloseq</code> object.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Subset phyloseq so it only contains CVP &amp; ENV samples (1st combination)</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>pseq_rarefy_subset <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq_rarefy, media <span class="sc">%in%</span> cbn[,<span class="dv">1</span>])</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Metadata data frame</span></span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>metadf_subset <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy_subset))</span></code></pre></div>
<p>With this info we can calculate the distances and then carry out the PERMANOVA.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Weighted unifrac distance calculation</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>wunifrac_dist_mat <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(pseq_rarefy_subset, <span class="at">method=</span><span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="co">#PERMANOVA/ADONIS of media</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>wunifrac_pairwise_adonis <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(wunifrac_dist_mat <span class="sc">~</span> media, <span class="at">data =</span> metadf_subset, <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>)</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>wunifrac_pairwise_adonis</span></code></pre></div>
<p>We can see that there is a significant difference between the CVP and ENV samples (P = 0.01).
However, it would be quite inefficient to type the above script for each combination.
We will therefore create a script with a loop to automatically run the code for each combination.
Additionally, we will create an empty data frame before the loop.
This will be used to store the comparison info and the P-values.</p>
<p>Create the below script in a new code block, reading the annotation to help understand how it works.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Pairwise comparison loop</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Get combinations of unique media values</span></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>cbn <span class="ot">&lt;-</span> <span class="fu">combn</span>(<span class="at">x =</span> <span class="fu">unique</span>(metadf<span class="sc">$</span>media), <span class="at">m =</span> <span class="dv">2</span>)</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Create empty final data frame with 4 columns</span></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a><span class="co"># and a number of rows equal to the the number of combinations</span></span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">matrix</span>(<span class="at">data =</span> <span class="cn">NA</span>, <span class="at">nrow =</span> <span class="fu">ncol</span>(cbn), <span class="at">ncol =</span> <span class="dv">4</span>))</span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Add column names</span></span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pairwise_permanova_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;1&quot;</span>,<span class="st">&quot;2&quot;</span>,<span class="st">&quot;p&quot;</span>,<span class="st">&quot;p.adj&quot;</span>)</span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Loop trhough the combinations</span></span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(cbn)){</span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Subset to media phyloseq</span></span>
<span id="cb99-13"><a href="#cb99-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Ensure to change metadata group if using different grouping (media)</span></span>
<span id="cb99-14"><a href="#cb99-14" aria-hidden="true" tabindex="-1"></a>  pseq_rarefy_subset <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq_rarefy, media <span class="sc">%in%</span> cbn[,i])</span>
<span id="cb99-15"><a href="#cb99-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Metadata data frame</span></span>
<span id="cb99-16"><a href="#cb99-16" aria-hidden="true" tabindex="-1"></a>  metadf_subset <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy_subset))</span>
<span id="cb99-17"><a href="#cb99-17" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Weighted unifrac distance calculation</span></span>
<span id="cb99-18"><a href="#cb99-18" aria-hidden="true" tabindex="-1"></a>  wunifrac_dist_mat <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(pseq_rarefy_subset, <span class="at">method=</span><span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb99-19"><a href="#cb99-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#PERMANOVA/ADONIS of media</span></span>
<span id="cb99-20"><a href="#cb99-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Ensure to change metadata group if using different grouping (media)</span></span>
<span id="cb99-21"><a href="#cb99-21" aria-hidden="true" tabindex="-1"></a>  wunifrac_pairwise_adonis <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(wunifrac_dist_mat <span class="sc">~</span> media, <span class="at">data =</span> metadf_subset, <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>)</span>
<span id="cb99-22"><a href="#cb99-22" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Add the gorup names and p-value to the main data frame</span></span>
<span id="cb99-23"><a href="#cb99-23" aria-hidden="true" tabindex="-1"></a>  pairwise_permanova_df[i,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> cbn[,i]</span>
<span id="cb99-24"><a href="#cb99-24" aria-hidden="true" tabindex="-1"></a>  pairwise_permanova_df[i,<span class="dv">3</span>] <span class="ot">&lt;-</span> wunifrac_pairwise_adonis[<span class="dv">1</span>,<span class="st">&quot;Pr(&gt;F)&quot;</span>]</span>
<span id="cb99-25"><a href="#cb99-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb99-26"><a href="#cb99-26" aria-hidden="true" tabindex="-1"></a><span class="co">#View data frame</span></span>
<span id="cb99-27"><a href="#cb99-27" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df</span></code></pre></div>
<p>All our media groupings are significantly different to each other (0.001).
You may have noticed that the <code>p.adj</code> column is empty.
We will add adjusted P-values to these columns.
It is always important to use adjusted P-values when carrying out multiple significant tests such as with pairwise comparisons.
We’ll use the function <code>p.adjust()</code> to add Benjamin-Hochberg (<code>method = "BH"</code>) adjusted values values.</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Add adjusted P-values</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df<span class="sc">$</span>p.adj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pairwise_permanova_df<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View data frame</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Write to a file</span></span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> pairwise_permanova_df, <span class="at">file =</span> <span class="st">&quot;pairwise_permanova_media_wunifrac.tsv&quot;</span>,</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Brilliant, the adjusted P-values are still significant, in fact they did not even change.
This shows us that the different media grouping are significantly different.
This makes sense from our taxonomy, alpha diversity, and beta diversity analysis.
However, it is hard to see the separation of the CVP and KBC samples in the MDS plot.</p>
</div>
</div>
<div id="beta-facet-zoom" class="section level2" number="16.5">
<h2><span class="header-section-number">16.5</span> Beta: Facet zoom</h2>
<p><img src="figures/telescope.png" width="15%" style="display: block; margin: auto;" /></p>
<p>The <a href="https://ggforce.data-imaginist.com/index.html"><code>ggforce</code> library</a> is a very powerful library to edit <code>ggplot2</code> objects.
One of its functions, <code>ggforce::facet_zoom()</code>, allows us to zoom into a specific part of a plot whilst also showing the main plot.
We will use this to zoom into the CVP and KBC samples within the weighted unifrac MDS plot.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Ordinate data</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>ord.mds.wunifrac <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">ordinate</span>(pseq_rarefy, <span class="at">method =</span> <span class="st">&quot;MDS&quot;</span>, <span class="at">distance =</span> <span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot ordination</span></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>nmds.wunifrac <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_ordination</span>(pseq_rarefy, ord.mds.wunifrac,</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">color =</span> <span class="st">&quot;site&quot;</span>, <span class="at">shape =</span> <span class="st">&quot;media&quot;</span>) <span class="sc">+</span></span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Zoom into CVP and KBC samples specified by the metadata category</span></span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Zoom.size = 1 so the zoom section is the same size as the main plot</span></span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Zoom.size = 2 would mean the zoom plot is 2 times the size of the original</span></span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>                  ggforce<span class="sc">::</span><span class="fu">facet_zoom</span>( <span class="at">xy =</span> media <span class="sc">==</span> <span class="fu">c</span>(<span class="st">&quot;CVP&quot;</span>,<span class="st">&quot;KBC&quot;</span>), <span class="at">zoom.size =</span> <span class="dv">1</span>)</span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Beta_diversity_mds_wunifrac_media_site_zoom.png&quot;</span>, <span class="at">plot =</span> nmds.wunifrac,</span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb101-13"><a href="#cb101-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb101-14"><a href="#cb101-14" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Beta_diversity_mds_wunifrac_media_site_zoom.png&quot;</span>)</span></code></pre></div>
<p>With the zoomed section we can more clearly see there is a difference between the CVP and KBC samples.
We can also see that the samples seem to slightly cluster by the site, especially for the KBC samples.</p>
<p>There are many other things you can do with <code>ggplot2</code> objects, too many to teach in this workshop.
How you want your plots will be highly dependant on your own data.
There are various resources in the APPEDNDIX to look at.</p>
</div>
<div id="beta-subset-samples" class="section level2" number="16.6">
<h2><span class="header-section-number">16.6</span> Beta: subset samples</h2>
<p><img src="figures/R_subset.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Another plot we can quickly make is to subset the <code>phyloseq</code> object to remove the ENV samples.
The ENV samples are causing the biggest difference between the samples.
To see the differences between the 3 media it will be good to remove it and produce an ordination plot.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Subset phyloseq to remove ENV samples</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>pseq_rarefy_no_env <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq_rarefy, media <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;CVP&quot;</span>,<span class="st">&quot;KBC&quot;</span>,<span class="st">&quot;TSA&quot;</span>))</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Ordinate data</span></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>ord.mds.wunifrac <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">ordinate</span>(pseq_rarefy_no_env, <span class="at">method =</span> <span class="st">&quot;MDS&quot;</span>, <span class="at">distance =</span> <span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot ordination</span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>nmds.wunifrac <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">plot_ordination</span>(pseq_rarefy_no_env, ord.mds.wunifrac,</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">color =</span> <span class="st">&quot;site&quot;</span>, <span class="at">shape =</span> <span class="st">&quot;media&quot;</span>) <span class="sc">+</span></span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>                  ggforce<span class="sc">::</span><span class="fu">facet_zoom</span>( <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.055</span>,<span class="sc">-</span><span class="fl">0.03</span>), <span class="at">zoom.size =</span> <span class="dv">1</span>)</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Beta_diversity_mds_wunifrac_media_site_no_env.png&quot;</span>, <span class="at">plot =</span> nmds.wunifrac,</span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">150</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb102-13"><a href="#cb102-13" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./Beta_diversity_mds_wunifrac_media_site_no_env.png&quot;</span>)</span></code></pre></div>
<p>This has changed our plot in a few ways.
The biggest difference is that our 1st axis explains 94.7% of variation, a very large amount.
This axis separates the media groups with TSA being the most different.</p>
<p>We can clearly see separation of the sites within the media groups.
There are a couple outliers within the CVP samples.</p>
<p>Thankfully our pairwise PERMANOVA results are still valid as they use the distances which did not change compared to the ordination values which did change.</p>
</div>
<div id="permanova-tasks" class="section level2" number="16.7">
<h2><span class="header-section-number">16.7</span> PERMANOVA tasks</h2>
<p><img src="figures/kettle_bell.png" width="15%" style="display: block; margin: auto;" /></p>
<p>For this task carry out pairwise PERMANOVA for the site metadata category.
I highly encourage you to copy, paste, and edit previous code you have created.</p>
<div class="webex-solution">
<button>
Site PERMANOVA solution
</button>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Pairwise comparison loop</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Get combinations of unique site values</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>cbn <span class="ot">&lt;-</span> <span class="fu">combn</span>(<span class="at">x =</span> <span class="fu">unique</span>(metadf<span class="sc">$</span>site), <span class="at">m =</span> <span class="dv">2</span>)</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Create empty final data frame with 4 columns</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a><span class="co"># and a number of rows equal to the the number of combinations</span></span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">matrix</span>(<span class="at">data =</span> <span class="cn">NA</span>, <span class="at">nrow =</span> <span class="fu">ncol</span>(cbn), <span class="at">ncol =</span> <span class="dv">4</span>))</span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Add column names</span></span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pairwise_permanova_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;1&quot;</span>,<span class="st">&quot;2&quot;</span>,<span class="st">&quot;p&quot;</span>,<span class="st">&quot;p.adj&quot;</span>)</span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Loop trhough the combinations</span></span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(cbn)){</span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Subset to site phyloseq</span></span>
<span id="cb103-13"><a href="#cb103-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Ensure to change metadata group if using different grouping (site)</span></span>
<span id="cb103-14"><a href="#cb103-14" aria-hidden="true" tabindex="-1"></a>  pseq_rarefy_subset <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq_rarefy, site <span class="sc">%in%</span> cbn[,i])</span>
<span id="cb103-15"><a href="#cb103-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Metadata data frame</span></span>
<span id="cb103-16"><a href="#cb103-16" aria-hidden="true" tabindex="-1"></a>  metadf_subset <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy_subset))</span>
<span id="cb103-17"><a href="#cb103-17" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Weighted unifrac distance calculation</span></span>
<span id="cb103-18"><a href="#cb103-18" aria-hidden="true" tabindex="-1"></a>  wunifrac_dist_mat <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(pseq_rarefy_subset, <span class="at">method=</span><span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb103-19"><a href="#cb103-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#PERMANOVA/ADONIS of site</span></span>
<span id="cb103-20"><a href="#cb103-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Ensure to change metadata group if using different grouping (site)</span></span>
<span id="cb103-21"><a href="#cb103-21" aria-hidden="true" tabindex="-1"></a>  wunifrac_pairwise_adonis <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(wunifrac_dist_mat <span class="sc">~</span> site, <span class="at">data =</span> metadf_subset, <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>)</span>
<span id="cb103-22"><a href="#cb103-22" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Add the gorup names and p-value to the main data frame</span></span>
<span id="cb103-23"><a href="#cb103-23" aria-hidden="true" tabindex="-1"></a>  pairwise_permanova_df[i,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> cbn[,i]</span>
<span id="cb103-24"><a href="#cb103-24" aria-hidden="true" tabindex="-1"></a>  pairwise_permanova_df[i,<span class="dv">3</span>] <span class="ot">&lt;-</span> wunifrac_pairwise_adonis[<span class="dv">1</span>,<span class="st">&quot;Pr(&gt;F)&quot;</span>]</span>
<span id="cb103-25"><a href="#cb103-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb103-26"><a href="#cb103-26" aria-hidden="true" tabindex="-1"></a><span class="co">#Add adjusted P-values</span></span>
<span id="cb103-27"><a href="#cb103-27" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df<span class="sc">$</span>p.adj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pairwise_permanova_df<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)</span>
<span id="cb103-28"><a href="#cb103-28" aria-hidden="true" tabindex="-1"></a><span class="co">#Write to a file</span></span>
<span id="cb103-29"><a href="#cb103-29" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> pairwise_permanova_df, <span class="at">file =</span> <span class="st">&quot;pairwise_permanova_site_wunifrac.tsv&quot;</span>,</span>
<span id="cb103-30"><a href="#cb103-30" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb103-31"><a href="#cb103-31" aria-hidden="true" tabindex="-1"></a><span class="co">#View data frame</span></span>
<span id="cb103-32"><a href="#cb103-32" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df</span></code></pre></div>
</div>
<p>Unfortunately, we cannot detect any significant differences between the sites.
Can we detect differences between the different combinations of sites and media?
Use the <code>site.media</code> metadata category to carry out pairwise PERMANOVA tests.</p>
<div class="webex-solution">
<button>
Site and media PERMANOVA solution
</button>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Pairwise comparison loop</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Get combinations of unique site &amp; media values</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>cbn <span class="ot">&lt;-</span> <span class="fu">combn</span>(<span class="at">x =</span> <span class="fu">unique</span>(metadf<span class="sc">$</span>site.media), <span class="at">m =</span> <span class="dv">2</span>)</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Create empty final data frame with 4 columns</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a><span class="co"># and a number of rows equal to the the number of combinations</span></span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">matrix</span>(<span class="at">data =</span> <span class="cn">NA</span>, <span class="at">nrow =</span> <span class="fu">ncol</span>(cbn), <span class="at">ncol =</span> <span class="dv">4</span>))</span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Add column names</span></span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pairwise_permanova_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;1&quot;</span>,<span class="st">&quot;2&quot;</span>,<span class="st">&quot;p&quot;</span>,<span class="st">&quot;p.adj&quot;</span>)</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Loop trhough the combinations</span></span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(cbn)){</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Subset to site &amp; media phyloseq</span></span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Ensure to change metadata group if using different grouping (site &amp; media)</span></span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>  pseq_rarefy_subset <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq_rarefy, site.media <span class="sc">%in%</span> cbn[,i])</span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Metadata data frame</span></span>
<span id="cb104-16"><a href="#cb104-16" aria-hidden="true" tabindex="-1"></a>  metadf_subset <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq_rarefy_subset))</span>
<span id="cb104-17"><a href="#cb104-17" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Weighted unifrac distance calculation</span></span>
<span id="cb104-18"><a href="#cb104-18" aria-hidden="true" tabindex="-1"></a>  wunifrac_dist_mat <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(pseq_rarefy_subset, <span class="at">method=</span><span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb104-19"><a href="#cb104-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#PERMANOVA/ADONIS of site &amp; media</span></span>
<span id="cb104-20"><a href="#cb104-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Ensure to change metadata group if using different grouping (site &amp; media)</span></span>
<span id="cb104-21"><a href="#cb104-21" aria-hidden="true" tabindex="-1"></a>  wunifrac_pairwise_adonis <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(wunifrac_dist_mat <span class="sc">~</span> site.media, <span class="at">data =</span> metadf_subset, <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>)</span>
<span id="cb104-22"><a href="#cb104-22" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Add the gorup names and p-value to the main data frame</span></span>
<span id="cb104-23"><a href="#cb104-23" aria-hidden="true" tabindex="-1"></a>  pairwise_permanova_df[i,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> cbn[,i]</span>
<span id="cb104-24"><a href="#cb104-24" aria-hidden="true" tabindex="-1"></a>  pairwise_permanova_df[i,<span class="dv">3</span>] <span class="ot">&lt;-</span> wunifrac_pairwise_adonis[<span class="dv">1</span>,<span class="st">&quot;Pr(&gt;F)&quot;</span>]</span>
<span id="cb104-25"><a href="#cb104-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb104-26"><a href="#cb104-26" aria-hidden="true" tabindex="-1"></a><span class="co">#Add adjusted P-values</span></span>
<span id="cb104-27"><a href="#cb104-27" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df<span class="sc">$</span>p.adj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(pairwise_permanova_df<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)</span>
<span id="cb104-28"><a href="#cb104-28" aria-hidden="true" tabindex="-1"></a><span class="co">#Write to a file</span></span>
<span id="cb104-29"><a href="#cb104-29" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> pairwise_permanova_df, <span class="at">file =</span> <span class="st">&quot;pairwise_permanova_site_media_wunifrac.tsv&quot;</span>,</span>
<span id="cb104-30"><a href="#cb104-30" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb104-31"><a href="#cb104-31" aria-hidden="true" tabindex="-1"></a><span class="co">#View data frame</span></span>
<span id="cb104-32"><a href="#cb104-32" aria-hidden="true" tabindex="-1"></a>pairwise_permanova_df</span></code></pre></div>
</div>
<p>We are getting very high adjusted P-values for these comparisons.
There is most likely not enough power for this analysis as there are only 3 replicates for each group.</p>
</div>
<div id="beta-summary" class="section level2" number="16.8">
<h2><span class="header-section-number">16.8</span> Beta: Summary</h2>
<p><img src="figures/sum_orange.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have produced Beta diversity plots and the statistics for weighted unifrac distances.
There are many other things you could do such as:</p>
<ul>
<li>Carry out beta diversity with taxa aggregated <code>phyloseq</code> objects.</li>
<li>Use different distance measures (<code>phyloseq::distanceMethodList</code>).
<ul>
<li>We had a phylogenetic tree in our <code>phyloseq</code> object which allows us to use <code>UniFrac</code> measures (“unifrac” and “wunifrac”).</li>
<li>If you do not have this you can use non-phylogenetic aware <code>betadiver</code> distances such as “jaccard” or “bray”.</li>
</ul></li>
<li>Use different different ordination methods (<code>?phyloseq:ordinate</code>).
<ul>
<li>Different methods will produce different” percentage of variation explained” values. If your plot has low values try another method to see if you can get higher values.</li>
</ul></li>
</ul>
<p>With our analysis we found out:</p>
<ul>
<li>The samples significantly cluster by media with ENV being the most different followed by TSA.</li>
<li>It appears that the samples cluster by site within each media group, however this is not backed up by statistics.</li>
<li>There appears to be different levels of variation within the media groups (tightness/looseness of clusters).
<ul>
<li>The CVP samples seem to have the lowest levels of variation (tight clusters), whilst the ENV samples seem to have the highest level of variation.</li>
<li>This could be explored further by <a href="https://microbiome.github.io/tutorials/PERMANOVA.html">“Checking the homogeneity condition”</a>.</li>
</ul></li>
</ul>
<p>We know there is a difference between the media groupings with the different medias causing different selections of certain taxa compared to the ENV samples.
Let’s try to figure out which taxa are affected in which ways.</p>

</div>
</div>
<div id="DA_chap" class="section level1" number="17">
<h1><span class="header-section-number">Chapter 17</span> Differential abundance (DA) analysis</h1>
<p><img src="figures/difference.png" width="20%" style="display: block; margin: auto;" /></p>
<p>We are going to carry out differential abundance analysis with <a href="https://www.nature.com/articles/s41467-020-17041-7">ANCOM-BC (Analysis of compositions of microbiomes with bias correction)</a>.
This is an improvement of the original ANCOM method but with added bias correction.
This bias correction aims to account for the bias caused by the difference in sampling/depth across the samples.</p>
<p>Differential abundance analysis aims to detect features (ASVs, taxa, etc.) that have different abundances between groups.
This can be used to detect organisms in high abundance in diseased states versus healthy states.
This helps determine what are the disease causing organisms.
However, it may also detect organisms that are able to survive in the environment rather than those that cause the difference in environment.</p>
<p>This will be shown in our data.
Organisms in higher abundance in the media samples compared to the environmental samples are most likely caused by selection pressures the media introduces.
To keep our data manageable in a workshop we will look at the phyla abundances.
If a phyla is detected to have higher abundance in a media compared to the environmental sample, we can say the phyla is a biomarker of the media compared to the environmental sample.
Additionally, we will be able to detect phyla that have been depleted in the media samples versus the environmental sample.</p>
<div id="da-setup" class="section level2" number="17.1">
<h2><span class="header-section-number">17.1</span> DA: setup</h2>
<p><img src="figures/packages_1.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We will use a new notebook called <strong>“7-differential_abundance.ipynb”</strong>.
Add the following to the top of this new notebook to load the required libraries/packages and data.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Packages</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ANCOMBC&quot;</span>)</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;DT&quot;</span>)</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;dplyr&quot;</span>)</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;tidyr&quot;</span>)</span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb105-10"><a href="#cb105-10" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
<p>We will be using the package <a href="https://github.com/FrederickHuangLin/ANCOM-BC-Code-Archive"><code>ANCOMBC</code></a> to carry out our DA analysis.
Instead of using our rarefied data we will use our original abundance data.
This is because ANCOM-BC uses absolute abundances and corrects for differences between samples.
This is a statistically heavy method and I recommend you read the <a href="https://www.nature.com/articles/s41467-020-17041-7">ANCOM-BC paper</a> if you are interested in how it works.
As bioinformaticians we are more interested in the output rather than the method.</p>
</div>
<div id="da-preprocess" class="section level2" number="17.2">
<h2><span class="header-section-number">17.2</span> DA: Preprocess</h2>
<p><img src="figures/water_surface.png" width="15%" style="display: block; margin: auto;" /></p>
<p>When carrying out ANCOM analysis you need to specify the reference group for the groupings you use.
The reference group will be the 1st level of the factor.
Therefore we will check our level orders for our media and site metadata categories.</p>
<div class="webex-solution">
<button>
What are factors?
</button>
<p>For an explanation on the factors class please see the following section in the R primer omics book:
<a href="https://neof-workshops.github.io/R_j4c0xh/14-Plots_scatterplot_and_box_plots.html#factors">Factors</a></p>
</div>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Check media and site levels</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)<span class="sc">$</span>media)</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)<span class="sc">$</span>site)</span></code></pre></div>
<p>The levels are not in the order we would like.
We’ll therefore change the order with the <code>factor()</code> function and <code>levels =</code> option.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Ensure ENV is our reference level for media</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(pseq)<span class="sc">$</span>media <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">sample_data</span>(pseq)<span class="sc">$</span>media,</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;ENV&quot;</span>, <span class="st">&quot;CVP&quot;</span>, <span class="st">&quot;KBC&quot;</span>, <span class="st">&quot;TSA&quot;</span>))</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)<span class="sc">$</span>media)</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Ensure UD is our reference level for site</span></span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(pseq)<span class="sc">$</span>site <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">sample_data</span>(pseq)<span class="sc">$</span>site,</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;UD&quot;</span>, <span class="st">&quot;MD&quot;</span>, <span class="st">&quot;LD&quot;</span>))</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)<span class="sc">$</span>site)</span></code></pre></div>
</div>
<div id="da-ancom-bc-analysis" class="section level2" number="17.3">
<h2><span class="header-section-number">17.3</span> DA: ANCOM-BC analysis</h2>
<p><img src="figures/play_blue.png" width="10%" style="display: block; margin: auto;" /></p>
<p>Now that we have our data the way we would like it we can carry out the ANCOM analysis.
We’ll use the function <code>ANCOMBC::ancombc2()</code> for our ANCOM-BC analysis.</p>
<p><strong>Note:</strong> This command will take a few minutes to run.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Run ANCOMBC</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>ancom_class_out <span class="ot">&lt;-</span> ANCOMBC<span class="sc">::</span><span class="fu">ancombc2</span>(<span class="at">data =</span> pseq, <span class="at">fix_formula =</span> <span class="st">&quot;media + site&quot;</span>,</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">p_adj_method =</span> <span class="st">&quot;holm&quot;</span>,</span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">tax_level =</span> <span class="st">&quot;Class&quot;</span>)</span></code></pre></div>
<p>The resulting object is very large so it can be a good idea to save it.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Save results</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(ancom_class_out, <span class="at">file =</span> <span class="st">&quot;ANCOMBC2_class_analysis.RData&quot;</span>)</span></code></pre></div>
<div id="ancombc2-options" class="section level3" number="17.3.1">
<h3><span class="header-section-number">17.3.1</span> ANCOMBC2 options</h3>
<p><img src="figures/parameter_blue.png" width="10%" style="display: block; margin: auto;" /></p>
<p>The options we use in the function are:</p>
<ul>
<li><code>data =</code>: The data to be analysed. This can be provided in various types of objects including the <code>phyloseq</code> object.</li>
<li><code>fix_formula =</code>: The groupings within your data. It is importnat to add all groupings that may have an influence on the microbial abundances.</li>
<li><code>p_adj_method =</code>: The p adjustment value to use.</li>
<li><code>tax_level =</code>: Te taxonomic level of interest.</li>
</ul>
<p>There are many more options that have been left as the default.
Please see the package <a href="https://bioc.ism.ac.jp/packages/devel/bioc/manuals/ANCOMBC/man/ANCOMBC.pdf">PDF</a> for more options and explanations.</p>
</div>
</div>
<div id="da-ancom-bc-results" class="section level2" number="17.4">
<h2><span class="header-section-number">17.4</span> DA: ANCOM-BC results</h2>
<p><img src="figures/output_file_blue.png" width="10%" style="display: block; margin: auto;" /></p>
<p>There is a lot of information in <code>ancom_class_out</code>.
Therefore, lets extract the results we want to look at and then remove the large object.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Output of analysis compared to reference groups</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>res_ancom_class <span class="ot">&lt;-</span> ancom_class_out<span class="sc">$</span>res</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove large object</span></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(ancom_class_out)</span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Check top rows of results</span></span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_ancom_class)</span></code></pre></div>
<p>These reuslts only contain taxon where a significant result was found in at least one comparison.
This is based on the adjusted p-values (&lt;0.05).
There are a lot of columns (37) in our result data frame.
Apart from taxon they are split by the comparison group.
These contain the following information:</p>
<ul>
<li>taxon: The taxa groups found to be a biomarker in at least one comparison.</li>
<li>lfc_*: This contains the log (natural log) fold changes with respect to the reference group.
<ul>
<li>For example The values for lfc_mediaCVP would equal LFC(CVP - ENV).</li>
<li>A positive value means the abundances are higher in the comparison group (CVP) and lower in the ref (ENV).</li>
<li>A negative value means the abundances are lower in the comparison group (CVP) and higher in the ref (ENV).</li>
<li>The intercept value is the grand mean and is likely not of interest.</li>
</ul></li>
<li>se_*: The standard error for the LFC values.</li>
<li>W_*: The W statistic. The higher this value the more significant the feature is differentially abundant between groups.</li>
<li>p_*: The P-value.</li>
<li>q_*: The Q-value. This is the adjusted P-value.</li>
<li>diff*: This indicates whether there is a significant difference in the abundances of the comparison and reference group. In other words, is the Q-value &lt; 0.05.</li>
</ul>
<p>It is quite a large table so let’s save it as a tsv (tab separated value) file.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Save as file</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> res_ancom_class, <span class="at">file =</span> <span class="st">&quot;ANCOMBC2_class_results.tsv&quot;</span>,</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>This is good to open in excel but we don’t have something like that on our cluster.
Therefore, we will use the package <code>DT</code> to create an interactive table with the function <code>datatable()</code>.
We can then save this object as an HTML with <code>htmlwidgets::saveWidget()</code>.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create data table</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> DT<span class="sc">::</span><span class="fu">datatable</span>(res_ancom_class)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Save data table as html</span></span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>htmlwidgets<span class="sc">::</span><span class="fu">saveWidget</span>(m, <span class="st">&quot;ANCOM_class_results.html&quot;</span>, <span class="at">selfcontained =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Go to the jupyter-notebook file explorer tab and click on the html file to open it in a new tab.
This gives a relatively nice way to view all the results.</p>
<p>One very clear pattern is that there are no biomarkers found for MD and LD.
I believe this is because we don’t have enough replicates for the sites within the media samples and the differences between the medias are so large.
However, there is a good amount of biomarkers for the three different medias compared to ENV.</p>
</div>
<div id="da-create-signifcant-lfc-long-data-frame" class="section level2" number="17.5">
<h2><span class="header-section-number">17.5</span> DA: Create signifcant LFC long data frame</h2>
<p><img src="figures/portrait.png" width="10%" style="display: block; margin: auto;" /></p>
<p>The results table is good but let’s make a heatmap showing the LFCs (Log fold changes) of the significant biomarkers. First, we need to do some preprecessing.</p>
<p>Remove columns unrelated to the media comparisons.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract the taxon column and the media columns from the results</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="co">#get positions of columns which contain &quot;media&quot;</span></span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>media_colnames_pos <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;media&quot;</span>, <span class="fu">colnames</span>(res_ancom_class))</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Taxon column and media column names</span></span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>media_colnames <span class="ot">&lt;-</span> <span class="fu">colnames</span>(res_ancom_class)[<span class="fu">c</span>(<span class="dv">1</span>,media_colnames_pos)]</span>
<span id="cb113-6"><a href="#cb113-6" aria-hidden="true" tabindex="-1"></a>media_colnames</span>
<span id="cb113-7"><a href="#cb113-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Create new df with only taxon and media columns</span></span>
<span id="cb113-8"><a href="#cb113-8" aria-hidden="true" tabindex="-1"></a>res_ancom_class_media <span class="ot">&lt;-</span> res_ancom_class[,media_colnames]</span>
<span id="cb113-9"><a href="#cb113-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_ancom_class_media)</span></code></pre></div>
<p>Because we had comparisons with more that 1 grouping (media &amp; site) we might have rows with no significnat results.
We will therefore remove these rows.</p>
<p><strong>Note:</strong> The next 2 code sections use the packages <code>dplyr</code> &amp; <code>tidyr</code> along with piping <code>%&gt;%</code>.
This is used as it would be very difficult to do these steps in base R and they are the code provided in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html"><code>ANCOMBC2</code> tutorial</a>.
If you would like to learn more about these packages please see the <a href="https://r4ds.had.co.nz/">R for Data Science book</a>.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove any taxon where there are no TRUE values</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="co">#I.e. no significant difference found for the media samples</span></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of biomarkers</span></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_ancom_class_media)</span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract rows that contain at least one TRUE value</span></span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a>res_ancom_class_media_sig <span class="ot">&lt;-</span> res_ancom_class_media <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter_all</span>(</span>
<span id="cb114-7"><a href="#cb114-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">any_vars</span>(. <span class="sc">%in%</span> <span class="cn">TRUE</span>)</span>
<span id="cb114-8"><a href="#cb114-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb114-9"><a href="#cb114-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Number of rows and the top of the new df</span></span>
<span id="cb114-10"><a href="#cb114-10" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_ancom_class_media_sig)</span>
<span id="cb114-11"><a href="#cb114-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_ancom_class_media_sig)</span></code></pre></div>
<p>Next we need to do a variety of preprocessing steps.
This will create an edited long data frame.
Long data frames are needed for <code>ggplot2()</code> when not using a <code>phyloseq</code> object.
Instead of having the LFC (log fold change) values over many columns there will be only one column for the values.
The long data frame will contain the following columns:</p>
<ul>
<li><strong>taxon</strong>: The biomarker name.</li>
<li><strong>group</strong>: The comparison group.</li>
<li><strong>value</strong>: The LFC value.</li>
</ul>
<p>Please see the code and its annotation below.
If you do not fully understand the code that is fine.
As long as you can edit it to work with your own future code.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create edited long df to plot significant log fold changes on heatmap</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>res_ancom_class_media_sig_lfc_fig <span class="ot">&lt;-</span> res_ancom_class_media_sig <span class="sc">%&gt;%</span></span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Change log change values to 0 if diff is false</span></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#test: Is diff TRUE?</span></span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#yes: Set LFC to original value (e.g. lfc_mediaCVP)</span></span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">#no: Set LFC to 0</span></span>
<span id="cb115-7"><a href="#cb115-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">lfc_mediaCVP =</span> <span class="fu">ifelse</span>(<span class="at">test =</span> diff_mediaCVP <span class="sc">==</span> <span class="cn">TRUE</span>,</span>
<span id="cb115-8"><a href="#cb115-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">yes =</span> lfc_mediaCVP, <span class="at">no =</span> <span class="dv">0</span>),</span>
<span id="cb115-9"><a href="#cb115-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">lfc_mediaKBC =</span> <span class="fu">ifelse</span>(<span class="at">test =</span> lfc_mediaKBC <span class="sc">==</span> <span class="cn">TRUE</span>,</span>
<span id="cb115-10"><a href="#cb115-10" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">yes =</span> lfc_mediaKBC, <span class="at">no =</span> <span class="dv">0</span>),</span>
<span id="cb115-11"><a href="#cb115-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">lfc_mediaTSA =</span> <span class="fu">ifelse</span>(<span class="at">test =</span> lfc_mediaTSA <span class="sc">==</span> <span class="cn">TRUE</span>,</span>
<span id="cb115-12"><a href="#cb115-12" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">yes =</span> lfc_mediaTSA, <span class="at">no =</span> <span class="dv">0</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb115-13"><a href="#cb115-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Only retain taxon and log fold changes columns</span></span>
<span id="cb115-14"><a href="#cb115-14" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Additionally, change column names and round LFC values to 2 decimal places</span></span>
<span id="cb115-15"><a href="#cb115-15" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">transmute</span>(taxon,</span>
<span id="cb115-16"><a href="#cb115-16" aria-hidden="true" tabindex="-1"></a>                   <span class="at">CVP =</span> <span class="fu">round</span>(lfc_mediaCVP,<span class="dv">2</span>),</span>
<span id="cb115-17"><a href="#cb115-17" aria-hidden="true" tabindex="-1"></a>                   <span class="at">KBC =</span> <span class="fu">round</span>(lfc_mediaKBC,<span class="dv">2</span>),</span>
<span id="cb115-18"><a href="#cb115-18" aria-hidden="true" tabindex="-1"></a>                   <span class="at">TSA =</span> <span class="fu">round</span>(lfc_mediaTSA,<span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb115-19"><a href="#cb115-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Convert to long data frame</span></span>
<span id="cb115-20"><a href="#cb115-20" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> CVP<span class="sc">:</span>TSA,</span>
<span id="cb115-21"><a href="#cb115-21" aria-hidden="true" tabindex="-1"></a>                      <span class="at">names_to =</span> <span class="st">&quot;group&quot;</span>,</span>
<span id="cb115-22"><a href="#cb115-22" aria-hidden="true" tabindex="-1"></a>                      <span class="at">values_to =</span> <span class="st">&quot;value&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb115-23"><a href="#cb115-23" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Arrange order of rows based on taxons</span></span>
<span id="cb115-24"><a href="#cb115-24" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(taxon)</span>
<span id="cb115-25"><a href="#cb115-25" aria-hidden="true" tabindex="-1"></a><span class="co">#Check new long df</span></span>
<span id="cb115-26"><a href="#cb115-26" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_ancom_class_media_sig_lfc_fig)</span></code></pre></div>
</div>
<div id="da-biomarker-lfc-heatmap" class="section level2" number="17.6">
<h2><span class="header-section-number">17.6</span> DA: Biomarker LFC heatmap</h2>
<p><img src="figures/tiles.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We now have a long data frame with the significant LFC values.
We can use this to create a heatmap.</p>
<p>There are some specific functions and packages for R that can produce heatmaps.
These can include dendrograms on the y and x axes to show clustering.
We are not interested in having all the bells and whistles so we will use the <code>ggplot2</code> layer <code>geom_tile()</code>.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co">#LFC heatmap</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>heatmap_media <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_ancom_class_media_sig_lfc_fig,</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> taxon, <span class="at">fill =</span> value)) <span class="sc">+</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Produce ggplot as tile/heatmap style plot</span></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">::</span><span class="fu">geom_tile</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Add the LFC values as text in the cells</span></span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">::</span><span class="fu">geom_text</span>(<span class="fu">aes</span>(group, taxon, <span class="at">label =</span> value),</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Remove the x and y labels (NULL) and add a title</span></span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">:</span>labs <span class="ot">=</span> <span class="cn">NULL</span>, y <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>                    title <span class="ot">=</span> <span class="st">&quot;Log fold changes compared to ENV samples&quot;</span><span class="er">)</span></span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./ANCOMBC_class_media_ref_ENV.png&quot;</span>, <span class="at">plot =</span> heatmap_media,</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">125</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./ANCOMBC_class_media_ref_ENV.png&quot;</span>)</span></code></pre></div>
<p>It is a good start but the colour gradient is not very useful.
We are going to use the layer <code>ggplot2::scale_fill_gradient2()</code> to customise this.
First, we need to decide on the minimum and maximum values.
Let’s generate these with some code based on the minimum and maximum LFC values.
We will use 0 as the midpoint so we can easily differentiate between higher and lower abundance biomarkers.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Get min and max log fold changes</span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>lo <span class="ot">&lt;-</span> <span class="fu">floor</span>(<span class="fu">min</span>(res_ancom_class_media_sig_lfc_fig<span class="sc">$</span>value))</span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>up <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fu">max</span>(res_ancom_class_media_sig_lfc_fig<span class="sc">$</span>value))</span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a>mid <span class="ot">&lt;-</span> <span class="dv">0</span></span></code></pre></div>
<p>With these values we can now make our upgraded heatmap.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co">#LFC heatmap</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>heatmap_media <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_ancom_class_media_sig_lfc_fig,</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> taxon, <span class="at">fill =</span> value)) <span class="sc">+</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Produce ggplot as tile/heatmap style plot</span></span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">::</span><span class="fu">geom_tile</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Customise colour gradient</span></span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">::</span><span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">high =</span> <span class="st">&quot;red&quot;</span>, <span class="at">mid =</span> <span class="st">&quot;white&quot;</span>,</span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">na.value =</span> <span class="st">&quot;white&quot;</span>,</span>
<span id="cb118-9"><a href="#cb118-9" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">midpoint =</span> mid, <span class="at">limit =</span> <span class="fu">c</span>(lo,up)) <span class="sc">+</span></span>
<span id="cb118-10"><a href="#cb118-10" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Add the LFC values as text in the cells</span></span>
<span id="cb118-11"><a href="#cb118-11" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">::</span><span class="fu">geom_text</span>(<span class="fu">aes</span>(group, taxon, <span class="at">label =</span> value),</span>
<span id="cb118-12"><a href="#cb118-12" aria-hidden="true" tabindex="-1"></a>                      <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb118-13"><a href="#cb118-13" aria-hidden="true" tabindex="-1"></a>                  <span class="co">#Remove the x and y labels (NULL) and add a title</span></span>
<span id="cb118-14"><a href="#cb118-14" aria-hidden="true" tabindex="-1"></a>                  ggplot2<span class="sc">:</span>labs <span class="ot">=</span> <span class="cn">NULL</span>, y <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb118-15"><a href="#cb118-15" aria-hidden="true" tabindex="-1"></a>                    title <span class="ot">=</span> <span class="st">&quot;Log fold changes compared to ENV samples&quot;</span><span class="er">)</span></span>
<span id="cb118-16"><a href="#cb118-16" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb118-17"><a href="#cb118-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./ANCOMBC_class_media_ref_ENV.png&quot;</span>, <span class="at">plot =</span> heatmap_media,</span>
<span id="cb118-18"><a href="#cb118-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">125</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb118-19"><a href="#cb118-19" aria-hidden="true" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb118-20"><a href="#cb118-20" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./ANCOMBC_class_media_ref_ENV.png&quot;</span>)</span></code></pre></div>
<p>Great, now we can the biomarker with higher abundances in the ENV sample (blue) and higher abundances in the comparison media (red).
Any 0 values are non-significant results we turned to 0 when we created the long data frame.</p>
<ul>
<li>Are there more biomarkers in higher abundance in ENV rather than the comparison media (i.e blue/negative results)?
<div id="radio_ROWXEWKYSH" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_ROWXEWKYSH" value="answer"></input> <span><strong>Yes</strong></span></label><label><input type="radio" autocomplete="off" name="radio_ROWXEWKYSH" value=""></input> <span><strong>No</strong></span></label>
</div></li>
<li>Which biomarker has the highest positive values?
<div id="radio_JJIRUEPOBA" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_JJIRUEPOBA" value=""></input> <span><strong>Phylum:Verrucomicrobia</strong></span></label><label><input type="radio" autocomplete="off" name="radio_JJIRUEPOBA" value="answer"></input> <span><strong>Class:Bacilli</strong></span></label><label><input type="radio" autocomplete="off" name="radio_JJIRUEPOBA" value=""></input> <span><strong>Class:Alphaproteobacteria</strong></span></label>
</div></li>
<li>Which biomarker has the lowest value?
<div id="radio_XXPGCXWLFK" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_XXPGCXWLFK" value=""></input> <span><strong>Phylum:Verrucomicrobia</strong></span></label><label><input type="radio" autocomplete="off" name="radio_XXPGCXWLFK" value=""></input> <span><strong>Class:Bacilli</strong></span></label><label><input type="radio" autocomplete="off" name="radio_XXPGCXWLFK" value="answer"></input> <span><strong>Class:Alphaproteobacteria</strong></span></label>
</div></li>
</ul>
<p>From these results we can see that there are a lot of biomarkers with low abundances in the comparison media samples.
There are only 2 biomarkers with higher abundance in the media samples.</p>
<p>The LFC values between the different media are quite similar.
There are some biomarkers that are only siginifcnat in 1 or 2 of the media though.
It seems that CVP and KBC are more simialr to each other than to TSA.
This matches the clustering we found in the beta diversity.</p>
<p>This seems to indicate that the media samples are not a good representation of the actual environmental community.
They appear to select for a few bacterial classes and select against a lot of bacterial classes.</p>
</div>
<div id="da-matching-bar-plot" class="section level2" number="17.7">
<h2><span class="header-section-number">17.7</span> DA: matching bar plot</h2>
<p><img src="figures/bar_chart.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Using code from the <a href="#chaptaxaplots">taxa plots chapter</a> we’ll make a quick class based taxa plot.
We will use the relative abundances for this.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Quick class based relative abundance heatmap</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Convert abundance table to class relative abundance (compositional) table</span></span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>pseq_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Class phyloseq</span></span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>class_relabund_pseq <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_taxa</span>(pseq_relabund, <span class="st">&quot;Class&quot;</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Produce heatmap ggplot</span></span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>class_heatmap <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(class_relabund_pseq, <span class="at">plot.type =</span> <span class="st">&quot;heatmap&quot;</span>) <span class="sc">+</span></span>
<span id="cb119-8"><a href="#cb119-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Class&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Sample&quot;</span>) <span class="sc">+</span></span>
<span id="cb119-9"><a href="#cb119-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Class relative abundance heatmap&quot;</span>)</span>
<span id="cb119-10"><a href="#cb119-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Save ggplot object as png file</span></span>
<span id="cb119-11"><a href="#cb119-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./class_relabund_heatmap.png&quot;</span>, <span class="at">plot =</span> class_heatmap,</span>
<span id="cb119-12"><a href="#cb119-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">200</span>)</span>
<span id="cb119-13"><a href="#cb119-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb119-14"><a href="#cb119-14" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./class_relabund_heatmap.png&quot;</span>)</span></code></pre></div>
<p>With this plot we can clearly see that there are many classess with abundances in the ENV samples that have very little in the media samples.
Most the the relative abundances of the media samples consist of Gammaproteobacteria.
The TSA and KBC samples also have a high abundance of Bacilli.
Overall, this is a nice visual confirmation of our ANCOM results.</p>
</div>
<div id="da-summary" class="section level2" number="17.8">
<h2><span class="header-section-number">17.8</span> DA: summary</h2>
<p><img src="figures/sum_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have used ANCOMBC to detect biomarkers in the media samples compared to the ENV samples.
This showed most biomarkers were caused by depletion of many classes and selection of a few classes by the media.</p>
<p>You could carry out more analysis in your future research in various ways:</p>
<ul>
<li>Carry out biomarker detection at different taxonomy levels such as genus.
<ul>
<li>We used class in this tutorial so we would have enough groups for a comparison but not too many to increase speed of analysis and so there wasn’t too many results to look at.</li>
</ul></li>
<li>There are many other types of tests and considerations to keep in mind. Check out the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html">ANCOMBC2 tutorial</a>, it is very in depth.
<ul>
<li>There is also a handy FAQ on the <a href="https://github.com/FrederickHuangLin/ANCOMBC">ANCOMBC github page</a>.</li>
</ul></li>
</ul>
<p>That is the last of the analysis. In the final chapter we will review the steps we carried out and the answers to the various questions we have about the dataset.</p>

</div>
</div>



<div id="summary" class="section level1" number="18">
<h1><span class="header-section-number">Chapter 18</span> Summary</h1>
<p>You have learnt a lot during this course. This chapter will therefore summarise what you have done.</p>
<div id="dataset-1" class="section level2" number="18.1">
<h2><span class="header-section-number">18.1</span> Dataset</h2>
<p><img src="figures/freshwater_france.png" width="50%" style="display: block; margin: auto;" /></p>
<p>You were introduce to the 16S dataset in the intro.
This consisted of samples from 3 parts of the Durance river (UD, MD, &amp; LD).
This is across a gradient of anthropisation.
Unfortunately we were not able to glean to much information on the differences due to low power, only 3 replicates per ENV sample.</p>
<p>The more interesting metadata grouping was that of the culture.
We were interested if 3 different media (TSA, KBC, and CVP) had different profiled communities than the ENV samples.
They were found to be very different compared to the ENV samples.
Some differences between the media groupings were also found.
Overall, any differences between the sites were hard to determine due to the large differences between media.</p>
<p>To properly determine the differences across the anthropisation gradient it would be best to use a good amount of environmental samples.
It may also help to sample more sites along the gradient.</p>
<p>We discovered these findings through our analysis.</p>
</div>
<div id="data-preparation" class="section level2" number="18.2">
<h2><span class="header-section-number">18.2</span> Data preparation</h2>
<p><img src="figures/half_tablespoon.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Briefly we carried out the following steps to prepare our data:</p>
<ul>
<li><a href="#import_chap">Chapter 7</a>: Imported a <code>phyloseq</code> object with <code>qiime2R</code>.</li>
<li><a href="#sum_phyloseq_chap">Chapter 8</a>: Checked our data with summarisations.</li>
<li><a href="#id_#mindepthchap">Chapter 9</a>: Checked the minimum depth we should use and removed samples with lower depth.</li>
</ul>
</div>
<div id="taxonomy" class="section level2" number="18.3">
<h2><span class="header-section-number">18.3</span> Taxonomy</h2>
<p><img src="figures/taxa.png" width="20%" style="display: block; margin: auto;" /></p>
<p>We investigated the taxonomic composition of our samples.</p>
<ul>
<li><a href="#taxa_relabund_chap">Chapter 10</a>: Creation of taxonomic relative abundance tables.</li>
<li><a href="#chaptaxaplots">Chapter 11</a>: Produced heat maps and bar plots of taxa relative abundances
<ul>
<li>We also learnt how to aggregate rare taxa to produce fewer taxa groups for plotting.</li>
</ul></li>
<li><a href="#family_genus_chap">Chapter 12</a>: Produced family and genus based taxa plots.</li>
</ul>
<p>With our taxa plots we saw that the biggest difference was caused by the media.
The ENV samples appear to have the most diversity with many different taxa groups.
The media samples were mostly composed of Proteobacteria &amp; Firmicutes.</p>
<p>Due to the difference of media we cannot see any clear differences between the sites.</p>
</div>
<div id="diversity-analysis" class="section level2" number="18.4">
<h2><span class="header-section-number">18.4</span> Diversity analysis</h2>
<p><img src="figures/diversity.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Prior to alpha &amp; beta diversity analysis we carried out rarefaction to normalise the depths between samples in <a href="#rarefaction_chap">Chapter 14</a>.
This is a controversial technique, we carry it out due to the logic and the fact the QIIME2 developers recommend it.
Please make you own decision.</p>
<p>Two of the leading papers are:</p>
<ul>
<li><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">Waste Not, Want Not: Why Rarefying Microbiome Data Is Inadmissible (McMurdie &amp; Holmes 2014)</a>.</li>
<li><a href="https://academic.oup.com/bioinformatics/article/38/9/2389/6536959">To rarefy or not to rarefy: robustness and efficiency trade-offs of rarefying microbiome data (Hong et al, 2022)</a>.</li>
</ul>
<div id="alpha-beta" class="section level3" number="18.4.1">
<h3><span class="header-section-number">18.4.1</span> Alpha &amp; beta</h3>
<p><img src="figures/alpha.png" width="15%" style="display: block; margin: auto;" /></p>
<p>With our rarefied table we carried out diversity analysis.</p>
<ul>
<li><a href="#alpha_chap">Chapter 15</a>: Produced alpha violin plots and statistics to compare the diversity of groups.</li>
<li><a href="#beta_chap">Chapter 16</a>: Produced MDS plots based on pairwise distance metrics. Along with statistics this allowed us to determine how similar/dissimilar samples were to each other.</li>
</ul>
<p>The alpha diversity showed significant differences between the media and ENV samples.
Additionally, the alpha metrics between the 3 media groupings were also significantly different.
This was the case for observed ASVs and the Chao1 estimator which had much higher values in the ENV samples.</p>
<p>Although the ENV samples had higher Shannon scores (diversity), the values were also wide spread leading to lack of significance when comparing the ENV to CVP and TSA samples.</p>
<p>No difference was found when comparing the sites.</p>
<p>The beta diversity analysis found significant difference between all the media groupings.
Clustering of the media groups was also clearly visible in the MDS plots.</p>
<p>Interestingly the CVP and KBC samples had relatively tight cluster that were quite near each other.
The TSA and ENV samples had very loose cluster indicating large differences within the group.
There is also some presence of clustering by site within each media group which appears to cause the main differences within the groups.</p>
<p>Unfortunately we are not able to detect any differences between sites due to power and the biggest differences cause by media.</p>
</div>
<div id="differential-abundance-anlaysis" class="section level3" number="18.4.2">
<h3><span class="header-section-number">18.4.2</span> Differential abundance anlaysis</h3>
<p><img src="figures/difference.png" width="20%" style="display: block; margin: auto;" /></p>
<p>After diversity analysis we carried out differential abundance analysis in <a href="#DA_chap">chapter 17</a>.
This was carried out with ANCOM whilst taking into account both the media and site information.</p>
<p>We found many bacterial classes with lower abundances in the media samples compared to the ENV samples.
We only found 2 classes with higher abundance in the media samples compared to the ENV samples.
This indicates the media are selecting for only a few classes and many bacteria cannot survive on the media.
This biologically makes sense but also corroborates what we have seen in our other analyses.</p>
</div>
</div>
<div id="findings" class="section level2" number="18.5">
<h2><span class="header-section-number">18.5</span> Findings</h2>
<p><img src="figures/compass.png" width="20%" style="display: block; margin: auto;" /></p>
<p>In <a href="#sum_and_qs">chapter 2</a> we asked some questions about our dataset.
Let’s try to answer them.</p>
<div id="anthropisation-gradient" class="section level3" number="18.5.1">
<h3><span class="header-section-number">18.5.1</span> Anthropisation gradient</h3>
<p><img src="figures/farm.png" width="15%" style="display: block; margin: auto;" /></p>
<p>How does the bacterial communities change across the anthropisation gradient?</p>
<p>Due to the vast difference caused by the media we were not able to determine this.
You could try to remove all but the ENV samples to determine this.
However, you would not have much power with only 9 samples (3 sites by 3 replicates).</p>
</div>
<div id="replicate-differences" class="section level3" number="18.5.2">
<h3><span class="header-section-number">18.5.2</span> Replicate differences</h3>
<p><img src="figures/replicate_difference.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Is there a difference in the replicates of one site and media combination? I.e. do any of the media produce inconsistent profiles?</p>
<p>There are inconsistencies between the replicates of site and media combinations.
However, the biggest inconsistencies are within the TSA samples. followed by the ENV samples
This is shown by the loose clustering in the beta diversity of the these sample and the tight clustering for CVP &amp; KBC samples.
This indicates that variation within groups may be caused by the sample variation rather than media inconsistencies.
In fact the KBC and CVP seem to “normalise” the community profiles.
This is again hard to determine due to small numbers of replicates.</p>
</div>
<div id="site-or-media" class="section level3" number="18.5.3">
<h3><span class="header-section-number">18.5.3</span> Site or media</h3>
<p><img src="figures/site_and_media.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Is there more difference between the sites or the media used?</p>
<p>It appears that there is more difference cause by the media than the site.
If you only look at the ENV samples there is a big difference between the sites.
This could be investigated further by checking for <a href="https://microbiome.github.io/tutorials/PERMANOVA.html">homogeneity</a>.</p>
</div>
<div id="media-versus-env-samples" class="section level3" number="18.5.4">
<h3><span class="header-section-number">18.5.4</span> Media versus ENV samples</h3>
<p><img src="figures/petri.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Do the media samples differ from the ENV samples? If so, how?</p>
<p>This is a definite yes.</p>
<p>Compared to the ENV samples the media samples:</p>
<ul>
<li>Have lower biodiversity.</li>
<li>Consist primarily of a few taxa.</li>
<li>There appear to be less differences between the sites within the CVP &amp; KBC samples.</li>
</ul>
</div>
</div>
<div id="conclusion" class="section level2" number="18.6">
<h2><span class="header-section-number">18.6</span> Conclusion</h2>
<p><img src="figures/recap.png" width="20%" style="display: block; margin: auto;" /></p>
<p>We have successfully analysed our data with <code>phyloseq</code> and various other R packages.
The main difference found was between the ENV samples and the media samples.
If we had more ENV samples we could try to deduce more about the differences between the anthropisation gradient.</p>
<p>If you have time you could try to subset the data to only contain the ENV samples and carry out some analysis.</p>
<p>Great work and thanks for reading.</p>

</div>
</div>



<div id="mamba-installs" class="section level1" number="19">
<h1><span class="header-section-number">A</span> Mamba installs</h1>
<p><img src="figures/mamba_logo.png" width="20%" style="display: block; margin: auto;" /></p>
<div id="mamba_install" class="section level2" number="19.1">
<h2><span class="header-section-number">A.1</span> Mamba installation and environment</h2>
<p>Mamba is a reimplementation of conda.
It is a great tool for installing bioinformatic packages including R packages.</p>
<p>Mamba github: <a href="https://github.com/mamba-org/mamba" class="uri">https://github.com/mamba-org/mamba</a></p>
<p>Mamba installation: <a href="https://github.com/conda-forge/miniforge#mambaforge" class="uri">https://github.com/conda-forge/miniforge#mambaforge</a></p>
<p>Mamba guide: <a href="https://mamba.readthedocs.io/en/latest/user_guide/mamba.html" class="uri">https://mamba.readthedocs.io/en/latest/user_guide/mamba.html</a></p>
<p>To create the mamba environment <code>r_community</code> run the below commands in your bash. You will need to have installed <code>mamba</code> first.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co">#R community</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create r_community</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate r_community</span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> bioconda bioconductor-microbiome</span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> r r-tidyverse</span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-devtools</span>
<span id="cb120-7"><a href="#cb120-7" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-ggforce</span>
<span id="cb120-8"><a href="#cb120-8" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> bioconda bioconductor-ancombc</span>
<span id="cb120-9"><a href="#cb120-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Update r-htmltools as loading devtools doesn’t work in R (from 0.5.2 -&gt; 0.5.4)</span></span>
<span id="cb120-10"><a href="#cb120-10" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-htmltools=0.5.4</span>
<span id="cb120-11"><a href="#cb120-11" aria-hidden="true" tabindex="-1"></a><span class="co">#If using jupyter-notebook, need to install r kernel for it</span></span>
<span id="cb120-12"><a href="#cb120-12" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-irkernel</span></code></pre></div>
<p>Activate the environment in bash.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate r_community</span></code></pre></div>
<p>Activate your <code>mamba's</code> <code>R</code> and install <code>Qiime2R</code>.</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;jbisanz/qiime2R&quot;</span>)</span></code></pre></div>
</div>
<div id="jupyter_appendix" class="section level2" number="19.2">
<h2><span class="header-section-number">A.2</span> Jupyter-notebook</h2>
<p><img src="figures/jupyter_logo.png" width="15%" style="display: block; margin: auto;" /></p>
<p>If you are running this on your own computer you can use <code>RStudio</code>.
However, you can also use <a href="https://jupyter.org/"><code>Jupyter-notebook</code></a> if you are using an HPC or prefer it.</p>
<p>If using bash you will need to create an environment with <code>Jupyter-notebook</code>.
Ensure you are in the <code>(base)</code> mamba environment.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> jupyter</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate jupyter</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> anaconda jupyter</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> deactivate</span></code></pre></div>
<p>To run <code>Jupyter-notebook</code> with your <code>r_community</code> environment you can run the following.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Activate you R_community env</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate r_community</span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Run jupyter-notebook (may be a slightly different path)</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a><span class="ex">~/mamba/envs/jupyter/bin/jupyter-notebook</span></span></code></pre></div>
</div>
</div>
<div id="r-packages-1" class="section level1" number="20">
<h1><span class="header-section-number">B</span> R packages</h1>
<p><img src="figures/r_package.png" width="20%" style="display: block; margin: auto;" /></p>
<div id="ggplot2_appendix" class="section level2" number="20.1">
<h2><span class="header-section-number">B.1</span> ggplot2</h2>
<p><img src="figures/ggplot2.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Below are some useful resource if you would like to learn ggplot2.</p>
<p>The <a href="https://r-graphics.org/">R Graphics Cookbook</a> is a good place to start. It contains a section called <a href="https://r-graphics.org/chapter-ggplot2">Understanding ggplot2</a> in its appendix which is useful for learning some key terminologies and concepts.</p>
<p><code>ggplot2</code> requires its input to be in long format. You will therefore need to know how to <a href="#http://www.cookbook-r.com/Manipulating_data/Converting_data_between_wide_and_long_format/">convert your wide data to long data</a>.</p>
<p><a href="https://ggplot2.tidyverse.org/reference/#layers"><strong>ggplot2 function reference</strong></a>.</p>
</div>
<div id="phyloseq_appendix" class="section level2" number="20.2">
<h2><span class="header-section-number">B.2</span> Phyloseq</h2>
<p><img src="figures/phyloseq_logo.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Below are various resources if you would like to learn more about phyloseq and associated packages.</p>
<ul>
<li><a href="https://joey711.github.io/phyloseq/">Phyloseq</a>: The original <code>phyloseq</code>. The R object is still great but the included tutorials and some functions are a bit outdated.</li>
<li><a href="https://microbiome.github.io/tutorials/">microbiome</a>: A supplemental package to <code>phyloseq</code>. It contains some much needed functions. The tutorials are much clearer than the <code>phyloseq</code> tutorials but still not always the clearest.</li>
</ul>
</div>
<div id="importing-into-phyloseq" class="section level2" number="20.3">
<h2><span class="header-section-number">B.3</span> Importing into phyloseq</h2>
<p><img src="figures/import.png" width="10%" style="display: block; margin: auto;" /></p>
<ul>
<li><a href="https://github.com/jbisanz/qiime2R">QIIME2R</a>: A very useful package to import QIIME2 artifacts into a <code>phyloseq</code> object.</li>
<li><a href="https://joey711.github.io/phyloseq/import-data.html">Phyloseqize R data</a>: A tutorial on how to import data frames into a <code>phyloseq</code> object.</li>
<li><strong>Shotgun metagenomics data</strong>
<ul>
<li>You can import Kraken2 data by converting the <a href="https://github.com/smdabdoub/kraken-biom">report file to a biom file</a> then importing the biom file into a <code>phyloseq</code> object with the <a href="https://github.com/jbisanz/qiime2R"><code>QIIME2R</code> function <code>read_q2biom()</code></a>.</li>
<li>You can also import the <a href="https://github.com/jenniferlu717/Bracken#output-kraken-style-bracken-report">Kraken-Style Bracken report</a> in the above manner.</li>
<li>The <a href="https://biom-format.org/">biom format</a> stands for the Biological Observation Matrix (BIOM) format. It is used and created by a variety of tools.</li>
</ul></li>
</ul>
</div>
</div>
<div id="diversity-metrics" class="section level1" number="21">
<h1><span class="header-section-number">C</span> Diversity metrics</h1>
<p><img src="figures/metric_square.png" width="20%" style="display: block; margin: auto;" /></p>
<div id="alpha_appendix" class="section level2" number="21.1">
<h2><span class="header-section-number">C.1</span> Alpha diversity</h2>
<p><img src="figures/alpha.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Alpha diversity measures assess the diversity of each sample separately. Generally a higher value of these indexes/measures indicates higher diversity/evenness.</p>
<div id="obvs" class="section level3" number="21.1.1">
<h3><span class="header-section-number">C.1.1</span> Observed features</h3>
<p>The observed number of features is defined as the number of distinct features, such as ASVs, within a sample.</p>
<p>The number of observed features can also be known as the feature richness of a sample. A sample with more present features than another would be said to be richer.</p>
</div>
<div id="chao" class="section level3" number="21.1.2">
<h3><span class="header-section-number">C.1.2</span> Chao</h3>
<p>This is an estimation of the actual number of features within a sample.</p>
<p><strong>Equation</strong>:
<span class="math display">\[
C =
O +
\frac{f1(f1-1)}
{2(f2 + 1)}
\]</span></p>
<ul>
<li>C = Chao1 estimator, i.e. estimated total feature richness.</li>
<li>O = Observed number of features.</li>
<li>f1 = Frequency of singletons (features with total abundance of 1).</li>
<li>f2 = Frequency of doubletons (features with total abundance of 2).</li>
</ul>
</div>
<div id="evenness" class="section level3" number="21.1.3">
<h3><span class="header-section-number">C.1.3</span> Evenness</h3>
<p>How evenly spread the abundances are across all present features.</p>
<ul>
<li>A sample with 4 ASVs, each with 25% relative abundance would be perfectly even.</li>
<li>A sample with 4 ASVs, where one ASV has 97% abundance and the other 3 have 1%, would be highly uneven.</li>
</ul>
</div>
<div id="faith" class="section level3" number="21.1.4">
<h3><span class="header-section-number">C.1.4</span> Faith’s PD (phylogenetic diversity)</h3>
<p>Faith’s PD represents the minimum total branch length that covers all taxa within the sample on a phylogenetic tree (Faith, 1992). A smaller PD value indicates a reduced expected taxonomic diversity whilst a large PD value indicates a higher expected diversity.</p>
<p>A sample with 10 ASVs could have a lower Faith’s PD than a sample with only 2 ASVs. This could occur if the 10 ASV sample only has ASVs from one genus whilst the 2 ASV samples consists of ASVs from 2 different Families.</p>
<p><img src="figures/faith_pd.png" width="60%" style="display: block; margin: auto;" /></p>
</div>
<div id="simpson" class="section level3" number="21.1.5">
<h3><span class="header-section-number">C.1.5</span> Simpson</h3>
<p>A measure of diversity based on number of features present and the relative abundance of each feature. If richness and evenness increase the diversity score increases.</p>
<p><strong>Equation</strong>:
<span class="math display">\[
D = 1 -
{(
\frac
{\sum n{(n-1)}}
{N{(N-1)}}
)}
\]</span></p>
<ul>
<li>D = Simpson diversity index</li>
<li>n = Abundance of feature</li>
<li>N = Total feature abundance of sample</li>
</ul>
<p>The values range from 0 (no diversity) to 1 (infinite diversity).</p>
</div>
<div id="simpsone" class="section level3" number="21.1.6">
<h3><span class="header-section-number">C.1.6</span> Simpson evenness measure (simpson_e)</h3>
<p>This is a measure of evenness based on the Simpson index. It ranges from 0 (lowest eveness) to 1 (complete evenness). It compares the calculated Simpson Index of a sample to its theoretical maximum if the sample was perfectly even but had the same amount of features.</p>
<p><strong>Equation</strong>:
<span class="math display">\[
E_D =
\frac{D}
{D_m}
\]</span></p>
<ul>
<li>E<sub>D</sub> = Simpson evenness measure</li>
<li>D = Simpson diversity index</li>
<li>D<sub>m</sub> = Max possible Simpson diveristy index given the number of features</li>
</ul>
</div>
<div id="shannon" class="section level3" number="21.1.7">
<h3><span class="header-section-number">C.1.7</span> Shannon</h3>
<p>A measure of diversity where a higher number means higher diversity. Shannon’s index accounts for both abundance and evenness of the feaures present.</p>
<p><strong>Equation</strong>:</p>
<p><span class="math display">\[
H =
-\sum_{i=1}^{n} p_i lnp_i
\]</span></p>
<ul>
<li>H = Shannon diversity index</li>
<li>p = n/N</li>
<li>n = Abundance of feature</li>
<li>N = Total feature abundance of sample</li>
</ul>
</div>
</div>
<div id="beta_appendix" class="section level2" number="21.2">
<h2><span class="header-section-number">C.2</span> Beta diversity</h2>
<p><img src="figures/beta.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Beta diversity compares 2 samples at a time. This is measured in terms of dissimilarity:</p>
<ul>
<li>A lower score indicate the 2 samples are more similar.</li>
<li>A higher score indicates the 2 samples are more dissimilar.</li>
</ul>
<div id="unifrac_explanation" class="section level3" number="21.2.1">
<h3><span class="header-section-number">C.2.1</span> Weighted and unweighted UniFrac distances</h3>
<p>The UniFrac metric is a phylogenetic distance measure between two samples. It is defined as “the sum of the unshared branch lengths between two samples divided by the total branch length (shared + unshared) of two samples” (Lozupone and Knight, 2005). This results in calculating the fraction of the branch lengths unique to each sample (ie. the higher this value is, the more dissimilar two samples are). The entire phylogenetic tree created for all the sequences of all the analysed samples is used for this calculation.</p>
<ul>
<li>Unweighted UniFrac: Only considers feature presence/absence.</li>
<li>Weighted UniFrac: Takes into account feature presence/absence &amp; abundance. I.e. the value is weighted by the abundances.</li>
</ul>
</div>
<div id="bray_explanation" class="section level3" number="21.2.2">
<h3><span class="header-section-number">C.2.2</span> Bray-Curtis</h3>
<p>The Bray-Curtis metric is a dissimilarity measure that can quantify the level of difference between two samples. Two identical samples would have a Bray-Curtis measure of 0 (i.e. they have 0 dissimilarity).</p>
<p>There are 2 versions but the one used by QIIME2 looks at the number of features shared by the 2 samples.</p>
<p><strong>Equation</strong>:</p>
<p><span class="math display">\[
CBC = 1 –
(\frac{2c}
{a + b})
\]</span></p>
<ul>
<li>c = # features present in both samples</li>
<li>a = # features present in sample a</li>
<li>b = # features present in sample b</li>
</ul>
</div>
</div>
<div id="diversity-resources" class="section level2" number="21.3">
<h2><span class="header-section-number">C.3</span> Diversity Resources</h2>
<p>For more diveristy measure please see: <a href="https://forum.qiime2.org/t/alpha-and-beta-diversity-explanations-and-commands/2282" class="uri">https://forum.qiime2.org/t/alpha-and-beta-diversity-explanations-and-commands/2282</a></p>

</div>
</div>
<p style="text-align: center;">
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
