<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Iterating rarefaction | R community advanced analysis</title>
  <meta name="description" content="NEOF book for the R community advanced analysis course" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Iterating rarefaction | R community advanced analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the R community advanced analysis course" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Iterating rarefaction | R community advanced analysis" />
  
  <meta name="twitter:description" content="NEOF book for the R community advanced analysis course" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew R. Gemmell" />


<meta name="date" content="2024-07-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="07-Rarefaction_and_random_seeds.html"/>
<link rel="next" href="09-Alpha_diversity.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/NEOF_bordered_small.png"></a></li>
<li><a>R community analysis</a></li>

<li class="divider"></li>
<li class="part"><span><b>Intro</b></span></li>
<li class="chapter" data-level="1" data-path="01-R_community_advanced_main_book.html"><a href="01-R_community_advanced_main_book.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="01-R_community_advanced_main_book.html"><a href="01-R_community_advanced_main_book.html#table-of-contents"><i class="fa fa-check"></i>Table of contents</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html"><i class="fa fa-check"></i><b>2</b> Dataset &amp; workflow</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#dataset"><i class="fa fa-check"></i><b>2.1</b> Dataset</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sites"><i class="fa fa-check"></i><b>2.1.1</b> Sites</a></li>
<li class="chapter" data-level="2.1.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#culture-media"><i class="fa fa-check"></i><b>2.1.2</b> Culture media</a></li>
<li class="chapter" data-level="2.1.3" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sum_and_qs"><i class="fa fa-check"></i><b>2.1.3</b> Summary &amp; questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#workflow"><i class="fa fa-check"></i><b>2.2</b> Workflow</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-R_packages.html"><a href="03-R_packages.html"><i class="fa fa-check"></i><b>3</b> R Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-R_packages.html"><a href="03-R_packages.html#r-packageslibraries"><i class="fa fa-check"></i><b>3.1</b> R packages/libraries</a></li>
<li class="chapter" data-level="3.2" data-path="03-R_packages.html"><a href="03-R_packages.html#the-grammar-of-graphics"><i class="fa fa-check"></i><b>3.2</b> The grammar of graphics</a></li>
<li class="chapter" data-level="3.3" data-path="03-R_packages.html"><a href="03-R_packages.html#phyloseq"><i class="fa fa-check"></i><b>3.3</b> phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Setup.html"><a href="04-Setup.html"><i class="fa fa-check"></i><b>4</b> Set-up</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Setup.html"><a href="04-Setup.html#cluster"><i class="fa fa-check"></i><b>4.1</b> Logon instructions</a></li>
<li class="chapter" data-level="4.2" data-path="04-Setup.html"><a href="04-Setup.html#mamba"><i class="fa fa-check"></i><b>4.2</b> Mamba</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Jupyter.html"><a href="05-Jupyter.html"><i class="fa fa-check"></i><b>5</b> Jupyter</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Jupyter.html"><a href="05-Jupyter.html#open-jupyter-notebook"><i class="fa fa-check"></i><b>5.1</b> Open Jupyter-notebook</a></li>
<li class="chapter" data-level="5.2" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-r-notebook"><i class="fa fa-check"></i><b>5.2</b> Create R notebook</a></li>
<li class="chapter" data-level="5.3" data-path="05-Jupyter.html"><a href="05-Jupyter.html#cells-and-code"><i class="fa fa-check"></i><b>5.3</b> Cells and code</a></li>
<li class="chapter" data-level="5.4" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-new-cells"><i class="fa fa-check"></i><b>5.4</b> Create new cells</a></li>
<li class="chapter" data-level="5.5" data-path="05-Jupyter.html"><a href="05-Jupyter.html#running-code"><i class="fa fa-check"></i><b>5.5</b> Running code</a></li>
<li class="chapter" data-level="5.6" data-path="05-Jupyter.html"><a href="05-Jupyter.html#saving-the-file"><i class="fa fa-check"></i><b>5.6</b> Saving the file</a></li>
<li class="chapter" data-level="5.7" data-path="05-Jupyter.html"><a href="05-Jupyter.html#title-cells-with-markdown"><i class="fa fa-check"></i><b>5.7</b> Title cells with markdown</a></li>
<li class="chapter" data-level="5.8" data-path="05-Jupyter.html"><a href="05-Jupyter.html#close-the-notebook"><i class="fa fa-check"></i><b>5.8</b> Close the notebook</a></li>
<li class="chapter" data-level="5.9" data-path="05-Jupyter.html"><a href="05-Jupyter.html#jup_vid_tut"><i class="fa fa-check"></i><b>5.9</b> Video tutorial</a></li>
</ul></li>
<li class="part"><span><b>Iterative rarefaction</b></span></li>
<li class="chapter" data-level="6" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html"><i class="fa fa-check"></i><b>6</b> Iterative rarefaction intro</a>
<ul>
<li class="chapter" data-level="6.1" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html#should-you-rarefy"><i class="fa fa-check"></i><b>6.1</b> Should you rarefy?</a></li>
<li class="chapter" data-level="6.2" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html#using-iterations"><i class="fa fa-check"></i><b>6.2</b> Using iterations</a></li>
<li class="chapter" data-level="6.3" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html#section-contents"><i class="fa fa-check"></i><b>6.3</b> Section contents</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html"><i class="fa fa-check"></i><b>7</b> Random seeds</a>
<ul>
<li class="chapter" data-level="7.1" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-notebook"><i class="fa fa-check"></i><b>7.1</b> Random seed notebook</a></li>
<li class="chapter" data-level="7.2" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-sampling"><i class="fa fa-check"></i><b>7.2</b> Random sampling</a></li>
<li class="chapter" data-level="7.3" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#sampling-with-replacement"><i class="fa fa-check"></i><b>7.3</b> Sampling with replacement</a></li>
<li class="chapter" data-level="7.4" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#setting-a-random-seed"><i class="fa fa-check"></i><b>7.4</b> Setting a random seed</a></li>
<li class="chapter" data-level="7.5" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#reset-seed"><i class="fa fa-check"></i><b>7.5</b> Reset seed</a></li>
<li class="chapter" data-level="7.6" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-practice"><i class="fa fa-check"></i><b>7.6</b> Random seed practice</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-question-1"><i class="fa fa-check"></i><b>7.6.1</b> Random seed: Question 1</a></li>
<li class="chapter" data-level="7.6.2" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-question-2"><i class="fa fa-check"></i><b>7.6.2</b> Random seed: Question 2</a></li>
<li class="chapter" data-level="7.6.3" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-question-3"><i class="fa fa-check"></i><b>7.6.3</b> Random seed: Question 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html"><i class="fa fa-check"></i><b>8</b> Iterating rarefaction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-dataset"><i class="fa fa-check"></i><b>8.1</b> Iterating rarefaction dataset</a></li>
<li class="chapter" data-level="8.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-setup"><i class="fa fa-check"></i><b>8.2</b> Iterating rarefaction setup</a></li>
<li class="chapter" data-level="8.3" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#rarefaction-iterations"><i class="fa fa-check"></i><b>8.3</b> Rarefaction iterations</a></li>
<li class="chapter" data-level="8.4" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#rng_vec_creation"><i class="fa fa-check"></i><b>8.4</b> RNG vector creation</a></li>
<li class="chapter" data-level="8.5" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#phyla-relative-abundance"><i class="fa fa-check"></i><b>8.5</b> Phyla relative abundance</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#subset-and-phyla-aggregation"><i class="fa fa-check"></i><b>8.5.1</b> Subset and phyla aggregation</a></li>
<li class="chapter" data-level="8.5.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#phyla-relative-abundance-bar-chart"><i class="fa fa-check"></i><b>8.5.2</b> Phyla relative abundance bar chart</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#one-round-1r-of-rarefaction"><i class="fa fa-check"></i><b>8.6</b> One Round (1R) of rarefaction</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#r-rarefaction"><i class="fa fa-check"></i><b>8.6.1</b> 1R: Rarefaction</a></li>
<li class="chapter" data-level="8.6.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#r-relative-abundance-bar-chart"><i class="fa fa-check"></i><b>8.6.2</b> 1R: Relative abundance bar chart</a></li>
<li class="chapter" data-level="8.6.3" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#r-difference-from-non-rarefied"><i class="fa fa-check"></i><b>8.6.3</b> 1R: Difference from non-rarefied</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#multiple-rarefaction-mr-iterations"><i class="fa fa-check"></i><b>8.7</b> Multiple Rarefaction (MR) iterations</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#mr-rarefaction"><i class="fa fa-check"></i><b>8.7.1</b> MR: Rarefaction</a></li>
<li class="chapter" data-level="8.7.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#mr-difference-from-non-rarefied"><i class="fa fa-check"></i><b>8.7.2</b> MR: Difference from non-rarefied</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-task"><i class="fa fa-check"></i><b>8.8</b> Iterating rarefaction task</a></li>
<li class="chapter" data-level="8.9" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-recap"><i class="fa fa-check"></i><b>8.9</b> Iterating rarefaction recap</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html"><i class="fa fa-check"></i><b>9</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="9.1" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-setup"><i class="fa fa-check"></i><b>9.1</b> <span class="math inline">\(\alpha\)</span>: Setup</a></li>
<li class="chapter" data-level="9.2" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-iterative-rarefaction-values"><i class="fa fa-check"></i><b>9.2</b> <span class="math inline">\(\alpha\)</span>: Iterative rarefaction values</a></li>
<li class="chapter" data-level="9.3" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-iterative-rarefaction-loop"><i class="fa fa-check"></i><b>9.3</b> <span class="math inline">\(\alpha\)</span>: Iterative rarefaction loop</a></li>
<li class="chapter" data-level="9.4" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-metric-and-metadata-data-frame"><i class="fa fa-check"></i><b>9.4</b> <span class="math inline">\(\alpha\)</span>: Metric and metadata data frame</a></li>
<li class="chapter" data-level="9.5" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha_long_df"><i class="fa fa-check"></i><b>9.5</b> <span class="math inline">\(\alpha\)</span>: Long data frame</a></li>
<li class="chapter" data-level="9.6" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-subset-metrics"><i class="fa fa-check"></i><b>9.6</b> <span class="math inline">\(\alpha\)</span>: Subset metrics</a></li>
<li class="chapter" data-level="9.7" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-violin-plot"><i class="fa fa-check"></i><b>9.7</b> <span class="math inline">\(\alpha\)</span>: Violin plot</a></li>
<li class="chapter" data-level="9.8" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-stats"><i class="fa fa-check"></i><b>9.8</b> <span class="math inline">\(\alpha\)</span>: Stats</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#kruskal-wallis-test"><i class="fa fa-check"></i><b>9.8.1</b> Kruskal Wallis test</a></li>
<li class="chapter" data-level="9.8.2" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#paired-wilcoxon-test"><i class="fa fa-check"></i><b>9.8.2</b> Paired wilcoxon test</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-plot-with-stats"><i class="fa fa-check"></i><b>9.9</b> <span class="math inline">\(\alpha\)</span>: Plot with stats</a>
<ul>
<li class="chapter" data-level="9.9.1" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#list-of-comparisons"><i class="fa fa-check"></i><b>9.9.1</b> List of comparisons</a></li>
<li class="chapter" data-level="9.9.2" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#violin-plot-with-stats"><i class="fa fa-check"></i><b>9.9.2</b> Violin plot with stats</a></li>
<li class="chapter" data-level="9.9.3" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#reorder-x-axis-and-stats"><i class="fa fa-check"></i><b>9.9.3</b> Reorder x-axis and stats</a></li>
</ul></li>
<li class="chapter" data-level="9.10" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-task"><i class="fa fa-check"></i><b>9.10</b> <span class="math inline">\(\alpha\)</span>: Task</a></li>
<li class="chapter" data-level="9.11" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-recap"><i class="fa fa-check"></i><b>9.11</b> <span class="math inline">\(\alpha\)</span>: Recap</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html"><i class="fa fa-check"></i><b>10</b> Beta Diversity</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R community advanced analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="iterating_rarefaction_chap" class="section level1 hasAnchor" number="8">
<h1><span class="header-section-number">Chapter 8</span> Iterating rarefaction<a href="08-Iterating_rarefaction.html#iterating_rarefaction_chap" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<center>
<img src="figures/france.png" style="width:200px" />
</center>
<p>In this chapter we will create code to carry out <strong>iterative rarefaction</strong>.
For this we create a vector of <strong>random seeds</strong>, each used for a different <strong>iteration</strong> of <strong>rarefaction</strong>.
We will loop through these <strong>random seeds</strong>, using each <strong>random seed</strong> to carry out one <strong>iteration</strong> of <strong>rarefaction</strong>.
In the next chapters we will utilise this code to produce alpha and beta diversity values that we will analyse.</p>
<div id="iterating-rarefaction-dataset" class="section level2 hasAnchor" number="8.1">
<h2><span class="header-section-number">8.1</span> Iterating rarefaction dataset<a href="08-Iterating_rarefaction.html#iterating-rarefaction-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/river.png" style="width:200px" />
</center>
<p>We will utilise the same dataset used in the <a href="https://neof-workshops.github.io/R_community_whqkt8/Course/02-Dataset_and_workflow.html#dataset">R community analysis workshop</a>.</p>
<p>Below are brief bullet points about the data:</p>
<ul>
<li>It is a 16S dataset of ASV counts with taxonomy and phylogeny produced by QIIME2</li>
<li>The samples come from surface water from the Durance River in the south-east of France</li>
<li>There are three sampling sites on an anthropisation gradient (low to high agriculture)
<ul>
<li>Upper Durance (UD)</li>
<li>Middle Durance (MD)</li>
<li>Lower Durance (LD)</li>
</ul></li>
<li>Four different media approaches were used to produce bacterial lawns that were sequenced
<ul>
<li>Environmental sample (ENV): No media used, frozen at -20°C.</li>
<li>TSA 10% incubated at 28°C for 2 days.</li>
<li>KBC incubated at 28°C for 2 days.</li>
<li>CVP incubated at 28°C for 3 days.</li>
</ul></li>
<li>There are three replicates for each sampling site and media combination (36 samples total)</li>
</ul>
</div>
<div id="iterating-rarefaction-setup" class="section level2 hasAnchor" number="8.2">
<h2><span class="header-section-number">8.2</span> Iterating rarefaction setup<a href="08-Iterating_rarefaction.html#iterating-rarefaction-setup" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/setup_5.png" style="width:200px" />
</center>
<p>First, create a new R jupyter-notebook called "Iterating_rarefaction.ipynb".</p>
<p>At the top of this notebook create a code cell to load in the various packages and data we need. The code is below:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb18-1"><a href="08-Iterating_rarefaction.html#cb18-1" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb18-2"><a href="08-Iterating_rarefaction.html#cb18-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb18-3"><a href="08-Iterating_rarefaction.html#cb18-3" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb18-4"><a href="08-Iterating_rarefaction.html#cb18-4" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb18-5"><a href="08-Iterating_rarefaction.html#cb18-5" tabindex="-1"></a><span class="co">#Load processed but unrarefied data from R community analysis workshop</span></span>
<span id="cb18-6"><a href="08-Iterating_rarefaction.html#cb18-6" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
</div>
<div id="rarefaction-iterations" class="section level2 hasAnchor" number="8.3">
<h2><span class="header-section-number">8.3</span> Rarefaction iterations<a href="08-Iterating_rarefaction.html#rarefaction-iterations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/number_pad.png" style="width:200px" />
</center>
<p>We need to choose the number of <strong>iterations</strong> we are going to carry out.</p>
<p>For our <strong>practice</strong> we will use <strong>10 iterations</strong> for speed.
In your real analysis I would recommend using <strong>1000 iterations</strong>.</p>
<p>Let's create a variable for our number of <strong>iterations</strong>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb19-1"><a href="08-Iterating_rarefaction.html#cb19-1" tabindex="-1"></a><span class="co">#Number of rarefaction iterations to be carried out</span></span>
<span id="cb19-2"><a href="08-Iterating_rarefaction.html#cb19-2" tabindex="-1"></a><span class="co">#Using 10 here for speed, real analysis should use 1000</span></span>
<span id="cb19-3"><a href="08-Iterating_rarefaction.html#cb19-3" tabindex="-1"></a>rarefaction_iters <span class="ot">&lt;-</span> <span class="dv">10</span></span></code></pre></div>
</div>
<div id="rng_vec_creation" class="section level2 hasAnchor" number="8.4">
<h2><span class="header-section-number">8.4</span> RNG vector creation<a href="08-Iterating_rarefaction.html#rng_vec_creation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/list.png" style="width:200px; background-color:white; border-radius:15px" />
</center>
<p>We can now carry out <strong>Random Number Generation (RNG)</strong> to create a number of <strong>random seeds</strong> equal to the number of iterations planned.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb20-1"><a href="08-Iterating_rarefaction.html#cb20-1" tabindex="-1"></a><span class="co">#Create rngseed vector</span></span>
<span id="cb20-2"><a href="08-Iterating_rarefaction.html#cb20-2" tabindex="-1"></a><span class="co">#Set seed for reproducibility</span></span>
<span id="cb20-3"><a href="08-Iterating_rarefaction.html#cb20-3" tabindex="-1"></a><span class="co">#This number was chosen randomly</span></span>
<span id="cb20-4"><a href="08-Iterating_rarefaction.html#cb20-4" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2605</span>)</span>
<span id="cb20-5"><a href="08-Iterating_rarefaction.html#cb20-5" tabindex="-1"></a><span class="co">#Sample 10 (number of iters) values from the number range 1-100,000</span></span>
<span id="cb20-6"><a href="08-Iterating_rarefaction.html#cb20-6" tabindex="-1"></a>rngseed_vec <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">100000</span>, <span class="at">size =</span> rarefaction_iters, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-7"><a href="08-Iterating_rarefaction.html#cb20-7" tabindex="-1"></a><span class="co">#Print out vector</span></span>
<span id="cb20-8"><a href="08-Iterating_rarefaction.html#cb20-8" tabindex="-1"></a>rngseed_vec</span>
<span id="cb20-9"><a href="08-Iterating_rarefaction.html#cb20-9" tabindex="-1"></a><span class="co">#Save our rngseed vector</span></span>
<span id="cb20-10"><a href="08-Iterating_rarefaction.html#cb20-10" tabindex="-1"></a><span class="fu">save</span>(rngseed_vec, <span class="at">file=</span><span class="st">&quot;rngseeds.RData&quot;</span>)</span>
<span id="cb20-11"><a href="08-Iterating_rarefaction.html#cb20-11" tabindex="-1"></a><span class="co">#Reset seed</span></span>
<span id="cb20-12"><a href="08-Iterating_rarefaction.html#cb20-12" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="cn">NULL</span>)</span></code></pre></div>
<p>There are a lot of steps above. These are:</p>
<ul>
<li><strong>Setting the random seed:</strong> We carry this out so we will always get the same rngseed vector that will be used for the rarefaction iterations. This is important so you will always get the same results if you need to rework some analysis, stats, or plots. Also useful here so you get the same results as the instructor and other attendees.</li>
<li><strong>Creating the rngseed vector:</strong> We use our old friend <code>sample()</code> to create a random number for each <strong>iteration</strong> we will carry out.
<ul>
<li>We arbitrarily sample from the numbers 1-100,000, you could change this to a larger range in your future research.</li>
<li>We use our previous object <code>rarefaction_iters</code> as <code>size=</code> to produce a random number for each of our <strong>iterations</strong>.</li>
<li>We carry this out <strong>without replacement</strong> so none of our <strong>rarefaction iterations</strong> are identical.</li>
</ul></li>
<li><strong>Save the rngseed vector:</strong> We save the vector as a file. We will load this in our alpha and beta diversity notebooks to be used for iterative rarefaction. This is also useful so you have a backup file of the rngseed vectors.</li>
<li><strong>Reset seed:</strong> Always good to reset the seed at the end of a cell.</li>
</ul>
</div>
<div id="phyla-relative-abundance" class="section level2 hasAnchor" number="8.5">
<h2><span class="header-section-number">8.5</span> Phyla relative abundance<a href="08-Iterating_rarefaction.html#phyla-relative-abundance" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/river_2.png" style="width:200px" />
</center>
<p>Prior to <strong>iterative rarefaction</strong> we will look at the phyla composition of the environmental samples.</p>
<div id="subset-and-phyla-aggregation" class="section level3 hasAnchor" number="8.5.1">
<h3><span class="header-section-number">8.5.1</span> Subset and phyla aggregation<a href="08-Iterating_rarefaction.html#subset-and-phyla-aggregation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/aggregate.png" style="width:200px" />
</center>
<p>For demonstrative purposes we will reduce the amount of samples and features in our data for this chapter. We will carry this out by:</p>
<ul>
<li>Subsetting the data so it only contains the 9 environmental samples.</li>
<li>Aggregate the taxa to phyla whilst aggregating rare taxa to one "other group"</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb21-1"><a href="08-Iterating_rarefaction.html#cb21-1" tabindex="-1"></a><span class="co">#Reduce data for demonstrative purposes</span></span>
<span id="cb21-2"><a href="08-Iterating_rarefaction.html#cb21-2" tabindex="-1"></a><span class="co">#Subset phyloseq object to only retain the ENV samples</span></span>
<span id="cb21-3"><a href="08-Iterating_rarefaction.html#cb21-3" tabindex="-1"></a><span class="co">#I.e. remove the media samples</span></span>
<span id="cb21-4"><a href="08-Iterating_rarefaction.html#cb21-4" tabindex="-1"></a>physeq_env <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pseq, media <span class="sc">==</span> <span class="st">&quot;ENV&quot;</span>)</span>
<span id="cb21-5"><a href="08-Iterating_rarefaction.html#cb21-5" tabindex="-1"></a></span>
<span id="cb21-6"><a href="08-Iterating_rarefaction.html#cb21-6" tabindex="-1"></a><span class="co">#Aggregate to phyla level whilst aggregating rare phyla</span></span>
<span id="cb21-7"><a href="08-Iterating_rarefaction.html#cb21-7" tabindex="-1"></a>pseq_env_phyla <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_rare</span>(pseq_env, <span class="at">level =</span> <span class="st">&quot;Phylum&quot;</span>,</span>
<span id="cb21-8"><a href="08-Iterating_rarefaction.html#cb21-8" tabindex="-1"></a>                                             <span class="at">detection =</span> <span class="fl">0.1</span>, <span class="at">prevalence =</span> <span class="fl">0.5</span>,</span>
<span id="cb21-9"><a href="08-Iterating_rarefaction.html#cb21-9" tabindex="-1"></a>                                             <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-10"><a href="08-Iterating_rarefaction.html#cb21-10" tabindex="-1"></a><span class="co">#View count table</span></span>
<span id="cb21-11"><a href="08-Iterating_rarefaction.html#cb21-11" tabindex="-1"></a><span class="fu">otu_table</span>(pseq_env_phyla)</span>
<span id="cb21-12"><a href="08-Iterating_rarefaction.html#cb21-12" tabindex="-1"></a><span class="co">#Sum count of samples</span></span>
<span id="cb21-13"><a href="08-Iterating_rarefaction.html#cb21-13" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_env_phyla)</span>
<span id="cb21-14"><a href="08-Iterating_rarefaction.html#cb21-14" tabindex="-1"></a><span class="co">#Remove unwanted objects</span></span>
<span id="cb21-15"><a href="08-Iterating_rarefaction.html#cb21-15" tabindex="-1"></a><span class="fu">rm</span>(pseq, pseq_env)</span></code></pre></div>
</div>
<div id="phyla-relative-abundance-bar-chart" class="section level3 hasAnchor" number="8.5.2">
<h3><span class="header-section-number">8.5.2</span> Phyla relative abundance bar chart<a href="08-Iterating_rarefaction.html#phyla-relative-abundance-bar-chart" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/stacked_bar_chart.png" style="width:200px" />
</center>
<p>Let's have a quick look at the non rarefied phyla relative abundance through a bar chart.</p>
<p><strong>Note:</strong> This is how you would normally look at this type of bar chart.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb22-1"><a href="08-Iterating_rarefaction.html#cb22-1" tabindex="-1"></a><span class="co">#Quick phyla bar chart of relative abundance</span></span>
<span id="cb22-2"><a href="08-Iterating_rarefaction.html#cb22-2" tabindex="-1"></a><span class="co">#Relative abundance transformation</span></span>
<span id="cb22-3"><a href="08-Iterating_rarefaction.html#cb22-3" tabindex="-1"></a>pseq_env_phyla_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq_env_phyla, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb22-4"><a href="08-Iterating_rarefaction.html#cb22-4" tabindex="-1"></a><span class="co">#Create, save, and display bar chart</span></span>
<span id="cb22-5"><a href="08-Iterating_rarefaction.html#cb22-5" tabindex="-1"></a>phylum_bar <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(pseq_env_phyla_relabund)</span>
<span id="cb22-6"><a href="08-Iterating_rarefaction.html#cb22-6" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./env_phyla_relabund.png&quot;</span>, <span class="at">plot =</span> phylum_bar,</span>
<span id="cb22-7"><a href="08-Iterating_rarefaction.html#cb22-7" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">100</span>, <span class="at">width =</span> <span class="dv">100</span>)</span>
<span id="cb22-8"><a href="08-Iterating_rarefaction.html#cb22-8" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file =</span> <span class="st">&quot;./env_phyla_relabund.png&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="one-round-1r-of-rarefaction" class="section level2 hasAnchor" number="8.6">
<h2><span class="header-section-number">8.6</span> One Round (1R) of rarefaction<a href="08-Iterating_rarefaction.html#one-round-1r-of-rarefaction" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/one_round.png" style="width:200px" />
</center>
<p>We will first carry out one round of <strong>rarefaction</strong> .
This is so we can get a reminder of how to carry it out and to compare the results of no <strong>rarefaction</strong> and only one round.</p>
<div id="r-rarefaction" class="section level3 hasAnchor" number="8.6.1">
<h3><span class="header-section-number">8.6.1</span> 1R: Rarefaction<a href="08-Iterating_rarefaction.html#r-rarefaction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/black_rhino.png" style="width:300px" />
</center>
<p>Carry out one round of <strong>rarefaction</strong> and view the <strong>rarefied</strong> counts.</p>
<p>We are using the environmental samples subsetted and phyla aggregated data.
Additionally, we are using the first of our <strong>random seeds</strong> in <code>rng_seed_vec</code> and the minimum read count as our <strong>rarefaction</strong> size.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb23-1"><a href="08-Iterating_rarefaction.html#cb23-1" tabindex="-1"></a><span class="co">#One round of rarefaction</span></span>
<span id="cb23-2"><a href="08-Iterating_rarefaction.html#cb23-2" tabindex="-1"></a>pseq_env_phyla_rarefy_1 <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(</span>
<span id="cb23-3"><a href="08-Iterating_rarefaction.html#cb23-3" tabindex="-1"></a>  pseq_env_phyla,</span>
<span id="cb23-4"><a href="08-Iterating_rarefaction.html#cb23-4" tabindex="-1"></a>  <span class="co">#Minimum read count as rarefaction size</span></span>
<span id="cb23-5"><a href="08-Iterating_rarefaction.html#cb23-5" tabindex="-1"></a>  <span class="at">sample.size =</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_env_phyla)),</span>
<span id="cb23-6"><a href="08-Iterating_rarefaction.html#cb23-6" tabindex="-1"></a>  <span class="co">#First random seed as the rng seed</span></span>
<span id="cb23-7"><a href="08-Iterating_rarefaction.html#cb23-7" tabindex="-1"></a>  <span class="at">rngseed =</span> rngseed_vec[<span class="dv">1</span>])</span>
<span id="cb23-8"><a href="08-Iterating_rarefaction.html#cb23-8" tabindex="-1"></a><span class="co">#View count table</span></span>
<span id="cb23-9"><a href="08-Iterating_rarefaction.html#cb23-9" tabindex="-1"></a><span class="fu">otu_table</span>(pseq_env_phyla_rarefy_1)</span>
<span id="cb23-10"><a href="08-Iterating_rarefaction.html#cb23-10" tabindex="-1"></a><span class="co">#Sum count of samples</span></span>
<span id="cb23-11"><a href="08-Iterating_rarefaction.html#cb23-11" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_env_phyla_rarefy_1)</span></code></pre></div>
<p>You should see that all the samples now have a total count of <strong>11046</strong>.</p>
</div>
<div id="r-relative-abundance-bar-chart" class="section level3 hasAnchor" number="8.6.2">
<h3><span class="header-section-number">8.6.2</span> 1R: Relative abundance bar chart<a href="08-Iterating_rarefaction.html#r-relative-abundance-bar-chart" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/stacked_bar_chart_2.png" style="width:200px" />
</center>
<p>Now to create a relative abundance bar chart with our <strong>rarefied</strong> data to compare to our non-rarefied data.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb24-1"><a href="08-Iterating_rarefaction.html#cb24-1" tabindex="-1"></a><span class="co">#Quick phyla bar chart of relative abundance</span></span>
<span id="cb24-2"><a href="08-Iterating_rarefaction.html#cb24-2" tabindex="-1"></a><span class="co">#Relative abundance transformation</span></span>
<span id="cb24-3"><a href="08-Iterating_rarefaction.html#cb24-3" tabindex="-1"></a>pseq_env_phyla_rarefy_1_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq_env_phyla_rarefy_1, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb24-4"><a href="08-Iterating_rarefaction.html#cb24-4" tabindex="-1"></a><span class="co">#Create, save, and display bar chart</span></span>
<span id="cb24-5"><a href="08-Iterating_rarefaction.html#cb24-5" tabindex="-1"></a>phylum_bar <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(pseq_env_phyla_rarefy_1_relabund)</span>
<span id="cb24-6"><a href="08-Iterating_rarefaction.html#cb24-6" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./env_phyla_rarefy_1_relabund.png&quot;</span>, <span class="at">plot =</span> phylum_bar,</span>
<span id="cb24-7"><a href="08-Iterating_rarefaction.html#cb24-7" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">100</span>, <span class="at">width =</span> <span class="dv">100</span>)</span>
<span id="cb24-8"><a href="08-Iterating_rarefaction.html#cb24-8" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file =</span> <span class="st">&quot;./env_phyla_rarefy_1_relabund.png&quot;</span>)</span></code></pre></div>
<p>Viewing the non-rarefied and <strong>rarefied</strong> based bar charts shows some differences. However, these are quite difficult to discern.</p>
</div>
<div id="r-difference-from-non-rarefied" class="section level3 hasAnchor" number="8.6.3">
<h3><span class="header-section-number">8.6.3</span> 1R: Difference from non-rarefied<a href="08-Iterating_rarefaction.html#r-difference-from-non-rarefied" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/histogram_1.png" style="width:300px" />
</center>
<p>To more easily see the differences we will subtract the two relative abundance tables from each other.
This will produce a matrix of differences.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb25-1"><a href="08-Iterating_rarefaction.html#cb25-1" tabindex="-1"></a><span class="co">#Value difference matrix</span></span>
<span id="cb25-2"><a href="08-Iterating_rarefaction.html#cb25-2" tabindex="-1"></a>single_rarefaction_diff <span class="ot">&lt;-</span> </span>
<span id="cb25-3"><a href="08-Iterating_rarefaction.html#cb25-3" tabindex="-1"></a>  phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_env_phyla_relabund) <span class="sc">-</span> phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_env_phyla_rarefy_1_relabund)</span>
<span id="cb25-4"><a href="08-Iterating_rarefaction.html#cb25-4" tabindex="-1"></a>single_rarefaction_diff</span></code></pre></div>
<p>We can see there are differences.
To make these differences even clearer let's make a histogram.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb26-1"><a href="08-Iterating_rarefaction.html#cb26-1" tabindex="-1"></a><span class="co">#Histogram of differences</span></span>
<span id="cb26-2"><a href="08-Iterating_rarefaction.html#cb26-2" tabindex="-1"></a><span class="fu">hist</span>(single_rarefaction_diff, <span class="at">main =</span> <span class="st">&quot;Single rarefaction&quot;</span>)</span></code></pre></div>
What is the range of the differences compared to the non rarefied relative abundance values?
<div id="radio_CELZPIELEN" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_CELZPIELEN" value=""></input> <span><strong>-0.0003(-3e-04) to 0.0003(3e-04)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_CELZPIELEN" value=""></input> <span><strong>-0.003 to 0.003</strong></span></label><label><input type="radio" autocomplete="off" name="radio_CELZPIELEN" value="answer"></input> <span><strong>-0.015 to 0.015</strong></span></label>
</div>
<p>Although these values appear quite small keep in mind we are working with relative abundance values.
Each sample has a total relative abundance of 1.00 so a relative abundance value of 0.01 is 1%.</p>
<p>Let's see if we can get these differences smaller with multiple rounds of rarefaction.</p>
</div>
</div>
<div id="multiple-rarefaction-mr-iterations" class="section level2 hasAnchor" number="8.7">
<h2><span class="header-section-number">8.7</span> Multiple Rarefaction (MR) iterations<a href="08-Iterating_rarefaction.html#multiple-rarefaction-mr-iterations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/ten_rounds.png" style="width:200px" />
</center>
<p>We will now carry out iterative rarefaction.</p>
<div id="mr-rarefaction" class="section level3 hasAnchor" number="8.7.1">
<h3><span class="header-section-number">8.7.1</span> MR: Rarefaction<a href="08-Iterating_rarefaction.html#mr-rarefaction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/tiger.png" style="width:300px" />
</center>
<p>The below loop creates a relative abundance table created by 10 rounds of iteration.
Type the code and read the annotations to understand what is going on.
Then run the code.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb27-1"><a href="08-Iterating_rarefaction.html#cb27-1" tabindex="-1"></a><span class="co">#Iterative rarefaction to produce an average rarefied relative abundance table</span></span>
<span id="cb27-2"><a href="08-Iterating_rarefaction.html#cb27-2" tabindex="-1"></a></span>
<span id="cb27-3"><a href="08-Iterating_rarefaction.html#cb27-3" tabindex="-1"></a><span class="co">#Assign rarefaction size</span></span>
<span id="cb27-4"><a href="08-Iterating_rarefaction.html#cb27-4" tabindex="-1"></a>rarefaction_size <span class="ot">&lt;-</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_env_phyla))</span>
<span id="cb27-5"><a href="08-Iterating_rarefaction.html#cb27-5" tabindex="-1"></a><span class="co">#Read in our rng seed vector</span></span>
<span id="cb27-6"><a href="08-Iterating_rarefaction.html#cb27-6" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;rngseeds.RData&quot;</span>)</span>
<span id="cb27-7"><a href="08-Iterating_rarefaction.html#cb27-7" tabindex="-1"></a></span>
<span id="cb27-8"><a href="08-Iterating_rarefaction.html#cb27-8" tabindex="-1"></a><span class="co">#Initalise where we will store the output</span></span>
<span id="cb27-9"><a href="08-Iterating_rarefaction.html#cb27-9" tabindex="-1"></a><span class="co">#In this case we create the first iteration</span></span>
<span id="cb27-10"><a href="08-Iterating_rarefaction.html#cb27-10" tabindex="-1"></a><span class="co">#Carry out first rarefaction</span></span>
<span id="cb27-11"><a href="08-Iterating_rarefaction.html#cb27-11" tabindex="-1"></a>pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(pseq_env_phyla,</span>
<span id="cb27-12"><a href="08-Iterating_rarefaction.html#cb27-12" tabindex="-1"></a>                                           <span class="at">sample.size =</span> rarefaction_size</span>
<span id="cb27-13"><a href="08-Iterating_rarefaction.html#cb27-13" tabindex="-1"></a>                                           <span class="co">#First random seed as the rng seed</span></span>
<span id="cb27-14"><a href="08-Iterating_rarefaction.html#cb27-14" tabindex="-1"></a>                                           <span class="at">rngseed =</span> rngseed_vec[<span class="dv">1</span>])</span>
<span id="cb27-15"><a href="08-Iterating_rarefaction.html#cb27-15" tabindex="-1"></a><span class="co">#Calculate relative abundance</span></span>
<span id="cb27-16"><a href="08-Iterating_rarefaction.html#cb27-16" tabindex="-1"></a>pseq_rarefy_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq_rarefy, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb27-17"><a href="08-Iterating_rarefaction.html#cb27-17" tabindex="-1"></a><span class="co">#Relabund phyla table object</span></span>
<span id="cb27-18"><a href="08-Iterating_rarefaction.html#cb27-18" tabindex="-1"></a>relabund_phyla_table <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy_relabund))</span>
<span id="cb27-19"><a href="08-Iterating_rarefaction.html#cb27-19" tabindex="-1"></a></span>
<span id="cb27-20"><a href="08-Iterating_rarefaction.html#cb27-20" tabindex="-1"></a><span class="co">#Loop through the next 9 iterations</span></span>
<span id="cb27-21"><a href="08-Iterating_rarefaction.html#cb27-21" tabindex="-1"></a><span class="co">#Add the relabund rarefied values to phyla_table</span></span>
<span id="cb27-22"><a href="08-Iterating_rarefaction.html#cb27-22" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(rngseed_vec)){</span>
<span id="cb27-23"><a href="08-Iterating_rarefaction.html#cb27-23" tabindex="-1"></a>  <span class="co">#Rarefy</span></span>
<span id="cb27-24"><a href="08-Iterating_rarefaction.html#cb27-24" tabindex="-1"></a>  pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(pseq_env_phyla,</span>
<span id="cb27-25"><a href="08-Iterating_rarefaction.html#cb27-25" tabindex="-1"></a>                                           <span class="at">sample.size =</span> rarefaction_size</span>
<span id="cb27-26"><a href="08-Iterating_rarefaction.html#cb27-26" tabindex="-1"></a>                                           <span class="at">rngseed =</span> rngseed_vec[i])</span>
<span id="cb27-27"><a href="08-Iterating_rarefaction.html#cb27-27" tabindex="-1"></a>  <span class="co">#Calculate relative abundance</span></span>
<span id="cb27-28"><a href="08-Iterating_rarefaction.html#cb27-28" tabindex="-1"></a>  pseq_rarefy_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq_rarefy, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb27-29"><a href="08-Iterating_rarefaction.html#cb27-29" tabindex="-1"></a>  <span class="co">#Sum values to phyla_table</span></span>
<span id="cb27-30"><a href="08-Iterating_rarefaction.html#cb27-30" tabindex="-1"></a>  relabund_phyla_table <span class="ot">&lt;-</span> </span>
<span id="cb27-31"><a href="08-Iterating_rarefaction.html#cb27-31" tabindex="-1"></a>    relabund_phyla_table <span class="sc">+</span> <span class="fu">as.data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy_relabund))</span>
<span id="cb27-32"><a href="08-Iterating_rarefaction.html#cb27-32" tabindex="-1"></a>}</span>
<span id="cb27-33"><a href="08-Iterating_rarefaction.html#cb27-33" tabindex="-1"></a><span class="co">#Average the values of the summed relabund phyla_table</span></span>
<span id="cb27-34"><a href="08-Iterating_rarefaction.html#cb27-34" tabindex="-1"></a>relabund_phyla_table <span class="ot">&lt;-</span> relabund_phyla_table <span class="sc">/</span> <span class="fu">length</span>(rngseed_vec)</span></code></pre></div>
</div>
<div id="mr-difference-from-non-rarefied" class="section level3 hasAnchor" number="8.7.2">
<h3><span class="header-section-number">8.7.2</span> MR: Difference from non-rarefied<a href="08-Iterating_rarefaction.html#mr-difference-from-non-rarefied" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/histogram_3.png" style="width:200px" />
</center>
<p>We'll skip the bar chart this time and only look at the difference of the values.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb28-1"><a href="08-Iterating_rarefaction.html#cb28-1" tabindex="-1"></a><span class="co">#Value difference matrix</span></span>
<span id="cb28-2"><a href="08-Iterating_rarefaction.html#cb28-2" tabindex="-1"></a>iterative_rarefaction_diff <span class="ot">&lt;-</span> </span>
<span id="cb28-3"><a href="08-Iterating_rarefaction.html#cb28-3" tabindex="-1"></a>  <span class="fu">as.matrix</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_env_phyla_relabund) <span class="sc">-</span> relabund_phyla_table)</span>
<span id="cb28-4"><a href="08-Iterating_rarefaction.html#cb28-4" tabindex="-1"></a>iterative_rarefaction_diff</span>
<span id="cb28-5"><a href="08-Iterating_rarefaction.html#cb28-5" tabindex="-1"></a><span class="co">#Histogram</span></span>
<span id="cb28-6"><a href="08-Iterating_rarefaction.html#cb28-6" tabindex="-1"></a><span class="fu">hist</span>(iterative_rarefaction_diff)</span></code></pre></div>
What is the range of the differences compared to the non rarefied relative abundance values?
<div id="radio_LXUTKAKJPC" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_LXUTKAKJPC" value=""></input> <span><strong>-0.0003(-3e-04) to 0.0003(3e-04)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_LXUTKAKJPC" value="answer"></input> <span><strong>-0.003 to 0.003</strong></span></label><label><input type="radio" autocomplete="off" name="radio_LXUTKAKJPC" value=""></input> <span><strong>-0.015 to 0.015</strong></span></label>
</div>
<p>You should notice that the differences are much smaller.
This indicates that the structure of the <strong>iterative rarefied</strong> data is much closer to the non-rarefied data than the one <strong>rarefied</strong> data.
This is what we want.</p>
</div>
</div>
<div id="iterating-rarefaction-task" class="section level2 hasAnchor" number="8.8">
<h2><span class="header-section-number">8.8</span> Iterating rarefaction task<a href="08-Iterating_rarefaction.html#iterating-rarefaction-task" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/per_thousand.png" style="width:200px" />
</center>
<p>Superlative! Now that you know how to carry out <strong>iterative rarefaction</strong> I'll ask you to do it once more for the phyla data.</p>
<p>Create a rarefaction averaged phyla relative abundance as we have done above but with 1000 <strong>rarefaction iterations</strong>.
For this task use <strong>153478</strong> as the <strong>seed</strong> when creating your vector of 1000 <strong>rng seeds</strong>.
Save this vector of rngseeds to a file called "rngseeds_1000.RData".</p>
<p><strong>Note:</strong> The iterative rarefaction step may take a few minutes.</p>
<p>After creating the relative abundance matrix determine how different the values are compared to the non-rarefied relative abundance with a histogram.</p>
What is the range of the differences compared to the non rarefied relative abundance values?
<div id="radio_GRBGLOHMDW" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_GRBGLOHMDW" value="answer"></input> <span><strong>-0.0003(-3e-04) to 0.0003(3e-04)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_GRBGLOHMDW" value=""></input> <span><strong>-0.003 to 0.003</strong></span></label><label><input type="radio" autocomplete="off" name="radio_GRBGLOHMDW" value=""></input> <span><strong>-0.015 to 0.015</strong></span></label>
</div>
<p>Please attempt the task yourself before looking at the solution code in the below expandable box.</p>
<div class="webex-solution">
<button>
Task solution code
</button>
<div class="sourceCode" id="cb29"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb29-1"><a href="08-Iterating_rarefaction.html#cb29-1" tabindex="-1"></a><span class="co">#Number of rarefaction iterations to be carried out</span></span>
<span id="cb29-2"><a href="08-Iterating_rarefaction.html#cb29-2" tabindex="-1"></a>rarefaction_iters <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb29-3"><a href="08-Iterating_rarefaction.html#cb29-3" tabindex="-1"></a><span class="co">#Create rngseed vector</span></span>
<span id="cb29-4"><a href="08-Iterating_rarefaction.html#cb29-4" tabindex="-1"></a><span class="co">#Set seed for reproducibility</span></span>
<span id="cb29-5"><a href="08-Iterating_rarefaction.html#cb29-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">153478</span>)</span>
<span id="cb29-6"><a href="08-Iterating_rarefaction.html#cb29-6" tabindex="-1"></a><span class="co">#Create the rngseed vector</span></span>
<span id="cb29-7"><a href="08-Iterating_rarefaction.html#cb29-7" tabindex="-1"></a><span class="co">#Sample 1000 (number of iters) values from the number range 1-100,000</span></span>
<span id="cb29-8"><a href="08-Iterating_rarefaction.html#cb29-8" tabindex="-1"></a>rngseed_vec <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">100000</span>, <span class="at">size =</span> rarefaction_iters, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb29-9"><a href="08-Iterating_rarefaction.html#cb29-9" tabindex="-1"></a><span class="co">#Save our rngseed vector</span></span>
<span id="cb29-10"><a href="08-Iterating_rarefaction.html#cb29-10" tabindex="-1"></a><span class="fu">save</span>(rngseed_vec, <span class="at">file=</span><span class="st">&quot;rngseeds_1000.RData&quot;</span>)</span>
<span id="cb29-11"><a href="08-Iterating_rarefaction.html#cb29-11" tabindex="-1"></a><span class="co">#Reset seed</span></span>
<span id="cb29-12"><a href="08-Iterating_rarefaction.html#cb29-12" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="cn">NULL</span>)</span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb30-1"><a href="08-Iterating_rarefaction.html#cb30-1" tabindex="-1"></a><span class="co">#Iterative rarefaction to produce an average rarefied relative abundance table</span></span>
<span id="cb30-2"><a href="08-Iterating_rarefaction.html#cb30-2" tabindex="-1"></a></span>
<span id="cb30-3"><a href="08-Iterating_rarefaction.html#cb30-3" tabindex="-1"></a><span class="co">#Assign rarefaction size</span></span>
<span id="cb30-4"><a href="08-Iterating_rarefaction.html#cb30-4" tabindex="-1"></a>rarefaction_size <span class="ot">&lt;-</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_env_phyla))</span>
<span id="cb30-5"><a href="08-Iterating_rarefaction.html#cb30-5" tabindex="-1"></a><span class="co">#Read in our rng seed vector</span></span>
<span id="cb30-6"><a href="08-Iterating_rarefaction.html#cb30-6" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;rngseeds_1000.RData&quot;</span>)</span>
<span id="cb30-7"><a href="08-Iterating_rarefaction.html#cb30-7" tabindex="-1"></a></span>
<span id="cb30-8"><a href="08-Iterating_rarefaction.html#cb30-8" tabindex="-1"></a><span class="co">#Initalise where we will store the output</span></span>
<span id="cb30-9"><a href="08-Iterating_rarefaction.html#cb30-9" tabindex="-1"></a><span class="co">#In this case we create the first iteration</span></span>
<span id="cb30-10"><a href="08-Iterating_rarefaction.html#cb30-10" tabindex="-1"></a><span class="co">#Carry out first rarefaction</span></span>
<span id="cb30-11"><a href="08-Iterating_rarefaction.html#cb30-11" tabindex="-1"></a>pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(pseq_env_phyla,</span>
<span id="cb30-12"><a href="08-Iterating_rarefaction.html#cb30-12" tabindex="-1"></a>                                           <span class="at">sample.size =</span> rarefaction_size</span>
<span id="cb30-13"><a href="08-Iterating_rarefaction.html#cb30-13" tabindex="-1"></a>                                           <span class="at">rngseed =</span> rngseed_vec[<span class="dv">1</span>])</span>
<span id="cb30-14"><a href="08-Iterating_rarefaction.html#cb30-14" tabindex="-1"></a><span class="co">#Calculate relative abundance</span></span>
<span id="cb30-15"><a href="08-Iterating_rarefaction.html#cb30-15" tabindex="-1"></a>pseq_rarefy_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq_rarefy, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb30-16"><a href="08-Iterating_rarefaction.html#cb30-16" tabindex="-1"></a><span class="co">#Relabund phyla table object</span></span>
<span id="cb30-17"><a href="08-Iterating_rarefaction.html#cb30-17" tabindex="-1"></a>relabund_phyla_table <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy_relabund))</span>
<span id="cb30-18"><a href="08-Iterating_rarefaction.html#cb30-18" tabindex="-1"></a></span>
<span id="cb30-19"><a href="08-Iterating_rarefaction.html#cb30-19" tabindex="-1"></a><span class="co">#Loop through the next 999 iterations</span></span>
<span id="cb30-20"><a href="08-Iterating_rarefaction.html#cb30-20" tabindex="-1"></a><span class="co">#Add the relabund rarefied values to phyla_table</span></span>
<span id="cb30-21"><a href="08-Iterating_rarefaction.html#cb30-21" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(rngseed_vec)){</span>
<span id="cb30-22"><a href="08-Iterating_rarefaction.html#cb30-22" tabindex="-1"></a>  <span class="co">#Rarefy</span></span>
<span id="cb30-23"><a href="08-Iterating_rarefaction.html#cb30-23" tabindex="-1"></a>  pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(pseq_env_phyla,</span>
<span id="cb30-24"><a href="08-Iterating_rarefaction.html#cb30-24" tabindex="-1"></a>                                           <span class="at">sample.size =</span> rarefaction_size</span>
<span id="cb30-25"><a href="08-Iterating_rarefaction.html#cb30-25" tabindex="-1"></a>                                           <span class="at">rngseed =</span> rngseed_vec[i])</span>
<span id="cb30-26"><a href="08-Iterating_rarefaction.html#cb30-26" tabindex="-1"></a>  <span class="co">#Calculate relative abundance</span></span>
<span id="cb30-27"><a href="08-Iterating_rarefaction.html#cb30-27" tabindex="-1"></a>  pseq_rarefy_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq_rarefy, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb30-28"><a href="08-Iterating_rarefaction.html#cb30-28" tabindex="-1"></a>  <span class="co">#Sum values to phyla_table</span></span>
<span id="cb30-29"><a href="08-Iterating_rarefaction.html#cb30-29" tabindex="-1"></a>  relabund_phyla_table <span class="ot">&lt;-</span> </span>
<span id="cb30-30"><a href="08-Iterating_rarefaction.html#cb30-30" tabindex="-1"></a>    relabund_phyla_table <span class="sc">+</span> <span class="fu">as.data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy_relabund))</span>
<span id="cb30-31"><a href="08-Iterating_rarefaction.html#cb30-31" tabindex="-1"></a>}</span>
<span id="cb30-32"><a href="08-Iterating_rarefaction.html#cb30-32" tabindex="-1"></a><span class="co">#Average the values of the summed relabund phyla_table</span></span>
<span id="cb30-33"><a href="08-Iterating_rarefaction.html#cb30-33" tabindex="-1"></a>relabund_phyla_table <span class="ot">&lt;-</span> relabund_phyla_table <span class="sc">/</span> <span class="fu">length</span>(rngseed_vec)</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb31-1"><a href="08-Iterating_rarefaction.html#cb31-1" tabindex="-1"></a><span class="co">#Value difference matrix</span></span>
<span id="cb31-2"><a href="08-Iterating_rarefaction.html#cb31-2" tabindex="-1"></a>iterative_rarefaction_1000_diff <span class="ot">&lt;-</span> </span>
<span id="cb31-3"><a href="08-Iterating_rarefaction.html#cb31-3" tabindex="-1"></a>  <span class="fu">as.matrix</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_env_phyla_relabund) <span class="sc">-</span> relabund_phyla_table)</span>
<span id="cb31-4"><a href="08-Iterating_rarefaction.html#cb31-4" tabindex="-1"></a>iterative_rarefaction_diff</span>
<span id="cb31-5"><a href="08-Iterating_rarefaction.html#cb31-5" tabindex="-1"></a><span class="co">#Histogram</span></span>
<span id="cb31-6"><a href="08-Iterating_rarefaction.html#cb31-6" tabindex="-1"></a><span class="fu">hist</span>(iterative_rarefaction_diff)</span></code></pre></div>
</div>
</div>
<div id="iterating-rarefaction-recap" class="section level2 hasAnchor" number="8.9">
<h2><span class="header-section-number">8.9</span> Iterating rarefaction recap<a href="08-Iterating_rarefaction.html#iterating-rarefaction-recap" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/recap.png" style="width:200px" />
</center>
<p>With this chapter you have learnt:</p>
<ul>
<li>How to create a vector of rngseeds</li>
<li>How to use rng seeds to carry out <strong>iterative rarefaction</strong></li>
</ul>
<p>In you real life analysis you would not use this method to create relative abundance taxonomy bar charts, you would use the non-rarefied relative abundance.
However, this hopefully gave you a good idea of how <strong>iterative rarefaction</strong> works so we can utilise it the next 2 chapters for alpha and beta diversity analysis.</p>
<p>Feel free to save then close and halt your "Iterating_rarefaction.ipynb" notebook.</p>

</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="07-Rarefaction_and_random_seeds.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="09-Alpha_diversity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
