<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Beta Diversity | R community advanced analysis</title>
  <meta name="description" content="NEOF book for the R community advanced analysis course" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Beta Diversity | R community advanced analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the R community advanced analysis course" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Beta Diversity | R community advanced analysis" />
  
  <meta name="twitter:description" content="NEOF book for the R community advanced analysis course" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew R. Gemmell" />


<meta name="date" content="2024-07-10" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="09-Alpha_diversity.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/NEOF_bordered_small.png"></a></li>
<li><a>R community analysis</a></li>

<li class="divider"></li>
<li class="part"><span><b>Intro</b></span></li>
<li class="chapter" data-level="1" data-path="01-R_community_advanced_main_book.html"><a href="01-R_community_advanced_main_book.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="01-R_community_advanced_main_book.html"><a href="01-R_community_advanced_main_book.html#table-of-contents"><i class="fa fa-check"></i>Table of contents</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html"><i class="fa fa-check"></i><b>2</b> Dataset &amp; workflow</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#dataset"><i class="fa fa-check"></i><b>2.1</b> Dataset</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sites"><i class="fa fa-check"></i><b>2.1.1</b> Sites</a></li>
<li class="chapter" data-level="2.1.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#culture-media"><i class="fa fa-check"></i><b>2.1.2</b> Culture media</a></li>
<li class="chapter" data-level="2.1.3" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sum_and_qs"><i class="fa fa-check"></i><b>2.1.3</b> Summary &amp; questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#workflow"><i class="fa fa-check"></i><b>2.2</b> Workflow</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-R_packages.html"><a href="03-R_packages.html"><i class="fa fa-check"></i><b>3</b> R Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-R_packages.html"><a href="03-R_packages.html#r-packageslibraries"><i class="fa fa-check"></i><b>3.1</b> R packages/libraries</a></li>
<li class="chapter" data-level="3.2" data-path="03-R_packages.html"><a href="03-R_packages.html#the-grammar-of-graphics"><i class="fa fa-check"></i><b>3.2</b> The grammar of graphics</a></li>
<li class="chapter" data-level="3.3" data-path="03-R_packages.html"><a href="03-R_packages.html#phyloseq"><i class="fa fa-check"></i><b>3.3</b> phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Setup.html"><a href="04-Setup.html"><i class="fa fa-check"></i><b>4</b> Set-up</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Setup.html"><a href="04-Setup.html#cluster"><i class="fa fa-check"></i><b>4.1</b> Logon instructions</a></li>
<li class="chapter" data-level="4.2" data-path="04-Setup.html"><a href="04-Setup.html#mamba"><i class="fa fa-check"></i><b>4.2</b> Mamba</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Jupyter.html"><a href="05-Jupyter.html"><i class="fa fa-check"></i><b>5</b> Jupyter</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Jupyter.html"><a href="05-Jupyter.html#open-jupyter-notebook"><i class="fa fa-check"></i><b>5.1</b> Open Jupyter-notebook</a></li>
<li class="chapter" data-level="5.2" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-r-notebook"><i class="fa fa-check"></i><b>5.2</b> Create R notebook</a></li>
<li class="chapter" data-level="5.3" data-path="05-Jupyter.html"><a href="05-Jupyter.html#cells-and-code"><i class="fa fa-check"></i><b>5.3</b> Cells and code</a></li>
<li class="chapter" data-level="5.4" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-new-cells"><i class="fa fa-check"></i><b>5.4</b> Create new cells</a></li>
<li class="chapter" data-level="5.5" data-path="05-Jupyter.html"><a href="05-Jupyter.html#running-code"><i class="fa fa-check"></i><b>5.5</b> Running code</a></li>
<li class="chapter" data-level="5.6" data-path="05-Jupyter.html"><a href="05-Jupyter.html#saving-the-file"><i class="fa fa-check"></i><b>5.6</b> Saving the file</a></li>
<li class="chapter" data-level="5.7" data-path="05-Jupyter.html"><a href="05-Jupyter.html#title-cells-with-markdown"><i class="fa fa-check"></i><b>5.7</b> Title cells with markdown</a></li>
<li class="chapter" data-level="5.8" data-path="05-Jupyter.html"><a href="05-Jupyter.html#close-the-notebook"><i class="fa fa-check"></i><b>5.8</b> Close the notebook</a></li>
<li class="chapter" data-level="5.9" data-path="05-Jupyter.html"><a href="05-Jupyter.html#jup_vid_tut"><i class="fa fa-check"></i><b>5.9</b> Video tutorial</a></li>
</ul></li>
<li class="part"><span><b>Iterative rarefaction</b></span></li>
<li class="chapter" data-level="6" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html"><i class="fa fa-check"></i><b>6</b> Iterative rarefaction intro</a>
<ul>
<li class="chapter" data-level="6.1" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html#should-you-rarefy"><i class="fa fa-check"></i><b>6.1</b> Should you rarefy?</a></li>
<li class="chapter" data-level="6.2" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html#using-iterations"><i class="fa fa-check"></i><b>6.2</b> Using iterations</a></li>
<li class="chapter" data-level="6.3" data-path="06-Iterative_rarefaction.html"><a href="06-Iterative_rarefaction.html#section-contents"><i class="fa fa-check"></i><b>6.3</b> Section contents</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html"><i class="fa fa-check"></i><b>7</b> Random seeds</a>
<ul>
<li class="chapter" data-level="7.1" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-notebook"><i class="fa fa-check"></i><b>7.1</b> Random seed notebook</a></li>
<li class="chapter" data-level="7.2" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-sampling"><i class="fa fa-check"></i><b>7.2</b> Random sampling</a></li>
<li class="chapter" data-level="7.3" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#sampling-with-replacement"><i class="fa fa-check"></i><b>7.3</b> Sampling with replacement</a></li>
<li class="chapter" data-level="7.4" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#setting-a-random-seed"><i class="fa fa-check"></i><b>7.4</b> Setting a random seed</a></li>
<li class="chapter" data-level="7.5" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#reset-seed"><i class="fa fa-check"></i><b>7.5</b> Reset seed</a></li>
<li class="chapter" data-level="7.6" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-practice"><i class="fa fa-check"></i><b>7.6</b> Random seed practice</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-question-1"><i class="fa fa-check"></i><b>7.6.1</b> Random seed: Question 1</a></li>
<li class="chapter" data-level="7.6.2" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-question-2"><i class="fa fa-check"></i><b>7.6.2</b> Random seed: Question 2</a></li>
<li class="chapter" data-level="7.6.3" data-path="07-Rarefaction_and_random_seeds.html"><a href="07-Rarefaction_and_random_seeds.html#random-seed-question-3"><i class="fa fa-check"></i><b>7.6.3</b> Random seed: Question 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html"><i class="fa fa-check"></i><b>8</b> Iterating rarefaction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-dataset"><i class="fa fa-check"></i><b>8.1</b> Iterating rarefaction dataset</a></li>
<li class="chapter" data-level="8.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-setup"><i class="fa fa-check"></i><b>8.2</b> Iterating rarefaction setup</a></li>
<li class="chapter" data-level="8.3" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#rarefaction-iterations"><i class="fa fa-check"></i><b>8.3</b> Rarefaction iterations</a></li>
<li class="chapter" data-level="8.4" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#rng_vec_creation"><i class="fa fa-check"></i><b>8.4</b> RNG vector creation</a></li>
<li class="chapter" data-level="8.5" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#phyla-relative-abundance"><i class="fa fa-check"></i><b>8.5</b> Phyla relative abundance</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#subset-and-phyla-aggregation"><i class="fa fa-check"></i><b>8.5.1</b> Subset and phyla aggregation</a></li>
<li class="chapter" data-level="8.5.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#phyla-relative-abundance-bar-chart"><i class="fa fa-check"></i><b>8.5.2</b> Phyla relative abundance bar chart</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#one-round-1r-of-rarefaction"><i class="fa fa-check"></i><b>8.6</b> One Round (1R) of rarefaction</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#r-rarefaction"><i class="fa fa-check"></i><b>8.6.1</b> 1R: Rarefaction</a></li>
<li class="chapter" data-level="8.6.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#r-relative-abundance-bar-chart"><i class="fa fa-check"></i><b>8.6.2</b> 1R: Relative abundance bar chart</a></li>
<li class="chapter" data-level="8.6.3" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#r-difference-from-non-rarefied"><i class="fa fa-check"></i><b>8.6.3</b> 1R: Difference from non-rarefied</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#multiple-rarefaction-mr-iterations"><i class="fa fa-check"></i><b>8.7</b> Multiple Rarefaction (MR) iterations</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#mr-rarefaction"><i class="fa fa-check"></i><b>8.7.1</b> MR: Rarefaction</a></li>
<li class="chapter" data-level="8.7.2" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#mr-difference-from-non-rarefied"><i class="fa fa-check"></i><b>8.7.2</b> MR: Difference from non-rarefied</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-task"><i class="fa fa-check"></i><b>8.8</b> Iterating rarefaction task</a></li>
<li class="chapter" data-level="8.9" data-path="08-Iterating_rarefaction.html"><a href="08-Iterating_rarefaction.html#iterating-rarefaction-recap"><i class="fa fa-check"></i><b>8.9</b> Iterating rarefaction recap</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html"><i class="fa fa-check"></i><b>9</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="9.1" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-setup"><i class="fa fa-check"></i><b>9.1</b> <span class="math inline">\(\alpha\)</span>: Setup</a></li>
<li class="chapter" data-level="9.2" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-iterative-rarefaction-values"><i class="fa fa-check"></i><b>9.2</b> <span class="math inline">\(\alpha\)</span>: Iterative rarefaction values</a></li>
<li class="chapter" data-level="9.3" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-iterative-rarefaction-loop"><i class="fa fa-check"></i><b>9.3</b> <span class="math inline">\(\alpha\)</span>: Iterative rarefaction loop</a></li>
<li class="chapter" data-level="9.4" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-metric-and-metadata-data-frame"><i class="fa fa-check"></i><b>9.4</b> <span class="math inline">\(\alpha\)</span>: Metric and metadata data frame</a></li>
<li class="chapter" data-level="9.5" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha_long_df"><i class="fa fa-check"></i><b>9.5</b> <span class="math inline">\(\alpha\)</span>: Long data frame</a></li>
<li class="chapter" data-level="9.6" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-subset-metrics"><i class="fa fa-check"></i><b>9.6</b> <span class="math inline">\(\alpha\)</span>: Subset metrics</a></li>
<li class="chapter" data-level="9.7" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-violin-plot"><i class="fa fa-check"></i><b>9.7</b> <span class="math inline">\(\alpha\)</span>: Violin plot</a></li>
<li class="chapter" data-level="9.8" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-stats"><i class="fa fa-check"></i><b>9.8</b> <span class="math inline">\(\alpha\)</span>: Stats</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#kruskal-wallis-test"><i class="fa fa-check"></i><b>9.8.1</b> Kruskal Wallis test</a></li>
<li class="chapter" data-level="9.8.2" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#paired-wilcoxon-test"><i class="fa fa-check"></i><b>9.8.2</b> Paired wilcoxon test</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-plot-with-stats"><i class="fa fa-check"></i><b>9.9</b> <span class="math inline">\(\alpha\)</span>: Plot with stats</a>
<ul>
<li class="chapter" data-level="9.9.1" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#list-of-comparisons"><i class="fa fa-check"></i><b>9.9.1</b> List of comparisons</a></li>
<li class="chapter" data-level="9.9.2" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#violin-plot-with-stats"><i class="fa fa-check"></i><b>9.9.2</b> Violin plot with stats</a></li>
<li class="chapter" data-level="9.9.3" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#reorder-x-axis-and-stats"><i class="fa fa-check"></i><b>9.9.3</b> Reorder x-axis and stats</a></li>
</ul></li>
<li class="chapter" data-level="9.10" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-task"><i class="fa fa-check"></i><b>9.10</b> <span class="math inline">\(\alpha\)</span>: Task</a></li>
<li class="chapter" data-level="9.11" data-path="09-Alpha_diversity.html"><a href="09-Alpha_diversity.html#alpha-recap"><i class="fa fa-check"></i><b>9.11</b> <span class="math inline">\(\alpha\)</span>: Recap</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html"><i class="fa fa-check"></i><b>10</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="10.1" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#beta-setup"><i class="fa fa-check"></i><b>10.1</b> <span class="math inline">\(\beta\)</span>: Setup</a></li>
<li class="chapter" data-level="10.2" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#beta-distance-matrices"><i class="fa fa-check"></i><b>10.2</b> <span class="math inline">\(\beta\)</span>: Distance matrices</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#unifrac-distances"><i class="fa fa-check"></i><b>10.2.1</b> Unifrac distances</a></li>
<li class="chapter" data-level="10.2.2" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#vegan-distances"><i class="fa fa-check"></i><b>10.2.2</b> Vegan distances</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#beta-iterative-rarefaction"><i class="fa fa-check"></i><b>10.3</b> <span class="math inline">\(\beta\)</span>: Iterative rarefaction</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#iterative-rarefaction-values"><i class="fa fa-check"></i><b>10.3.1</b> Iterative rarefaction values</a></li>
<li class="chapter" data-level="10.3.2" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#iterative-beta-diversity-calculation"><i class="fa fa-check"></i><b>10.3.2</b> Iterative beta diversity calculation</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#beta-ordination"><i class="fa fa-check"></i><b>10.4</b> <span class="math inline">\(\beta\)</span>: Ordination</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#load-weighted-unifrac-matrix"><i class="fa fa-check"></i><b>10.4.1</b> Load weighted unifrac matrix</a></li>
<li class="chapter" data-level="10.4.2" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#nmds"><i class="fa fa-check"></i><b>10.4.2</b> NMDS</a></li>
<li class="chapter" data-level="10.4.3" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#pcoa"><i class="fa fa-check"></i><b>10.4.3</b> PCoA</a></li>
<li class="chapter" data-level="10.4.4" data-path="10-Beta_diversity.html"><a href="10-Beta_diversity.html#which-ordination-method-to-use"><i class="fa fa-check"></i><b>10.4.4</b> Which ordination method to use?</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R community advanced analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="beta_chap" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Beta Diversity<a href="10-Beta_diversity.html#beta_chap" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<center>
<img src="figures/beta_iterative_rarefaction.png" style="width:300px" />
</center>
<p>This chapter will teach you how to carry out <strong>beta diversity</strong> analysis.
The steps will include:</p>
<ul>
<li>Calculating <strong>beta diversity</strong> distances</li>
<li>Using <strong>iterative rarefaction</strong> to create averaged distance values</li>
<li>Carrying various ordination techniques to visualise the distances between samples</li>
</ul>
<div id="beta-setup" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> <span class="math inline">\(\beta\)</span>: Setup<a href="10-Beta_diversity.html#beta-setup" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/setup_7.png" style="width:200px" />
</center>
<p>Create a new R jupyter notebook called "Beta_diversity.ipynb".</p>
<p>Load the required data and libraries.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb51-1"><a href="10-Beta_diversity.html#cb51-1" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb51-2"><a href="10-Beta_diversity.html#cb51-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb51-3"><a href="10-Beta_diversity.html#cb51-3" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb51-4"><a href="10-Beta_diversity.html#cb51-4" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb51-5"><a href="10-Beta_diversity.html#cb51-5" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;vegan&quot;</span>)</span>
<span id="cb51-6"><a href="10-Beta_diversity.html#cb51-6" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;rbiom&quot;</span>)</span>
<span id="cb51-7"><a href="10-Beta_diversity.html#cb51-7" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ape&quot;</span>)</span>
<span id="cb51-8"><a href="10-Beta_diversity.html#cb51-8" tabindex="-1"></a><span class="co">#Load processed but unrarefied ASV data from main R community workshop</span></span>
<span id="cb51-9"><a href="10-Beta_diversity.html#cb51-9" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
</div>
<div id="beta-distance-matrices" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> <span class="math inline">\(\beta\)</span>: Distance matrices<a href="10-Beta_diversity.html#beta-distance-matrices" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/measuring_tape_2.png" style="width:300px" />
</center>
<p>Here we'll find out how to produce a paired distance matrix with our <strong>beta diversity</strong> metric of choice.
Unfortunately there is not one function that can calculate all the metrics we may want.</p>
<div id="unifrac-distances" class="section level3 hasAnchor" number="10.2.1">
<h3><span class="header-section-number">10.2.1</span> Unifrac distances<a href="10-Beta_diversity.html#unifrac-distances" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/weighted_blanket.png" style="width:200px" />
</center>
<p>Unifrac distances are preferred by many when a phylogenetic tree is available for your data.
We are working with 16S data so a phylogenetic tree was created.
Other barcodes or types of data may not have a phylogenetic such as ITS data or complex shotgun metagenomic data.</p>
<p>To produce a Unifrac distance matrix we will use the function <code>unifrac()</code> from the package <code>rbiom</code>.</p>
<p>The function <code>rbiom::unifrac()</code> was created to work with biom files and not <strong>phyloseq</strong> objects.
Thankfully we don't need to convert our objects to a biom object, instead only needing to extract our count/abundance data and phylogenetic tree.
We can carry this out with the following 2 functions:</p>
<ul>
<li><code>phyloseq::otu_table()</code>: Extracts the feature count/abundances table</li>
<li><code>phyloseq::phy_tree()</code>: Extract the phylogenetic tree</li>
</ul>
<p>There are 2 types of Unifrac distances:</p>
<ul>
<li>Weighted: Incorporate relative abundances</li>
<li>Unweighted: Does not incorporate relative abundances</li>
</ul>
<p>We'll calculate weighted Unifrac distances by specifying the parameter <code>weighted = TRUE</code>.</p>
<p>So our output works with subsequent ordination step we'll convert the output of <code>rbiom::unifrac()</code> into a matrix with <code>as.matrix()</code>.</p>
<p>Carry out the matrix production with the below script:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb52-1"><a href="10-Beta_diversity.html#cb52-1" tabindex="-1"></a><span class="co">#Calculate weighted unifrac values</span></span>
<span id="cb52-2"><a href="10-Beta_diversity.html#cb52-2" tabindex="-1"></a>unifrac_rbiom_microbiome <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(</span>
<span id="cb52-3"><a href="10-Beta_diversity.html#cb52-3" tabindex="-1"></a>  rbiom<span class="sc">::</span><span class="fu">unifrac</span>(<span class="at">biom =</span> phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq),</span>
<span id="cb52-4"><a href="10-Beta_diversity.html#cb52-4" tabindex="-1"></a>                 <span class="at">tree =</span> phyloseq<span class="sc">::</span><span class="fu">phy_tree</span>(pseq),</span>
<span id="cb52-5"><a href="10-Beta_diversity.html#cb52-5" tabindex="-1"></a>                 <span class="at">weighted =</span> <span class="cn">TRUE</span>))</span>
<span id="cb52-6"><a href="10-Beta_diversity.html#cb52-6" tabindex="-1"></a><span class="co">#Check the first 6 rows and columns of the resulting distance matrix</span></span>
<span id="cb52-7"><a href="10-Beta_diversity.html#cb52-7" tabindex="-1"></a><span class="fu">head</span>(unifrac_rbiom_microbiome)</span></code></pre></div>
</div>
<div id="vegan-distances" class="section level3 hasAnchor" number="10.2.2">
<h3><span class="header-section-number">10.2.2</span> Vegan distances<a href="10-Beta_diversity.html#vegan-distances" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/vegan_fork.png" style="width:200px" />
</center>
<p>To calculate non-unifrac <strong>beta diversity</strong> distances we can use the <code>vegan</code> package and its function <code>vegdist()</code>.</p>
<p>The function <code>rbiom::unifrac()</code> calculates paired distances, this is the samples in the <code>otu_table</code> of a phyloseq object.
Unfortunately, <code>vegan::vegdist()</code> calculates paired distances by rows (features), this being the features in the <code>otu_table</code> of a phyloseq object.
We do not want this but thankfully all we need to add is the function <code>t()</code> to our abundance object to transpose the data.</p>
<p>Create a <strong>Bray-Curtis</strong> distance matrix with the below code:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb53-1"><a href="10-Beta_diversity.html#cb53-1" tabindex="-1"></a><span class="co">#Calculate Bray-Curtis distance matrix</span></span>
<span id="cb53-2"><a href="10-Beta_diversity.html#cb53-2" tabindex="-1"></a>bray_curtis_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(</span>
<span id="cb53-3"><a href="10-Beta_diversity.html#cb53-3" tabindex="-1"></a>  vegan<span class="sc">::</span><span class="fu">vegdist</span>(<span class="at">x =</span> <span class="fu">t</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq)),</span>
<span id="cb53-4"><a href="10-Beta_diversity.html#cb53-4" tabindex="-1"></a>                 <span class="at">method =</span> <span class="st">&quot;bray&quot;</span>))</span>
<span id="cb53-5"><a href="10-Beta_diversity.html#cb53-5" tabindex="-1"></a><span class="co">#Check first 6 rows and columns of matrix</span></span>
<span id="cb53-6"><a href="10-Beta_diversity.html#cb53-6" tabindex="-1"></a>bray_curtis_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code></pre></div>
<p>The function can calculate a plethora of metrics with the full list available at the following <a href="https://rdrr.io/cran/vegan/man/vegdist.html">webpage</a>.</p>
</div>
</div>
<div id="beta-iterative-rarefaction" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> <span class="math inline">\(\beta\)</span>: Iterative rarefaction<a href="10-Beta_diversity.html#beta-iterative-rarefaction" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/iterative_distances.png" style="width:400px" />
</center>
<p>Now that we know how to create a <strong>beta diveristy</strong> paired distance matrix, we can create one with averaged values created by <strong>iterative rarefaction</strong>.</p>
<div id="iterative-rarefaction-values" class="section level3 hasAnchor" number="10.3.1">
<h3><span class="header-section-number">10.3.1</span> Iterative rarefaction values<a href="10-Beta_diversity.html#iterative-rarefaction-values" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/ten_rng_seeds.png" style="width:200px" />
</center>
<p>We need to set our <strong>rarefaction</strong> values and rngseeds.
We can use the same code as we used in the <strong>alpha_diversity</strong> analysis.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb54-1"><a href="10-Beta_diversity.html#cb54-1" tabindex="-1"></a><span class="co">#Rarefaction values</span></span>
<span id="cb54-2"><a href="10-Beta_diversity.html#cb54-2" tabindex="-1"></a><span class="co">#Rarefaction size</span></span>
<span id="cb54-3"><a href="10-Beta_diversity.html#cb54-3" tabindex="-1"></a><span class="co">#Minimum sample depth in this case</span></span>
<span id="cb54-4"><a href="10-Beta_diversity.html#cb54-4" tabindex="-1"></a>rarefaction_size <span class="ot">&lt;-</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq))</span>
<span id="cb54-5"><a href="10-Beta_diversity.html#cb54-5" tabindex="-1"></a><span class="co">#Load the vector of 10 rngseeds created in the previous chapter</span></span>
<span id="cb54-6"><a href="10-Beta_diversity.html#cb54-6" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;rngseeds.RData&quot;</span>)</span>
<span id="cb54-7"><a href="10-Beta_diversity.html#cb54-7" tabindex="-1"></a><span class="co">#Number of rarefaction iterations to be carried out</span></span>
<span id="cb54-8"><a href="10-Beta_diversity.html#cb54-8" tabindex="-1"></a><span class="co">#Based on length of rng seed vector</span></span>
<span id="cb54-9"><a href="10-Beta_diversity.html#cb54-9" tabindex="-1"></a>rarefaction_iters <span class="ot">&lt;-</span> <span class="fu">length</span>(rngseed_vec)</span></code></pre></div>
</div>
<div id="iterative-beta-diversity-calculation" class="section level3 hasAnchor" number="10.3.2">
<h3><span class="header-section-number">10.3.2</span> Iterative beta diversity calculation<a href="10-Beta_diversity.html#iterative-beta-diversity-calculation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/mean.png" style="width:200px" />
</center>
<p>The below code carries out <strong>itertaive rarefaction</strong> and produces an averaged <strong>weighted unifrac</strong> paired distance matrix.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb55-1"><a href="10-Beta_diversity.html#cb55-1" tabindex="-1"></a><span class="co">#Loop to create iteration based rarefied weighted unifrac values</span></span>
<span id="cb55-2"><a href="10-Beta_diversity.html#cb55-2" tabindex="-1"></a></span>
<span id="cb55-3"><a href="10-Beta_diversity.html#cb55-3" tabindex="-1"></a><span class="co">#Create matrix to contain summed wunifrac beta diversity values</span></span>
<span id="cb55-4"><a href="10-Beta_diversity.html#cb55-4" tabindex="-1"></a><span class="co">#In this case we&#39;ll run the first rarefied beta diversity analysis</span></span>
<span id="cb55-5"><a href="10-Beta_diversity.html#cb55-5" tabindex="-1"></a>pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(</span>
<span id="cb55-6"><a href="10-Beta_diversity.html#cb55-6" tabindex="-1"></a>  pseq,</span>
<span id="cb55-7"><a href="10-Beta_diversity.html#cb55-7" tabindex="-1"></a>  <span class="at">sample.size =</span> rarefaction_size,</span>
<span id="cb55-8"><a href="10-Beta_diversity.html#cb55-8" tabindex="-1"></a>  <span class="at">rngseed =</span> rngseed_vec[<span class="dv">1</span>],</span>
<span id="cb55-9"><a href="10-Beta_diversity.html#cb55-9" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb55-10"><a href="10-Beta_diversity.html#cb55-10" tabindex="-1"></a><span class="co">#wunifrac beta diversity</span></span>
<span id="cb55-11"><a href="10-Beta_diversity.html#cb55-11" tabindex="-1"></a>beta_df_sum <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(</span>
<span id="cb55-12"><a href="10-Beta_diversity.html#cb55-12" tabindex="-1"></a>  rbiom<span class="sc">::</span><span class="fu">unifrac</span>(</span>
<span id="cb55-13"><a href="10-Beta_diversity.html#cb55-13" tabindex="-1"></a>    <span class="at">biom =</span> phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy),</span>
<span id="cb55-14"><a href="10-Beta_diversity.html#cb55-14" tabindex="-1"></a>    <span class="at">tree =</span> phyloseq<span class="sc">::</span><span class="fu">phy_tree</span>(pseq_rarefy),</span>
<span id="cb55-15"><a href="10-Beta_diversity.html#cb55-15" tabindex="-1"></a>    <span class="at">weighted =</span> <span class="cn">TRUE</span>))</span>
<span id="cb55-16"><a href="10-Beta_diversity.html#cb55-16" tabindex="-1"></a></span>
<span id="cb55-17"><a href="10-Beta_diversity.html#cb55-17" tabindex="-1"></a><span class="co">#Loop through 2 to the number of iterations</span></span>
<span id="cb55-18"><a href="10-Beta_diversity.html#cb55-18" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>rarefaction_iters){</span>
<span id="cb55-19"><a href="10-Beta_diversity.html#cb55-19" tabindex="-1"></a>  <span class="co">#Rarefaction</span></span>
<span id="cb55-20"><a href="10-Beta_diversity.html#cb55-20" tabindex="-1"></a>  pseq_rarefy <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(</span>
<span id="cb55-21"><a href="10-Beta_diversity.html#cb55-21" tabindex="-1"></a>    pseq,</span>
<span id="cb55-22"><a href="10-Beta_diversity.html#cb55-22" tabindex="-1"></a>    <span class="at">sample.size =</span> rarefaction_size,</span>
<span id="cb55-23"><a href="10-Beta_diversity.html#cb55-23" tabindex="-1"></a>    <span class="at">rngseed =</span> rngseed_vec[i],</span>
<span id="cb55-24"><a href="10-Beta_diversity.html#cb55-24" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb55-25"><a href="10-Beta_diversity.html#cb55-25" tabindex="-1"></a>  <span class="co">#Beta diversity</span></span>
<span id="cb55-26"><a href="10-Beta_diversity.html#cb55-26" tabindex="-1"></a>  beta_df <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(</span>
<span id="cb55-27"><a href="10-Beta_diversity.html#cb55-27" tabindex="-1"></a>    rbiom<span class="sc">::</span><span class="fu">unifrac</span>(</span>
<span id="cb55-28"><a href="10-Beta_diversity.html#cb55-28" tabindex="-1"></a>      <span class="at">biom =</span> phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_rarefy),</span>
<span id="cb55-29"><a href="10-Beta_diversity.html#cb55-29" tabindex="-1"></a>      <span class="at">tree =</span> phyloseq<span class="sc">::</span><span class="fu">phy_tree</span>(pseq_rarefy),</span>
<span id="cb55-30"><a href="10-Beta_diversity.html#cb55-30" tabindex="-1"></a>      <span class="at">weighted =</span> <span class="cn">TRUE</span>))</span>
<span id="cb55-31"><a href="10-Beta_diversity.html#cb55-31" tabindex="-1"></a>  <span class="co">#Add/sum the new data frame values to the sum data frame</span></span>
<span id="cb55-32"><a href="10-Beta_diversity.html#cb55-32" tabindex="-1"></a>  beta_df_sum <span class="ot">&lt;-</span> beta_df_sum <span class="sc">+</span> beta_df</span>
<span id="cb55-33"><a href="10-Beta_diversity.html#cb55-33" tabindex="-1"></a>}</span>
<span id="cb55-34"><a href="10-Beta_diversity.html#cb55-34" tabindex="-1"></a><span class="co">#Divide by number of rarefaction iterations to get average</span></span>
<span id="cb55-35"><a href="10-Beta_diversity.html#cb55-35" tabindex="-1"></a>beta_df_mean <span class="ot">&lt;-</span> beta_df_sum <span class="sc">/</span> rarefaction_iters</span>
<span id="cb55-36"><a href="10-Beta_diversity.html#cb55-36" tabindex="-1"></a><span class="co">#Save alpha mean data frame</span></span>
<span id="cb55-37"><a href="10-Beta_diversity.html#cb55-37" tabindex="-1"></a><span class="fu">save</span>(beta_df_mean, <span class="at">file =</span> <span class="st">&quot;wunifrac_df_mean.RData&quot;</span>)</span>
<span id="cb55-38"><a href="10-Beta_diversity.html#cb55-38" tabindex="-1"></a><span class="co">#View first 6 rows and columns of matrix</span></span>
<span id="cb55-39"><a href="10-Beta_diversity.html#cb55-39" tabindex="-1"></a>beta_df_mean[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb55-40"><a href="10-Beta_diversity.html#cb55-40" tabindex="-1"></a><span class="co">#Remove unneeded objects</span></span>
<span id="cb55-41"><a href="10-Beta_diversity.html#cb55-41" tabindex="-1"></a><span class="fu">rm</span>(beta_df_sum, beta_df_mean, pseq_rarefy)</span></code></pre></div>
<div class="webex-solution">
<button>
verbose = FALSE option
</button>
<p>We include the option <code>verbose = FALSE</code> in the <code>phyloseq::rarefy_even_depth()</code> to prevent a lot of text to be displayed.
This text says which rngseed was used in the rarefaction.
We don't need this message as we already have a record of the rngseeds we used in <code>rngseed_vec</code>.</p>
</div>
<p>You'll notice that each value is duplicated, once in the bottom left triangle and once in the top right triangle.
This is fine as the subsequent functions in this chapter will work with this in the same manner as if the matrix was de-duplicated.</p>
<p>In the above case we saved our final distance matrix and removed it.
We will then load the object in the next section.
This is convenient as it means we only need to run this code cell once.
With higher numbers of rarefaction iterations it can take a while.</p>
</div>
</div>
<div id="beta-ordination" class="section level2 hasAnchor" number="10.4">
<h2><span class="header-section-number">10.4</span> <span class="math inline">\(\beta\)</span>: Ordination<a href="10-Beta_diversity.html#beta-ordination" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/navigation.png" style="width:200px" />
</center>
<p>There are various ways to ordinate paired dissimilarity distances.
We are going to use two of the most popular for community based data: <strong>NMDS</strong> and <strong>PCoA</strong>.</p>
<div id="load-weighted-unifrac-matrix" class="section level3 hasAnchor" number="10.4.1">
<h3><span class="header-section-number">10.4.1</span> Load weighted unifrac matrix<a href="10-Beta_diversity.html#load-weighted-unifrac-matrix" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<center>
<img src="figures/loading.png" style="width:300px" />
</center>
<p>Prior to ordination we need to load the weighted unifrac matrix.
This is a useful cell to have as we don't need to rerun the <strong>iterative rarefaction</strong> to reacquire this object if we save, close + halt, and then reopen this notebook.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb56-1"><a href="10-Beta_diversity.html#cb56-1" tabindex="-1"></a><span class="co">#Load wunifrac matrix</span></span>
<span id="cb56-2"><a href="10-Beta_diversity.html#cb56-2" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;wunifrac_df_mean.RData&quot;</span>)</span></code></pre></div>
</div>
<div id="nmds" class="section level3 hasAnchor" number="10.4.2">
<h3><span class="header-section-number">10.4.2</span> NMDS<a href="10-Beta_diversity.html#nmds" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The first ordination technique we will use is <strong>NMDS</strong> (Non-metric MultiDimensional Scaling).
To carry this out we can use the function <code>vegan::monoMDS()</code>.</p>
<p>We provide the function with our <strong>beta diversity dissimilarity matrix</strong>.
Additionally, we provide it with the parameter <code>k = 2</code>.
This is the number of dimensions we want, we want 2 as we will only be plotting an x and y axis i.e. 2 dimensions.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb57-1"><a href="10-Beta_diversity.html#cb57-1" tabindex="-1"></a><span class="co">#NMDS ordinate</span></span>
<span id="cb57-2"><a href="10-Beta_diversity.html#cb57-2" tabindex="-1"></a>nmds_res <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">monoMDS</span>(beta_df_mean, <span class="at">k =</span> <span class="dv">2</span>)</span>
<span id="cb57-3"><a href="10-Beta_diversity.html#cb57-3" tabindex="-1"></a><span class="co">#Structure of nmds_res</span></span>
<span id="cb57-4"><a href="10-Beta_diversity.html#cb57-4" tabindex="-1"></a><span class="fu">str</span>(nmds_res)</span></code></pre></div>
<p>The function produces a <strong>list</strong> containing 28 different objects.
Here we are only interested in the <code>points</code> data frame.
This contains the points we will be plotting in our NMDS plot.
Lets extract this data frame.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb58-1"><a href="10-Beta_diversity.html#cb58-1" tabindex="-1"></a><span class="co">#Extract plot points</span></span>
<span id="cb58-2"><a href="10-Beta_diversity.html#cb58-2" tabindex="-1"></a>nmds_points <span class="ot">&lt;-</span> nmds_res<span class="sc">$</span>points</span>
<span id="cb58-3"><a href="10-Beta_diversity.html#cb58-3" tabindex="-1"></a><span class="co">#Check head</span></span>
<span id="cb58-4"><a href="10-Beta_diversity.html#cb58-4" tabindex="-1"></a><span class="fu">head</span>(nmds_points)</span></code></pre></div>
<p>Before plotting with <code>ggplot2</code> we need to combine the metadata and points into a long data frame.
There is no need to "longify" our data frame as it will start in the long format we require.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb59-1"><a href="10-Beta_diversity.html#cb59-1" tabindex="-1"></a><span class="co">#Create long data frame with metadata and points</span></span>
<span id="cb59-2"><a href="10-Beta_diversity.html#cb59-2" tabindex="-1"></a><span class="co">#Extract metadata and ensure row names order matches</span></span>
<span id="cb59-3"><a href="10-Beta_diversity.html#cb59-3" tabindex="-1"></a>metadf <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)</span>
<span id="cb59-4"><a href="10-Beta_diversity.html#cb59-4" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">identicial</span>(<span class="fu">row.names</span>(metadf), <span class="fu">row.names</span>(nmds_points)) <span class="sc">==</span> <span class="cn">FALSE</span>) {</span>
<span id="cb59-5"><a href="10-Beta_diversity.html#cb59-5" tabindex="-1"></a>  metadf <span class="ot">&lt;-</span> metadf[<span class="fu">row.names</span>(nmds_points),]</span>
<span id="cb59-6"><a href="10-Beta_diversity.html#cb59-6" tabindex="-1"></a>}</span>
<span id="cb59-7"><a href="10-Beta_diversity.html#cb59-7" tabindex="-1"></a><span class="co">#Make points with metadata data frame</span></span>
<span id="cb59-8"><a href="10-Beta_diversity.html#cb59-8" tabindex="-1"></a>nmds_points_metadata <span class="ot">&lt;-</span> <span class="fu">cbind</span>(nmds_points, metadf)</span>
<span id="cb59-9"><a href="10-Beta_diversity.html#cb59-9" tabindex="-1"></a><span class="co">#Change colnames MDS1 and MDS2 to NMDS1 and NMDS2</span></span>
<span id="cb59-10"><a href="10-Beta_diversity.html#cb59-10" tabindex="-1"></a><span class="fu">colnames</span>(nmds_points_metadata)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;NMDS1&quot;</span>,<span class="st">&quot;NMDS2&quot;</span>)</span>
<span id="cb59-11"><a href="10-Beta_diversity.html#cb59-11" tabindex="-1"></a><span class="co">#Check head of data frame</span></span>
<span id="cb59-12"><a href="10-Beta_diversity.html#cb59-12" tabindex="-1"></a><span class="fu">head</span>(nmds_points_metadata)</span>
<span id="cb59-13"><a href="10-Beta_diversity.html#cb59-13" tabindex="-1"></a><span class="co">#Save object we want, remove ones we don&#39;t</span></span>
<span id="cb59-14"><a href="10-Beta_diversity.html#cb59-14" tabindex="-1"></a><span class="fu">save</span>(nmds_points_metadata, <span class="at">file =</span> <span class="st">&quot;wunifrac_NMDS.RData&quot;</span>)</span>
<span id="cb59-15"><a href="10-Beta_diversity.html#cb59-15" tabindex="-1"></a><span class="fu">rm</span>(metadf, nmds_res, nmds_points)</span></code></pre></div>
<p>Now we can produce a NMDS scatter plot.
We will colour the points by site and the point shapes will be determined by the media.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r Rchunk"><code class="sourceCode r"><span id="cb60-1"><a href="10-Beta_diversity.html#cb60-1" tabindex="-1"></a><span class="co">#Produce NMDS scatter plot</span></span>
<span id="cb60-2"><a href="10-Beta_diversity.html#cb60-2" tabindex="-1"></a><span class="co">#Plot ordination</span></span>
<span id="cb60-3"><a href="10-Beta_diversity.html#cb60-3" tabindex="-1"></a>nmds.wunifrac <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> nmds_points_metadata,</span>
<span id="cb60-4"><a href="10-Beta_diversity.html#cb60-4" tabindex="-1"></a>                        <span class="fu">aes</span>(<span class="at">x =</span> NMDS1, <span class="at">y =</span> NMDS2, <span class="at">color =</span> site, <span class="at">shape =</span> media)) <span class="sc">+</span></span>
<span id="cb60-5"><a href="10-Beta_diversity.html#cb60-5" tabindex="-1"></a>                ggplot2<span class="sc">::</span><span class="fu">geom_point</span>()</span>
<span id="cb60-6"><a href="10-Beta_diversity.html#cb60-6" tabindex="-1"></a><span class="co">#Save ggplot2 object with ggsave</span></span>
<span id="cb60-7"><a href="10-Beta_diversity.html#cb60-7" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">&quot;./Beta_diversity_NMDS_wunifrac_media_site.png&quot;</span>,</span>
<span id="cb60-8"><a href="10-Beta_diversity.html#cb60-8" tabindex="-1"></a>       <span class="at">plot =</span> nmds.wunifrac,</span>
<span id="cb60-9"><a href="10-Beta_diversity.html#cb60-9" tabindex="-1"></a>       <span class="at">device =</span> <span class="st">&quot;png&quot;</span>, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">125</span>, <span class="at">width =</span> <span class="dv">150</span>)</span>
<span id="cb60-10"><a href="10-Beta_diversity.html#cb60-10" tabindex="-1"></a><span class="co">#Display plot</span></span>
<span id="cb60-11"><a href="10-Beta_diversity.html#cb60-11" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file =</span> <span class="st">&quot;./Beta_diversity_NMDS_wunifrac_media_site.png&quot;</span>)</span></code></pre></div>
<p>Super!
You should see that the plot is mainly separated by NMDS1 splitting the ENV samples compared to the media samples.
The next major difference is in NMDS2 which separates the TSA samples compared to the CVP and KBC samples.
The CVP and KBC samples also appear to form distinct clusters.</p>
</div>
<div id="pcoa" class="section level3 hasAnchor" number="10.4.3">
<h3><span class="header-section-number">10.4.3</span> PCoA<a href="10-Beta_diversity.html#pcoa" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The second, and last, ordination method we will look at is <strong>PCoA</strong> (Principal Correspondence Analysis).</p>
</div>
<div id="which-ordination-method-to-use" class="section level3 hasAnchor" number="10.4.4">
<h3><span class="header-section-number">10.4.4</span> Which ordination method to use?<a href="10-Beta_diversity.html#which-ordination-method-to-use" class="anchor-section" aria-label="Anchor link to header"></a></h3>

</div>
</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="09-Alpha_diversity.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
