# Data & Setup
<center>
![](images/setup_8.png){style="width:200px; background: white; border-radius:5px; border: white solid 5px"}
</center>


Before you carry out any practice you will need to setup your environment  with the instructions in this page. The steps are:

- Ensuring you have setup your directory, environment, and jupyter-notebook as shown in the overall [setup](/setup/setup.qmd).
- Creating a jupyter-notebook in a new work directory.
- Loading in the required libraries.
- Loading in the data.

## Jupyter notebook
<center>
![](images/jupyter_logo.png){style="width:200px; background: white; border-radius:5px; border: white solid 5px"}
</center>

In the jupyter-notebook file explorer create and move into a new directory called `divergence`, ensure this is a subdiretcory of the main workshop directory (`adv_R_comm`).
We'll keep each section in its own subdirectory so it is easier to keep the output files separate and tidy.

Create a new R based jupyter-notebook called "divergence.ipynb".
At the top of the jupyter-notebook add the following headers:

- A first level header of "Divergence".
- A second level header of "Libraries & data".

## Libraries
<center>
![](images/packages.png){style="width:200px; background: white; border-radius:5px; border: white solid 5px"}
</center>

Load the required libraries/packages in a new code cell.

```{r, eval=FALSE}
#Libraries
library("phyloseq")
library("microbiome")
library("IRdisplay")
library("tidyverse")
library("ggforce")
library("patchwork")
```

## Data

<center>
![](images/river.png){style="width:200px"}
</center>

We have a phyloseq object containing ASV abundance data through DADA2 denoising in QIIME2 ([theory](https://neof-workshops.github.io/16S_yg84o9/Course/03-Intro_to_qiime2.html#amplicon-sequence-variants-identification) & [code](https://neof-workshops.github.io/16S_yg84o9/Course/09-ASV_identification.html)).
For our divergence analysis we will be utilising the Durance river data used in our R community analysis workshop.
This is a 16S dataset with two main metadata groupings:

- Sites: Upper, Middle, and Lower Durance sampling site (UD, MD, & LD).
- Culture media: Environmental sample (ENV), TSA, KBC, and CVP.

More info in the [R community analysis workbook](https://neof-workshops.github.io/R_community_whqkt8/Course/02-Dataset_and_workflow.html#dataset)

Load in the phyloseq object.

```{r, eval=FALSE}
#Original phyloseq object
load("/pub14/tea/nsc206/NEOF/R_community/data/phyloseq.RData")
```

Additionally, we will load a [dissimilarity matrix of weighted unifrac values created through iterative rarefaction](https://neof-workshops.github.io/R_community_whqkt8/iterative_rarefaction/20-Beta_rarefaction.html):

```{r, eval=FALSE}
#Iterative rarefied weighted unifrac data frame
load("/pub14/tea/nsc206/NEOF/R_community_advanced/wunifrac_df_mean.RData")
```